{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import subprocess\n",
    "from scipy import sparse\n",
    "import numpy as np\n",
    "\n",
    "portal_obs_fields = [\n",
    "    'assay',\n",
    "    'cell_type',\n",
    "    'development_stage',\n",
    "    'disease',\n",
    "    'self_reported_ethnicity',\n",
    "    'organism',\n",
    "    'sex',\n",
    "    'tissue'\n",
    "]\n",
    "full_obs_standards = portal_obs_fields + [e + '_ontology_term_id' for e in portal_obs_fields] + ['donor_id','suspension_type','is_primary_data']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_obs_standards"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to load the following libraries. Make sure that the anndata version you have installed matches the one used by [cellxgene-schema](https://github.com/chanzuckerberg/single-cell-curation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: anndata\n",
      "Version: 0.8.0\n",
      "Summary: Annotated data.\n",
      "Home-page: \n",
      "Author: Philipp Angerer, Alex Wolf, Isaac Virshup, Sergei Rybakov\n",
      "Author-email: \n",
      "License: \n",
      "Location: /opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages\n",
      "Requires: h5py, natsort, numpy, packaging, pandas, scipy\n",
      "Required-by: cellxgene-schema, scanpy\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip show anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: cellxgene-schema\n",
      "Version: 3.0.1\n",
      "Summary: Tool for applying and validating cellxgene integration schema to single cell datasets\n",
      "Home-page: https://github.com/chanzuckerberg/single-cell-curation\n",
      "Author: Chan Zuckerberg Initiative\n",
      "Author-email: cellxgene@chanzuckerberg.com\n",
      "License: MIT\n",
      "Location: /opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages\n",
      "Requires: anndata, click, coverage, Cython, numpy, owlready2, pandas, pytest, PyYaml, wheel\n",
      "Required-by: \n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip show cellxgene-schema \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading the AnnData object\n",
    "**Update the name of the file (without the .h5ad extension)**<br>\n",
    "*The sample `my_matrix.h5ad` that is in this repo is subsampled from https://cellxgene.cziscience.com/e/f15e263b-6544-46cb-a46e-e33ab7ce8347.cxg/ with some metadata alterations for the purpose of this tutorial*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = 'lung_5loc_sc_sn_cellxgene_23092022'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Load the AnnData object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<193108x20922 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 277931444 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(file + '.h5ad')\n",
    "adata.raw.X "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# data layers\n",
    "**If needed, transfer to sparse matrix format, this saves space**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "count_nonzero = counts the number of non-zero values in the array "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/4r/275m7hv11jv8zl5946vzl1c00000gp/T/ipykernel_78384/963552313.py:2: DeprecationWarning: Please use `csc_matrix` from the `scipy.sparse` namespace, the `scipy.sparse.csc` namespace is deprecated.\n",
      "  if isinstance(x, sparse.coo_matrix) or isinstance(x, sparse.csr_matrix) or isinstance(x, sparse.csc.csc_matrix):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X sparsity: 0.0\n",
      "raw.X sparsity: 0.9312085885799986\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/4r/275m7hv11jv8zl5946vzl1c00000gp/T/ipykernel_78384/963552313.py:23: DeprecationWarning: Please use `csr_matrix` from the `scipy.sparse` namespace, the `scipy.sparse.csr` namespace is deprecated.\n",
      "  if sparsity > max_sparsity and type(adata.raw.X) != sparse.csr.csr_matrix:\n"
     ]
    }
   ],
   "source": [
    "def determine_sparsity(x):\n",
    "    if isinstance(x, sparse.coo_matrix) or isinstance(x, sparse.csr_matrix) or isinstance(x, sparse.csc.csc_matrix):\n",
    "        sparsity = 1 - x.count_nonzero() / float(np.cumprod(x.shape)[-1])\n",
    "    elif isinstance(x, np.ndarray):\n",
    "        sparsity = 1 - np.count_nonzero(x) / float(np.cumprod(x.shape)[-1])\n",
    "    else:\n",
    "        print(f'matrix is of type {type(x)}, sparsity calculation has not been implemented')\n",
    "\n",
    "    return sparsity\n",
    "\n",
    "\n",
    "max_sparsity = 0.5\n",
    "\n",
    "sparsity = determine_sparsity(adata.X)\n",
    "print(f'X sparsity: {sparsity}')\n",
    "if sparsity > max_sparsity and type(adata.X) != sparse.csr.csr_matrix:\n",
    "    print('converting X to sparse')\n",
    "    adata.X = sparse.csr_matrix(adata.X)\n",
    "\n",
    "if adata.raw:\n",
    "    sparsity = determine_sparsity(adata.raw.X)\n",
    "    print(f'raw.X sparsity: {sparsity}')\n",
    "    if sparsity > max_sparsity and type(adata.raw.X) != sparse.csr.csr_matrix:\n",
    "        print('converting raw.X to sparse')\n",
    "        raw_adata = ad.AnnData(adata.raw.X, var=adata.raw.var, obs=adata.obs)\n",
    "        raw_adata.X = sparse.csr_matrix(raw_adata.X)\n",
    "        adata.raw = raw_adata\n",
    "        del raw_adata\n",
    "\n",
    "for l in adata.layers:\n",
    "    sparsity = determine_sparsity(adata.layers[l])\n",
    "    print(f'layers[{l}] sparsity: {sparsity}')\n",
    "    if sparsity > max_sparsity and type(adata.layers[l]) != sparse.csr.csr_matrix:\n",
    "        print(f'converting layers[{l}] to sparse')\n",
    "        adata.layers[l] = sparse.csr_matrix(adata.layers[l])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Check the min/max of each layer**<br>\n",
    "*Look for duplicated or other unnecessary layers*<br>\n",
    "*Raw should be whole, positive, ~10<sup>3*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raw min = 0.0\n",
      "raw max = 9.026224\n",
      "ERROR: raw contains non-integer values\n",
      "X min = -7.6582603\n",
      "X max = 10.0\n"
     ]
    }
   ],
   "source": [
    "if adata.raw:\n",
    "    print('raw min = ' + str(adata.raw.X.min()))\n",
    "    print('raw max = ' + str(adata.raw.X.max()))\n",
    "    non_integer = np.any(~np.equal(np.mod(adata.raw.X.data, 1), 0))\n",
    "else:\n",
    "    non_integer = np.any(~np.equal(np.mod(adata.X.data, 1), 0))\n",
    "\n",
    "if non_integer == False:\n",
    "    print('raw is all integers')\n",
    "else:\n",
    "    print('ERROR: raw contains non-integer values')\n",
    "\n",
    "print('X min = ' + str(adata.X.min()))\n",
    "print('X max = ' + str(adata.X.max()))\n",
    "\n",
    "for l in adata.layers:\n",
    "    print(f'layers[{l}] min = ' + str(adata.layers[l].min()))\n",
    "    print(f'layers[{l}] max = ' + str(adata.layers[l].max()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** delete a layer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata.layers['other_raw']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# obsm\n",
    "**Confirm at least one set of embeddings is present**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AxisArrays with keys: X_pca, X_pca_hm, X_umap, X_umap_BBKNN_scDonor_snBatch, X_umap_Harmony_Donor_n_Material, X_umap_Harmony_Material, X_umap_Harmony_scDonor_snBatch"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obsm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**View embeddings to identify which matches paper figures**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Celltypes_colors'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/pandas/core/indexes/base.py:3629\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3628\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3629\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3630\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/pandas/_libs/index.pyx:136\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/pandas/_libs/index.pyx:163\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Celltypes_colors'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [9]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m sc\u001b[38;5;241m.\u001b[39mset_figure_params(dpi\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m100\u001b[39m)\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m e \u001b[38;5;129;01min\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mobsm:\n\u001b[0;32m----> 3\u001b[0m     \u001b[43msc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membedding\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbasis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43me\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mCelltypes_colors\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlegend_loc\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mon data\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:257\u001b[0m, in \u001b[0;36membedding\u001b[0;34m(adata, basis, color, gene_symbols, use_raw, sort_order, edges, edges_width, edges_color, neighbors_key, arrows, arrows_kwds, groups, components, dimensions, layer, projection, scale_factor, color_map, cmap, palette, na_color, na_in_legend, size, frameon, legend_fontsize, legend_fontweight, legend_loc, legend_fontoutline, colorbar_loc, vmax, vmin, vcenter, norm, add_outline, outline_width, outline_color, ncols, hspace, wspace, title, show, save, ax, return_fig, **kwargs)\u001b[0m\n\u001b[1;32m    250\u001b[0m \u001b[38;5;66;03m# use itertools.product to make a plot for each color and for each component\u001b[39;00m\n\u001b[1;32m    251\u001b[0m \u001b[38;5;66;03m# For example if color=[gene1, gene2] and components=['1,2, '2,3'].\u001b[39;00m\n\u001b[1;32m    252\u001b[0m \u001b[38;5;66;03m# The plots are: [\u001b[39;00m\n\u001b[1;32m    253\u001b[0m \u001b[38;5;66;03m#     color=gene1, components=[1,2], color=gene1, components=[2,3],\u001b[39;00m\n\u001b[1;32m    254\u001b[0m \u001b[38;5;66;03m#     color=gene2, components = [1, 2], color=gene2, components=[2,3],\u001b[39;00m\n\u001b[1;32m    255\u001b[0m \u001b[38;5;66;03m# ]\u001b[39;00m\n\u001b[1;32m    256\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m count, (value_to_plot, dims) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[38;5;28mzip\u001b[39m(color, dimensions)):\n\u001b[0;32m--> 257\u001b[0m     color_source_vector \u001b[38;5;241m=\u001b[39m \u001b[43m_get_color_source_vector\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    258\u001b[0m \u001b[43m        \u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    259\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalue_to_plot\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    260\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlayer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlayer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    261\u001b[0m \u001b[43m        \u001b[49m\u001b[43muse_raw\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_raw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    262\u001b[0m \u001b[43m        \u001b[49m\u001b[43mgene_symbols\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_symbols\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    263\u001b[0m \u001b[43m        \u001b[49m\u001b[43mgroups\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroups\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    264\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    265\u001b[0m     color_vector, categorical \u001b[38;5;241m=\u001b[39m _color_vector(\n\u001b[1;32m    266\u001b[0m         adata,\n\u001b[1;32m    267\u001b[0m         value_to_plot,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    270\u001b[0m         na_color\u001b[38;5;241m=\u001b[39mna_color,\n\u001b[1;32m    271\u001b[0m     )\n\u001b[1;32m    273\u001b[0m     \u001b[38;5;66;03m# Order points\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:1167\u001b[0m, in \u001b[0;36m_get_color_source_vector\u001b[0;34m(adata, value_to_plot, use_raw, gene_symbols, layer, groups)\u001b[0m\n\u001b[1;32m   1163\u001b[0m     value_to_plot \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mindex[adata\u001b[38;5;241m.\u001b[39mvar[gene_symbols] \u001b[38;5;241m==\u001b[39m value_to_plot][\n\u001b[1;32m   1164\u001b[0m         \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   1165\u001b[0m     ]  \u001b[38;5;66;03m# TODO: Throw helpful error if this doesn't work\u001b[39;00m\n\u001b[1;32m   1166\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m use_raw \u001b[38;5;129;01mand\u001b[39;00m value_to_plot \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[0;32m-> 1167\u001b[0m     values \u001b[38;5;241m=\u001b[39m \u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mraw\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobs_vector\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalue_to_plot\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1168\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1169\u001b[0m     values \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mobs_vector(value_to_plot, layer\u001b[38;5;241m=\u001b[39mlayer)\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/anndata/_core/raw.py:171\u001b[0m, in \u001b[0;36mRaw.obs_vector\u001b[0;34m(self, k)\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mobs_vector\u001b[39m(\u001b[38;5;28mself\u001b[39m, k: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m np\u001b[38;5;241m.\u001b[39mndarray:\n\u001b[1;32m    170\u001b[0m     \u001b[38;5;66;03m# TODO decorator to copy AnnData.obs_vector docstring\u001b[39;00m\n\u001b[0;32m--> 171\u001b[0m     idx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_normalize_indices\u001b[49m\u001b[43m(\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mslice\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    172\u001b[0m     a \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mX[idx]\n\u001b[1;32m    173\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m issparse(a):\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/anndata/_core/raw.py:162\u001b[0m, in \u001b[0;36mRaw._normalize_indices\u001b[0;34m(self, packed_index)\u001b[0m\n\u001b[1;32m    160\u001b[0m obs, var \u001b[38;5;241m=\u001b[39m unpack_index(packed_index)\n\u001b[1;32m    161\u001b[0m obs \u001b[38;5;241m=\u001b[39m _normalize_index(obs, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_adata\u001b[38;5;241m.\u001b[39mobs_names)\n\u001b[0;32m--> 162\u001b[0m var \u001b[38;5;241m=\u001b[39m \u001b[43m_normalize_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_names\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m obs, var\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/anndata/_core/index.py:75\u001b[0m, in \u001b[0;36m_normalize_index\u001b[0;34m(indexer, index)\u001b[0m\n\u001b[1;32m     73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m indexer\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(indexer, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m---> 75\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# int\u001b[39;00m\n\u001b[1;32m     76\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(indexer, (Sequence, np\u001b[38;5;241m.\u001b[39mndarray, pd\u001b[38;5;241m.\u001b[39mIndex, spmatrix, np\u001b[38;5;241m.\u001b[39mmatrix)):\n\u001b[1;32m     77\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mshape\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[1;32m     78\u001b[0m         (indexer\u001b[38;5;241m.\u001b[39mshape \u001b[38;5;241m==\u001b[39m (index\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m], \u001b[38;5;241m1\u001b[39m))\n\u001b[1;32m     79\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m (indexer\u001b[38;5;241m.\u001b[39mshape \u001b[38;5;241m==\u001b[39m (\u001b[38;5;241m1\u001b[39m, index\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]))\n\u001b[1;32m     80\u001b[0m     ):\n",
      "File \u001b[0;32m/opt/anaconda3/envs/cellxgene/lib/python3.8/site-packages/pandas/core/indexes/base.py:3631\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3629\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3630\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3631\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3632\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3633\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3634\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3635\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3636\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Celltypes_colors'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvwAAALNCAYAAABTfrWHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAB7CAAAewgFu0HU+AABAsklEQVR4nO3de5gtZ10n+u+P7L3DJYEAEnMnQsQ84CWjwBEHJDEHAkpUBi9cFHAUPB4OGJVBjgoM92EUGJgRFFEBJahBGYwIBNHo6HAfRLkECWZDAkNCwiHZSUj23sl7/lir2bVX+rq6V6/utz+f56knVbXeql9lvd1rf7tW1VvVWgsAANCn2837AAAAgNkR+AEAoGMCPwAAdEzgBwCAjgn8AADQMYEfAAA6JvADAEDHBH4AAOiYwA8AAB0T+AEAoGMCPwAAdEzgBwCAjgn8AADQMYEfAAA6JvADAEDHBH4AAOiYwA8AAB0T+AEAoGMCPwAAdGyugb+q7llVN1ZVq6onz7DO7arqh6vqgqr6XFV9raq+WlWfqKrXV9X3zqo2AADM0655Fa6qI5K8IckdZlznlCRvSfI9Ey/dPsldktw3yU9X1VuT/Gxr7SuzPB4AANhM8zzD/5okZ86yQFWdnOTvc3jYb0muTPLVieY/kuS9VXWnWR4TAABspk0P/FV1RFX9VpKnzrjO7ZL8cZKTx6taklclOam1dlxr7a5JviPJXw42OyPJ78zyuAAAYDNtauCvqm9M8t4kP7sJ5Z6Y5EGD5V9srZ3XWvviworW2j8leVSS3xu0e1xVPXATjg8AAGZu0wJ/VT0qyUeTPHSTSj5zMP8/Wmv/ZbFGrbWW5P9K8unB6l+b4XEBAMCmmXngr6rvqKq/SnJhkuMHL83s0pmq+jdJ7jdY9erl2rfWDmR0uc+CR1TVMTM4NAAA2FSbcYb/jUnOHizfkOTnWmuzvIb/EYP5g0neuYptLhzM705y7oYeEQAAzMFm37T7jiTf2lr7rRnXecBg/uOttRtW2qC1dkWSLw5WTQ7jCQAA285mBP6W5OIkZ7XWHtVa27sJNe87mP/UGrb7zGD+9A06FgAAmJvNePDWD7bWLt+EOkMnDeavWMN2XxjMn7JBxwIAAHMz8zP8mx32q+qOSYYPz7pmDZsPn7J7t405IgAAmJ/NOMO/2Y6aWL5+DdsO207uZ1Wq6qQVmuzJ6HKhq5J8Ockt09QBAGDTHJHkHuP5f26t3TzPg1mrHgP/nonlA2vYdth22vdmsy9fAgBg8zwgyYfnfRBrsdmj9GyGmlhuU++oanJfAACwrfR4hn//xPLuNWw7bHvz+Cm8a3XyCq+fmOT9SfK3f/u3Oemkla4AYjs7ePBgLr300iTJaaedll27evyVY0if7zz6fOfR5zvPFVdckYc+9KELi1+e57FMo8ef0Osmlu+0aKvFDa/bX3Hs/sWMx/Nf0vBLg5NOOin3ute9pinDNrF///7s27cvSXLqqadmz57JK87ojT7fefT5zqPPd7xtd/9ld5f0tNa+luTGwaq1jLYzbHvVxhwRAADMT3eBf2zvYP6ENWx34mB+LeP3AwDAltRr4B8+Xfc+a9hu2PYTG3QsAAAwN70G/g8N5r+jqla8uG48fv7xS+wDAAC2pV4D/0WD+TskOXsV25w7mG9J/mpDjwgAAOagy8DfWvtokk8PVv3Scu2raneSZwxWXdRau3IWxwYAAJupy8A/9srB/NlV9Zxl2r42yemD5ZfP5pAAAGBzbbvAX1VtMO1dpunvJvnHwfILqur3q+rUwb7uV1V/keSnB+3e1lp7z0YeMwAAzMu2C/yr1Vo7mORHcvjwmk9OcllVXVlVVyf5eJIfGLz+yST/ftMOEgAAZqzbwJ8krbXPJnlokv858dKxSe4+se7dSR7aWvvqJhwaAABsil3zPoBZa639a1U9OMkPJ/mxJA9MclxGf+z87yTvS/Km1tq753aQAAAwI3MN/K212qRtWpK3jScAANgxur6kBwAAdjqBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAd29TAX1VHVdXTquqiqrqyqvaP//uRqnpxVX3zDGvfrqoeXVVvrqpLq2pfVd1UVV+oqndW1XlVdedZ1QcAgHnYtVmFquqsJG9KctLES8eOp+9M8qyqemGSF7fWbtnA2t+S5I+SnLHIyyeMp0ckeV5VPbW1dsFG1QYAgHnalDP8VXVOknfm8LB/MMkXktw4WLcryfOTvHYDa98vyftz27B/dZIrkgz/sDgmyZ9U1dM3qj4AAMzTzAN/VZ2Q5PwkR45X7UvytCTHtNZOSnJ0kkcm+ZfBZk+pqqduQO3dSf4soyC/4A1JTmut3aO1dnKSuyU5L4f/4fGqqvru9dYHAIB524wz/C/JKFQnyU1Jzmmtvaa1dkOStNZuba29K8kDknxssN0Lq+qoddZ+YpL7DJZf0Fr7qdbaZxdWtNaua629KsnZSfaPV9f4uAEAYFubaeCvquOTPGGw6tdba+9brG1r7bokj0lyYLzq2CTrPcv/Q4P5zyd5wVINW2vvT/I7g1UPraq7LdUeAAC2g1mf4X9sDt0YfGuS31yu8fjM+9sGqx63zvqnDebft4obgd87mL9dkm9aZ30AAJirWQf+RwzmP9hau3IV21w4mL9/VZ2yjvpHDubvMMX2nlMAAMC2NutAe//B/PtXuc0HJpYftI76lw7mH1JVd1qh/cMH8weTfHodtQEAYO5mFvir6rgculk3ST61yk0vyyhsLzh9HYfxp4P5uyZ5ZVXVYg3Hzwn46cGqt47vKwAAgG1rlmf4Jx+wdcVqNmqtHUxy1WDVei7p+f0kHx4sPyXJe6rqYVV156raXVWnV9WLkrwrye5xu8uT/NI66gIAwJYwyyftHjuxfM0atv1KRk+/TQ7/lmBNWmsHqur7k/xJkjPHq88eT0t5Z5KntNa+OE3Nqpr8Q2fScQszBw8ezP79+5dryzZ34MCBRefplz7fefT5zqPPd56DBw+u3GgLm2XgnxxD//o1bDtsu66x+FtrX05yVlX9X0lemeT2yzT/8yQ/N23YH7t8tQ0vvfTS7Nu3bx2l2E4uueSSeR8Cm0yf7zz6fOfR5zvDlVeuZtyZrWuWl/TsmVhey5/Aw7a7l2y1ClV1bFW9Kcl/y6Gw35J8OckXc/j9Aj+Y5DNV9YvrqQkAAFvFLM/wT94c22ZYa/EDqDotyd/m0OVB+5K8OMkbFoYIHY/cc25GD+X65iR3TPLyqjq+tfYfpih78gqvH5fkQ0ly2mmn5dRTT52iBNvFgQMHvn725/TTT8/u3ev6+5VtQJ/vPPp859HnO8/RRx8970NYl1kG/smL09fy2zBse9M0xatqT5K351DYvybJma21jw/btdZuSPJHVfWOjC7pOXP80jOr6uLW2jvWUre1tuzNycNBgnbt2pU9eya/CKFXu3fv1t87jD7fefT5zqPPd4Zdu2YZmWdvlpf0TA5pudIY+EPD6/ZvmLL+E5Pcd7D805Nhf6i1ti/JY5JcPVj9/ClrAwDAljDLwH/1xPJaRtsZtr1qyVbL+/HB/Kdba29faYPW2leS/OZg1Xet80m/AAAwV7MM/Hsnlk9YrNGkqtqVw4f0XNX4/Ys4YzD/d2vY7m8mlr99yvoAADB3Mwv84+Ewh2Pv32eVm94rh99b8IkpD+GYwfxX17DdlyeW7zxlfQAAmLtZnuFPxqPRjD1wldtMtvvwoq1W9tXB/Kq+XRi768TyV6asDwAAczfrwH/RYP7BVXXMKrY5dzD/yZVGvVnGpYP576uqI1a53UMmlv9lyvoAADB3sw78FyS5ZTy/J8nTl2tcVfdO8ujBqj9YR+3hcJrHJ/mJlTYYj8n/tMGqT7bW/nUdxwAAAHM108A/Pjt/wWDVc6rq7MXaVtXRSd6aQ2PwX5vk9eso/7qMHrS14NVVteRlRVV1ZJI3JzlpsPpl66gPAABzN+sz/EnyK0muH8/vTvKOqnp2Vd0lSWrk4Rld73/GYLvnttYmh/bMeJu9VdUWpsXatNauSjJ8Uu6dk/yPqnppVd1rsK8jq+pRSd6f5IcG7d+T5A/X8j8KAABbzcwDf2vtsiSPz6En7x6Z5KVJrqmqKzI6C//uJN8y2Oz81tqrN6D2byd58WDVniTPTvLZqrquqj6f0QPCLszhf2x8OMmPttZuXe8xAADAPG3GGf601i5M8qgklw9WH5HkxBz+BN5bk7w8o6fkblTtX0vy2CRfmnjp6CQnZ/RHwIKDGT146yGttWs36hgAAGBedq3cZGO01t5TVacneVJGN+beN8k9ktyc0UO6Lk7yutbax2dQ+4+r6u0ZfdPwyCTfNa69J6NnBXwmyV8neWNrbe9G1wcAgHnZtMCfJK21G5O8djytZz+nTrHNTUl+bzwBAMCOsCmX9AAAAPMh8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHdvUwF9VR1XV06rqoqq6sqr2j//7kap6cVV984zr36WqnlpV76iqf62qG6vq+qr6TFX9YVU9oqpqlscAAACbaddmFaqqs5K8KclJEy8dO56+M8mzquqFSV7cWrtlg+s/IcnLk3zjIi+fNp6ekORvq+onW2uXb2R9AACYh005w19V5yR5Zw4P+weTfCHJjYN1u5I8P8lrN7j+ryb5wxwe9m8d179uovlDk/yPqjpxI48BAADmYeaBv6pOSHJ+kiPHq/YleVqSY1prJyU5Oskjk/zLYLOnVNVTN6j+U5K8aLDq00l+PMntx/WPSfJ/JPnrQZt7Jnn9RtQHAIB52owz/C9Jcrfx/E1Jzmmtvaa1dkOStNZuba29K8kDknxssN0Lq+qo9RSuquOTvHKw6v1Jvqe19iettQPj+q219sEkD0vyl4O2j6iq+6+nPgAAzNtMA/84cD9hsOrXW2vvW6xta+26JI9JcmC86tgk6z3L/6wkdxrPX5PkUa21ryxR/9YkP5ekDVb/yDrrAwDAXM36DP9jc+jG4FuT/OZyjVtrn03ytsGqx01buKr2JHnyYNWzW2vXrFD/8+Nj/JMkr0nyz9PWBwCArWDWo/Q8YjD/wdbalavY5sIkPzaev39VnTIO4mt1ZkbX5yejs/tvXM1GrbWnT1ELAAC2pFmf4R9eA//+VW7zgYnlB01Z+8zB/HsXrtkHAICdZGZn+KvquBy6WTdJPrXKTS/LaMjOhWM7fcpD+K7B/EcHx3WPjEbpOTvJyUmOyGh4zr9K8pZVfgsBAADbwiwv6Zl8wNYVq9motXawqq5KcsJ41SlT1r/PYP7zVXVEkmdmNM7/kRNtz0jyA0leUFXPa629MgAA0IFZXtJz7MTysjfMThiOpHO3JVstb/jgrOuS/GmS/5Tbhv2ho5O8oqp+p6pqyroAALBlzPIM/+QY+tevYdth2zWPxV9Vd0mye7DquRmN85+MLu95aZK/TXJtRt8g/GiSX05y53Gbn0lyaZKXTVF78puNScctzBw8eDD79+9fawm2kQMHDiw6T7/0+c6jz3cefb7zHDx4cN6HsC6zDPx7JpbX8hsxbLt7yVZLu+PE8kLYf0OSp7TWhr32mSQvqao/zehpuwuXEr2oqv64tbZ3jbUvX23DSy+9NPv27Vvj7tmuLrnkknkfAptMn+88+nzn0ec7w5VXbu9bPGd5Sc/kJTFt0VazsdgfCe/PbcP+17XWPp3kJwardmV0zT8AAGxbszzDP3mtylrO1A/b3jRF7cW2ee5SYX9Ba+1vqurvknzveNW5Sf6fNdY+eYXXj0vyoSQ57bTTcuqpp65x92wnBw4c+PrZn9NPPz27d0/zhRXbiT7fefT5zqPPd56jjz563oewLrMM/NdNLN9pDdsOr9u/YYrak9fJ7EvyN6vc9i9zKPCfUlUntta+sNrCrbVlRyMa3gu8a9eu7NkzeeUTvdq9e7f+3mH0+c6jz3cefb4z7No162fVztYsL+m5emJ5LaPtDNtetdbCrbWvJbllsOqylc7uD3x6Yvm4RVsBAMA2MMvAv3di+YTFGk2qql05fEjPVY3fv4jPD+bXcsPw5GhCax4lCAAAtoqZBf7W2pdz+Nj791mq7YR75fBLjT4x5SH802D++DVsd+eJ5WunrA8AAHM3yzP8yfjm1LEHrnKbyXYfnrL2xwbzJ1TVap/Ye9/BfMtoPH4AANiWZh34LxrMP7iqjlnFNucO5j+50k2wy/jzieUfW+V23z+Y/2hrbS0PDAMAgC1l1oH/ghy6eXZPkqcv17iq7p3k0YNVfzBt4dbaR5J8crDqWVW17I3DVXVWkgcNVr1l2voAALAVzDTwj8/OXzBY9ZyqOnuxtlV1dJK35tAY/Ncmef06D+HFg/l7JDm/qhYdHrSq7pnkjYNVX9mA+gAAMFezPsOfJL+SQyPf7E7yjqp6dlXdJUlq5OEZXe9/xmC757bWJof2zHibvVXVFqalCrfWzk/yF4NV5yT5n1X1w1W1e7yvI6vqJzJ6Eu/woVn/obX21bX8jwIAwFYz88DfWrssyeNz6Mm7RyZ5aZJrquqKjB6K9e4k3zLY7PzW2qs36BAen+RvB8vfnuRtSW6sqssz+ibhD3L4ePuvaq393gbVBwCAudmMM/xprV2Y5FFJLh+sPiLJiTn8Cby3Jnl5kiduYO19SR4x3u9wPP5dSU7K6A+QBTcm+fnW2nkbVR8AAOZp054T3Fp7T1WdnuRJGd2Ye9+Mrqu/OaOHdF2c5HWttY/PoPZNSZ5ZVa9N8oQkj0xyapK7ZxTyP5XkL5P8TmvtSxtdHwAA5mXTAn+StNZuTPLa8bSe/Zw65XafTfKC8QQAAN3blEt6AACA+RD4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANAxgR8AADom8AMAQMcEfgAA6JjADwAAHRP4AQCgYwI/AAB0TOAHAICOCfwAANCxTQ38VXVUVT2tqi6qqiurav/4vx+pqhdX1Tdv5vGMj6mq6l1V1cbTkzf7GAAAYFZ2bVahqjoryZuSnDTx0rHj6TuTPKuqXpjkxa21Wzbp0J6e5JxNqgUAAJtqU87wV9U5Sd6Zw8P+wSRfSHLjYN2uJM9P8tpNOq77JnnZZtQCAIB5mHngr6oTkpyf5Mjxqn1JnpbkmNbaSUmOTvLIJP8y2OwpVfXUGR/XniRvTnL7WdYBAIB52owz/C9Jcrfx/E1Jzmmtvaa1dkOStNZuba29K8kDknxssN0Lq+qoGR7Xi5KcMcP9AwDA3M008FfV8UmeMFj166219y3WtrV2XZLHJDkwXnVskpmc5a+qM5P80njx6lnUAACArWDWZ/gfm0M3Bt+a5DeXa9xa+2yStw1WPW6jD6iqjknyxhz6f5/ppUMAADBPsw78jxjMf7C1duUqtrlwMH//qjplg4/pNUkW9vmG1trblmsMAADb2awD//0H8+9f5TYfmFh+0AYdS6rq8Tn0rcHnkvz8Ru0bAAC2opkF/qo6Lodu1k2ST61y08syGrJzwekbdDwn59AlRbcmeeL4vgEAAOjWLM/wTz5g64rVbNRaO5jkqsGqdV/SU1W3y+ihX8eMV72itfZ3690vAABsdbN80u6xE8vXrGHbryQ5YTx/t+UartIzk5w5nv/nJL+2AftcVFVN/qEz6biFmYMHD2b//v2zOhS2gAMHDiw6T7/0+c6jz3cefb7zHDx4cOVGW9gsA//kGPrXr2HbYdt1jcVfVWckeeF4cX+Sn2yt3byefa7g8tU2vPTSS7Nv374ZHgpbySWXXDLvQ2CT6fOdR5/vPPp8Z7jyytWMO7N1zfKSnj0Ty2v5E3jYdve0B1BVt0/yh4NjeW5r7WPLbAIAAF2Z5Rn+mlhuM6y1lJclud94/u+T/Pom1Dx5hdePS/KhJDnttNNy6qmnzvyAmJ8DBw58/ezP6aefnt27p/77lW1Cn+88+nzn0ec7z9FHHz3vQ1iXWQb+yYvT1/LbMGx70zTFq+rhSZ4+XtyX0ag8t06zr7VorS17c3LVob+Ddu3alT17Jr8IoVe7d+/W3zuMPt959PnOo893hl27ZhmZZ2+Wl/RMDnl5pzVsO7xu/4a1Fq6quyf5/Rz6luG81tpla90PAABsd7MM/FdPLK9ltJ1h26uWbLW0386hUX7e3lr7vSn2AQAA294sv5/YO7F8wmKNJlXVrhw+pOeqxu8fbH9KkscMVj2wqi5dwy5eVlXDYTsf2lr7wlqOAQAAtoqZBf7W2per6pokdx+vus8qN71XDj+uT6yx9OS3Fsevcftjc/gfHO7EAQBg25rlJT3JeDSasQeucpvJdh/eoGMBAIAdZ9aB/6LB/IOr6phVbHPuYP6TK416M6m1tre1VmuZJnbxUxOv711LfQAA2EpmHfgvSHLLeH5PDg2TuaiquneSRw9W/cGMjgsAAHaEmQb+8dn5CwarnlNVZy/WtqqOTvLWHLpm/tokr5/l8QEAQO9mfYY/SX4lyfXj+d1J3lFVz66quyRJjTw8o+v9zxhs99zW2uTQnhlvs7eq2sI0w2MHAIBtbeaBf/zAq8fn0JN3j0zy0iTXVNUVGT0F991JvmWw2fmttVfP+tgAAKB3m3GGP621C5M8Ksnlg9VHJDkxhz+B99YkL0/yxM04LgAA6N0sH7x1mNbae6rq9CRPyujG3PsmuUeSmzN6SNfFSV7XWvv4Zh0TAAD0btMCf5K01m5M8trxtJ79nLohB3Rof5NDcwIAQBc25ZIeAABgPgR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjmxr4q+qoqnpaVV1UVVdW1f7xfz9SVS+uqm+eYe07VtVTq+pPq+qyqtpXVTdX1Rer6uKqel5VnTSr+gAAMA+7NqtQVZ2V5E1JJkP1sePpO5M8q6pemOTFrbVbNrD245P81yR3W+Tl48fTQ5P8SlW9PMlzNrI+AADMy6ac4a+qc5K8M4eH/YNJvpDkxsG6XUmen+S1G1j7OUnenMPDfkty1bj+wcH6PUn+3yT/vaqO2KhjAACAeZl54K+qE5Kcn+TI8ap9SZ6W5JjW2klJjk7yyCT/MtjsKVX11A2ofW6SFwxWXZvkF5Ic21r7xnH9Y5I8PsnnBu0eleSl660PAADzthln+F+SQ2fXb0pyTmvtNa21G5KktXZra+1dSR6Q5GOD7V5YVUdNW7SqbpfkFYNVX0ny3a21/9Jau3phZWvthtbaW5J8V5JPDtqfV1WnTVsfAAC2gpkG/qo6PskTBqt+vbX2vsXattauS/KYJAfGq45Nsp6z/GcmGQb2X2itXbJU49baNeNjbeNVu3P4sQMAwLYz6zP8j82hG4NvTfKbyzVurX02ydsGqx63jto/MJjfl9FlRctqrf1jkn8crPreddQHAIC5m3Xgf8Rg/oOttStXsc2Fg/n7V9UpU9a+32D+o621g0u2PNxnB/PHT1kbAAC2hFkPy3n/wfz7V7nNByaWH5Tk81PUfnaS301yYpJr1rDdXQfzN09RFwAAtoyZBf6qOi6HD4X5qVVuellGQ2UuHNvp09Rf5PKcFVXVHZJ892DV55ZqCwAA28EsL+mZfMDWFavZaHzpzVWDVdNe0jONpya502D53ZtYGwAANtwsA/+xE8truazmK4P5xZ6Ou+HGIwr9x8Gq65O8dTNqAwDArMzyGv7JMfSvX8O2w7ZTj8W/WlV1ZJI/zeghXAte0Vr78hT7mvxmY9JxCzMHDx7M/v3711qCbeTAgQOLztMvfb7z6POdR5/vPAcPrnbsl61ploF/z8TyWn4jhm13b8CxLKmqdif544xuDl7wvzJ6YNg0Ll9tw0svvTT79u2bsgzbzSWXLPkYCDqlz3cefb7z6POd4corVzPQ5NY1y0t6amK5LdpqjsZn9v8kyQ8NVn8lyY+01ozQAwDAtjfLM/yT16qs5Uz9sO1NG3Ast1FVd07y9oyeyLtgX5JHttYuW8euT17h9eOSfChJTjvttJx66qnrKMVWd+DAga+f/Tn99NOze/dMv7BiC9DnO48+33n0+c5z9NFHz/sQ1mWWgf+6ieU7LdpqccPr9m/YgGM5TFXdM8lfJPnWweprMwr7H1zPvltry45GVHXoi49du3Zlz57JK5/o1e7du/X3DqPPdx59vvPo851h165ZP7pqtmZ5Sc/VE8trGW1n2PaqJVtNoaoelOSDOTzsX5XkzNba+zayFgAAzNssA//eieUTVrNRVe3K4UN6rmr8/lXu+7FJ/npi/59N8m/HD+oCAICuzCzwj4e0HI69f59VbnqvHH6p0Sc24niq6rwk5ye5/WD1B5I8qLV26UbUAACArWaWZ/iT8c2pYw9c5TaT7T683oOoql9N8socPnLQnyU5a5qx9gEAYLuYdeC/aDD/4Ko6ZhXbnDuY/+RKN8GupKqekeRFE6tfleRHW2tfW8++AQBgq5t14L8gyS3j+T1Jnr5c46q6d5JHD1b9wXqKV9X3ZnRmf+hXWmvntdZuXc++AQBgO5hp4B+fnb9gsOo5VXX2Ym2r6ugkb82hMfivTfL6aWuPv014cw7/f3xea+2l0+4TAAC2m1mf4U+SX0ly/Xh+d5J3VNWzq+ouSVIjD8/oev8zBts9t7U2ObRnxtvsraq2MC1R95eSnDRYvrC19oL1/I8AAMB2M/OnCLTWLquqx2d09n5PkiOTvDTJi6rqS0mOyW0fynV+a+3V09asqqNy28uHHlBVax2N54rW2pnTHgcAAMzbpjw2rLV2YVU9KsnvJjl5vPqIJCdONL01o2vuf3mdJb8vyV0m1h03xX6292PVAADY8TYt0LbW3lNVpyd5UkY35t43yT2S3JzRQ7ouTvK61trHN6Dcasf8BwCArm3qGezW2o1JXjue1rOfU1d4/TeS/MZ6agAAQA8246ZdAABgTgR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRM4AcAgI4J/AAA0DGBHwAAOibwAwBAxwR+AADomMAPAAAdE/gBAKBjAj8AAHRsUwN/VR1VVU+rqouq6sqq2j/+70eq6sVV9c0zrn92Vb2pqi6tqhuqal9VXVJVb6mq768qfwABANCVXZtVqKrOSvKmJCdNvHTsePrOJM+qqhcmeXFr7ZYNrH3Xce1HLfLyt4ynxya5uKqe1Fr7/EbVBgCAedqUM9pVdU6Sd+bwsH8wyReS3DhYtyvJ85O8dgNr3znJxblt2L96PA2dmeTvq+q4jaoPAADzNPPAX1UnJDk/yZHjVfuSPC3JMa21k5IcneSRSf5lsNlTquqpG3QIr0vy7YPlNyc5rbV2j9baPZLcO6Oz/wtOTvLWqqoNqg8AAHOzGWf4X5LkbuP5m5Kc01p7TWvthiRprd3aWntXkgck+dhguxdW1VHrKVxVZyb58cGqV7fWfqK19tmFFa21f22tPSnJ8wbt/m2SH11PbQAA2ApmGvir6vgkTxis+vXW2vsWa9tauy7JY5IcGK86Nsl6z/L/h8H8Z5M8c6mGrbUXJPnrwapfW2dtAACYu1mf4X9sDt0YfGuS31yu8fjM+9sGqx43beGq+oYkjxis+q3W2oGl2o/9xmD+26rqftPWBwCArWDWgX8YuD/YWrtyFdtcOJi/f1WdMmXth+Xw/78Ll2o48N4kXxss/7spawMAwJYw68B//8H8+1e5zQcmlh80Ze0HDOa/2lr79EobtNb2J/noYNX3TFkbAAC2hJkF/vHQlncbrPrUKje9LKMhOxecPuUh3HeK2knymQ2oDQAAW8Isz/BPPmDritVs1Fo7mOSqwappL+kZ1l9V7bEvDPdheE4AALazWQb+YyeWr1nDtl8ZzN9tyVarrz9t7V1J7jxlfQAAmLtdKzeZ2uQY+tevYdth22nH4h9uN23thf1cu9qNq2rym41JJy7MXHHFWr54YDs6ePBgrrxydK/60UcfnV27Zvkrx1agz3cefb7z6POdZyKzHTGv45jWLH9C90wsrzQk5lJtd29A/WlrT1P/8tU2fOhDH7rGXQMAMGf3SPK5eR/EWszykp7Ja9/bDGutVH+zawMA0KfJy9a3vFme4d8/sbyWM+XDtjeto/7t11l7mvonr/D6KUn+YTz/3Tn8JmH6c1ySD43nH5DkS3M8FjaHPt959PnOo893nhNzaIj5S+Z5INOYZeC/bmL5TmvYdnj9/Q3rqL8Q+Ketveb6rbVlL8yfGPTnCyu1Z3ub6O8v6e/+6fOdR5/vPPp855no88mT2lveLC/puXpieS2j7QzbXrVkq9XXn7b211pr+6asDwAAczfLwL93YvmE1WxUVbty+LVR0/7VPKy/qtpjJw7mXW4DAMC2NrPA31r7cg4f//4+q9z0Xjn8UqNPTHkIw6frrrb2ZNtpawMAwJYwyzP8yaEbWpLkgavcZrLdhzeg9nFVtdLNtKmqPUnOWGIfAACw7cw68F80mH9wVR2zim3OHcx/ch03wrw3ya2D5UetYpuzk9xhsHzRUg0BAGA7mHXgvyDJLeP5PUmevlzjqrp3kkcPVv3BtIVba1cn+avBqmdU1UrDcz5zMH9Ja80ZfgAAtrWZBv7x2fkLBqueU1VnL9a2qo5O8tYcGgf/2iSvX+chvGIwf3qS1y7VsKqek+T7Bqtevs7aAAAwd9XabB9CW1XflOSfcmh8+5uT/Mckr22tXVujgU0fluTVSb5lsOnPt9ZevcQ+9ya558Jya23yqb7Dthfm8Mt53pHkl1trnxi/fmqS5yV58qDNh5I8qLV2SwAAYBubeeBPkqo6N6Oz93sGq2/J6Ml0x+S2D8Y6v7X2hGX2tzerD/zfkOSvk3zbxEvXjI9h8vHIX0zyb1tre5faJwAAbBezvoY/SdJaWzjLfvlg9REZjXk/DPu3ZnQpzRM3sPbVGV2q8xcTL909tw37H0ryEGEfAIBebMoZ/q8Xq7pjkidldGPufZPcI6NLfPYmuTjJ61prH1/FfvZmlWf4J7Y7O8njkzwkyfEZfeNwVUZB/y1J/sxlPAAA9GRTAz8AALC5NuWSHgAAYD4EfgAA6JjADwAAHRP4AQCgYwL/KlTVUVX1tKq6qKqurKr94/9+pKpeXFXfPOP6Z1fVm6rq0qq6oar2VdUlVfWWqvr+qtKPG2yefV5Vd6yqp1bVn1bVZeP+vrmqvlhVF1fV86rqpFnV36nm/Xu+xDFVVb2rqtp4evJmH0PP5t3nVXWX8e/6O6rqX6vqxqq6vqo+U1V/WFWPGD+ckg0w58/121XVo6vqzeN/y/dV1U1V9YWqemdVnVdVd55VfQ6pqnuOf9dm+pk67vMfrqoLqupzVfW1qvpqVX2iql5fVd87q9qLaq2ZlpmSnJXR8wPaMtOBJM9NcsQG175rkgtXqN2S/E2SU+b9XvUyzbnPH5/RQ+FW6vObk7xko+vv1Gmefb7CcT1j4hiePO/3qpdp3n2e5AkZPXxypd/1i5OcPO/3a7tPc/5c/5YkH11FX/9/SX503u9Vz1NGz4D6m1l/piY5Jck/rKLPL0hyt035f5/3m7+VpyTnJLlpkQ+EK5LcsEjHvW4Da985yccWqfHl8TS5/vNJjpv3e7bdpzn3+XMW2f+tSa4c1z+wyOsXziKM7KRpnn2+wnHdN8nXJmo/ed7vVw/TvPs8ya8uUuOWcf1rF3ltb5IT5/2+bddpzp/r98soyC/2b/nlSQ4u8trT5/2e9Tol+e1Zf6YmOXmcySb/Lf/SEj8LH01yp5n/v8/7zd+qU5ITcviZ1uuS/N8LnZLR5VCPSPLpiY576gbV/6OJ/f5hknsPXr9XkjdOtPn7jJ+tYNpefZ7k3Il9fjXJeUm+YdDmTkkeN/7Hf9j2P8/7vduu07x/z5c5rj1Z/Izgk+f9nm33ad59nuQpE/u9JMmPJdk9fr2SPDDJeyfavXPe7912nOb8ub57kf3+/sS/5XdO8vM5/A+PW5N897zfu56mjM7s/9asP1PHP0//c6Iv/0uSEwZtvj3JOyaO4/yZvwfz7oStOiV5w6AjvpbkQUu0u3OSfxy0vTLJUeusfebED8Krlmn73Im2Pzbv9267TvPq8/EHxGcG+7smyenLtL97kk8M2u9Pctq837/tOM3z93yF4/rPi/zDJPBv8z7P6Anv1w/2+b4s8XX++HNhMhTcf97v33ab5tzfPz3Rf89fpu13Z3Sp5kLbv573e9fLlOQbM7o0buafqUmePLH/85ZoV0l+d6LtA2f6Psy7I7biNP5QHl4+8YIV2t97HLoW2v/iOusPP+QvzfjMzzLth2eC/mne7992nObZ50m+b+KX/omr2OaMjM4cLGzzvHm/h9ttmvfv+TJ1zszo8o6W216+9+R5v2/beZp3nyd55WBfVye5+wrtT5n4Pf9P834Pt9O0Bfr7zwf7+lxWuPwyyX8btL8lm3Rtd89Tkkcl+eLE5+gsA//HB/v+uxXa7s7oG76F9n8+y/fC6C6Le2ySXeP5W5P85nKNW2ufTfK2warHTVu4qr4ho68XF/xWa+3ACpv9xmD+26rqftPW38Hm1udJfmAwvy/J+Stt0Fr7x4zORi3Y3Lv9+zDPPl9UVR2T0aV6C5/NT93oGjvcPD/b92R09m/Bs1tr16xQ//PjY/yTJK9J8s/T1t+h5v07ftpg/n2ttVtWaP/ewfztknzTOuvvWFX1HVX1Vxnd53b84KXfmWHNf5PRPRsLXr1c+3G2e9Vg1SPG/wbMhMC/uGHg/mBr7cpVbHPhYP7+VXXKlLUflsP75cKlGg68N6OvKhf8uylr72Tz7PPhB8RHW2sHV7ndZwfzxy/ZiqXMs8+X8pqMzuomyRtaa29brjFrNs8+PzPJMeP5azL6w25FrbWnt9Z+vLX2tNbam6esvVPN+3f8yMH8HabYXkab3huTnD1YviHJz7XWZnkSZfjzdjDJO1exzfDnbXdG9/PNhB+mxd1/MP/+VW7zgYnlB01Z+wGD+a+21j690gattf0Z3eC34HumrL2TzbPPn53RTXu/kOT1a9juroP5m6esvZPNs89vo6oen0NnFD+X0Y18bKx59vmZg/n3ruKbW9Zv3r/jlw7mH1JVd1qh/cMH8wczuuGX9XtHkm9trf3WjOsM89vHW2s3rLRBa+2KjC45WjCz/LZr5SY7S1Udl+Rug1WfWuWml2X0C7rwnp4+5SHcd4rayeimz4UflGlr70jz7vNFLs9ZUVXdIaObvBZ8bpraO9W8+3yR4zk5hy43uDWj+ziu24h9M7IF+vy7BvNfP0FTVfdI8uMZnY08OaPRRL6Q5K+SvGWVZ6WZsAX6O0n+NIdC/F2TvLKqfraNL+AeqqqzMrrJd8FbfQasS8voRt3nt9Yu3qSa68lvJ4znZ5bfnOG/rcknmF6xmo3Gl2FcNVg17deAw/qrqj32heE+PJ1xTebd59N4akbDdC549ybW7sGW6fPxk7LflEOXe7yitfZ3690vtzHvPr/PYP7zVXVEVf1yRmOx/9ckP5zRHwVnZHRfzyuTfKaqfmHKejvdvPs7GQ3B+eHB8lOSvKeqHlZVd66q3VV1elW9KMm7MrqkIxn9TPzSOuqS/GBr7axNDPvJxuS3meUIgf+2jp1YXvamqglfGczfbclWq68/be1dGQ0xxurMu8/XpKqOT/IfB6uuT/LWzajdka3U58/Mocs9/jnJr23APrmteff5iYP56zI6+/ufcvh13pOOTvKKqvodJ3HWbN79vXBT5vdndKZ5wdlJLsroAWv7MzoT/KsZPXsjGV33/aDW2vAyD9aotXb5Ztarqjvm8JNwWy5HCPy3ddTE8vVr2HbYdnI/09SftvZ66u9E8+7zVauqIzMKCscMVr+itfblWdfuzJbo86o6I8kLx4v7k/xka839GLMxtz6vqrvk0NnbZPT8lB8az380o3t4vjHJ7TP6JuBXM/qjYMHPJHnWWuvucFvid7y19uXW2llJfi6jp/0u58+T/Exr7QsrtGPr2RI/b8sR+G9rz8TyWm6sGrbdvWSr1deftvZ66u9E8+7zVamq3Un+OIffRPa/krxklnU7Nfc+r6rbZ/QE7YVjeW5r7WPT7o8VzbPP7zixvHBz3xsyetjOBa21q1prN7fWPtNae0lGT9sdnuV9UVWdOkXtnWruv+NJUlXHVtWbMhpj//bj1QvP2PhiRvcLLPjBjC7j+sX11GQuNurnbWb31gr8tzX5teltbq7ZxPqbXXunmnefr2h8Zv9PcuisYDL6GvBHnBGeylbo85fl0JCsf5/k1+dwDDvJPPt8sdD4/iRPWWoY3vEIbT8xWLUro8u/WJ25/45X1WkZfYPzkxndjL0vo1HZjm+tHdtaOzGjb2sfl9GNm8noj8OXV5XPg+1lw37eZnX5nsB/W/snltfy1/2w7Upf3a2m/rS111N/J5p3ny+rqu6c0Q1dPzxYvS/JI1trl82i5g4w1z6vqocnefp4cV9Go/LcOs2+WLV59vli2zx3pWdutNb+JsnwBu6ZjdHdoXn/ju9J8vYcGn3lmiTf01p72XDkpdbaDa21P8rohu2LB7t4ZlUNH8rI1rZRP283LzaK00YQ+G9rchislcbNHRpee7Xi+KurqD9t7fXU34nm3edLqqp7JvmHHD6G97VJzmmtfXCj6+0gc+vzqrp7RqN3LJzFOc8fbptinr/n+xZZ/ptVbvuXg/lTqurEJVsyNO/P9Sfm8GEaf7q19vGlGrfW9iV5TJKrB6ufP2VtNt+8f95WJPDf1tUTy2u5Y3rY9qolW62+/rS1vzb+8GB15t3ni6qqByX5YJJvnahxZmvtfRtZaweaZ5//dg6d9Xt7a+33ptgHaze3Pm+tfS3JLYNVl63hidqTD186bq31d6h5f67/+GD+0621t6+0QWvtKzn0PI4k+a4ZPM2bGRj/jt84WLUlcsSQwH9beyeWT1is0aSq2pXDhwFbyxisS9VfVe2x4Vkfd/ivzd6J5c3u88X2/dgkfz2x/88m+bfjB3WxPnsnljelz8f/eD9msOqBVXXpctPELl428bqzvau3d2J5s3/PPz+YX8sNfUZgm87eieXN7u8zBvNrea7G5Dc/3z5lfTbf3sH8tPltw3LEJE/andBa+3JVXZPk7uNV91mu/cC9cvj7+YkpD+FTGY3bu5bak22nrb0jbYE+P0xVnZfkFTn8JqAPJDnX8JsbY459PnmS5fg1bn9sDg8jRuNapS3we/5PSb5pPL+Wfp98psq1U9bfUbZAfx8zmP/qGrab/Iz3TJ3t41M5dBnXlstvzvAv7kOD+QeucpvJdh9etNXaah9XVSevtMH45qAzltgHqzPPPv+6qvrVjJ6wOQz7f5bkLGF/w22JPmdTzbPPh0OunrCGSzWG14G3JJPf+rC0efb3Vwfzaznbe9eJ5a8s2oqtaPjz9h3jbLasqjoph58AmFl+E/gXd9Fg/sFVdcwqthmOnvDJ1tq0X8u8N8lwtI5HrWKbs5PcYbB80VINWdI8+zxJUlXPSPKiidWvSvKj4+sD2Vib3uettb2ttVrLNLGLn5p4fe9a6jPX3/M/n1j+sVVu9/2D+Y+21tbyQJ+dbp79PfzD7Puq6ohVbveQieV/mbI+m2/483aHjLLZSoY/by3JX23oEQ0I/Iu7IIdusNqTQ8PnLaqq7p3k0YNVfzBt4dba1Tm8w58xfuDScoZjM1/SWnOGf+3m1ufj/X1vRmf2h36ltXae4RpnZq59zlzM87P9I0k+OVj1rKpa9sa+qjorhz9o7y3T1t+h5vk7/o7B/PE5/JkKS9W/U5KnDVZ9srX2r+s4BjZRa+2jOfwm+19arv042z1jsOqi4ZCtG661ZlpkyuiDtY2n/UnOXqLd0Rk9WGOh7VeTfMM6a58z2F9L8vpl2j5nou3PzPu9267TvPo8o2s9L5/ox+fO+/3YCdM8f8/XcIzDn4snz/s92+7TnD/bHz/Rn+9Kcqcl2t4zoxt9F9pek+SYeb9/222a4+f6sRkN1biwv2szeqryUu2PTPLfJ34+njjv96/HaZafqUl+dmL/z1mm7esn2j5spv/f837jt+qU0c1V+wYdcVNGT8i7y/j1SvLwJJdMdNgzltnn3mHbFepfOLHfv0hyv8Hrp2Y0lvewzQeTHDHv9267TvPq8yQvnNjfn8/7vdgp07x/z1d5jAJ/R32+yGf7xzJ6qN7u8etHZnQ2+H9PtPv3837vtuM0z/7ObcPfzUlemuRegzZHZnTp7kcn2l6U5Hbzfv96nKb5TJ3YZu8y7XYt0pe/n+TUQZv7ZZTphm3+bNb/3zUuziKq6twkb83oq8AFtyT5UkZnZScfrHB+a+0Jy+xvb0ZnbZIk7bbX5w7bfkNGwzJ+28RL14yP4diJ9V/MaMjGvUvtk5Vtdp9X1VEZDcN1l8HqL2XtD9+4orV25hq3IfP9PV/l8Q0/pH+qtfaG9eyPuX+2H51R6H/oxEsHx/XvkVEIHHpVa+28pfbJ8ubc3y9K8quLvLQvo28RvnHiuJLRjcL/Z2vNiEwzMM1n6sQ2n2utnbpM23tn9NTkkyZeuirJETk0ctSCT2aU37660nGsy7z/0trqU5KH5fCvVRebbknyG1nh7HrWeOYvyTfktmeDFps+mMEZA9P26fMkP7iK/l3NtHfe79t2nub5e76KY1vz2SjT1u7zJLcf73f/CvVvyDJnmk3bpr9/PLf9xmax6UCS/5bk9vN+v3qepvlMXeu/txkN7/oPq+jzd2WTLg81Dv8KWmvvqarTkzwpo5t57pvRGZibM/qlvzjJ69oyj8xeR+2rk5xbVWdndO3nQzK6+WdPRn8pfiij6xP/rLV2y5I7Yk02uc/XMlYvMzLP33PmY86f7TcleWZVvTbJE5I8MqPLNO+e0dM6P5XkL5P8TmvtSxtdfyeac3//cVW9PaN/xx+Z5LvGtfdk9K39ZzL6Rv+Nzbf0XWit/WtVPTijy/V+LKPhXo/LaLCc/53kfUne1Fp792Ydk0t6AACgY4blBACAjgn8AADQMYEfAAA6JvADAEDHBH4AAOiYwA8AAB0T+AEAoGMCPwAAdEzgBwCAjgn8AADQMYEfAAA6JvADAEDHBH4AAOiYwA8AAB0T+AEAoGMCPwAAdEzgBwCAjgn8AADQMYEfAAA6JvADAEDHBH4AAOiYwA8AAB0T+AEAoGMCPwAAdEzgBwCAjgn8AADQMYEfAAA69v8DeMjeRZO/U44AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 400x400 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 358,
       "width": 382
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sc.set_figure_params(dpi=100)\n",
    "for e in adata.obsm:\n",
    "    sc.pl.embedding(adata, basis=e, color='Celltypes_colors', legend_loc='on data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Check that the default_embedding value, if defined, is in obsm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'default_embedding' in adata.uns:\n",
    "    de = adata.uns['default_embedding']\n",
    "    if de not in adata.obsm_keys():\n",
    "        print('ERROR:' + de + ' not in ' + ','.join(adata.obsm_keys()))\n",
    "    else:\n",
    "        print(de + ' is in ' + ','.join(adata.obsm_keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** update uns.default_embedding\\\n",
    "Ideally, the default_embedding matches the figures in the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['default_embedding'] = 'X_umap'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OverloadedDict, wrapping:\n",
       "\t{'Celltypes_colors': array(['#43cd80', '#54ff9f', '#cd6600', '#ff8c00', '#ff83fa', '#9400d3',\n",
       "       '#ff34b3', '#6e8b3d', '#4169e1', '#add8e6', '#0000ee', '#009acd',\n",
       "       '#2f4f4f', '#00868b', '#00cdcd', '#8b3626', '#ffbbff', '#ffaeb9',\n",
       "       '#c71585', '#ee7621', '#ff6347', '#d02090', '#7cfc00', '#c0ff3e',\n",
       "       '#ffd39b', '#ff7256', '#cd3333', '#cd661d', '#8b4513', '#db7093',\n",
       "       '#cd919e', '#ff0000', '#388e8e', '#7d9ec0', '#bc8f8f', '#ffdead',\n",
       "       '#e3a869', '#cd69c9', '#bcee68', '#473c8b', '#00bfff', '#00cd66',\n",
       "       '#668b8b', '#ee00ee', '#eeaeee', '#68228b', '#87cefa', '#53868b',\n",
       "       '#ff69b4', '#9a32cd', '#ffd700', '#ff00ff', '#ffa54f', '#ff9912',\n",
       "       '#ffd39b', '#cdcd00', '#ffa500', '#ffec8b', '#00c957', '#daa520',\n",
       "       '#8b6914', '#006400', '#8b4c39', '#fa8072', '#97ffff', '#7fffd4',\n",
       "       '#79cdcd', '#00c78c', '#7fff00', '#173e17', '#8fbc8f', '#c1ffc1',\n",
       "       '#228b22', '#8b8b00', '#cdcd00', '#ff7256', '#ffa07a', '#a2cd5a',\n",
       "       '#6495ed', '#00f5ff'], dtype=object), 'Celltypes_int_F_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF', '#3B9700', '#04F757', '#C8A1A1',\n",
       "       '#1E6E00'], dtype=object), 'Celltypes_int_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF', '#3B9700', '#04F757', '#C8A1A1'],\n",
       "      dtype=object), 'Celltypes_int_updB_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF', '#3B9700', '#04F757', '#C8A1A1',\n",
       "       '#1E6E00'], dtype=object), 'Celltypes_master_high_colors': array(['#43cd80', '#54ff9f', '#ff8c00', '#cd00cd', '#8b3626', '#ffbbff',\n",
       "       '#ffd39b', '#ff7256', '#6e8b3d', '#ffa07a', '#ff0000', '#388e8e',\n",
       "       '#ffd700', '#daa520', '#9a32cd', '#228b22', '#4169e1'],\n",
       "      dtype=object), 'ID_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF'], dtype=object), 'Loc_true_colors': array(['#8b5a00', '#cd8500', '#ffa500', '#eecfa1', '#ffefd5'],\n",
       "      dtype=object), 'hvg': {'flavor': 'seurat'}, 'leiden': {'params': {'n_iterations': -1, 'random_state': 0, 'resolution': 2.8}}, 'leiden_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648'], dtype=object), 'neighbors': {'connectivities_key': 'connectivities', 'distances_key': 'distances', 'params': {'method': 'umap', 'metric': 'euclidean', 'n_neighbors': 10, 'n_pcs': 40, 'random_state': 0, 'use_rep': 'X_pca_hm'}}, 'pca': {'params': {'use_highly_variable': True, 'zero_center': True}, 'variance': array([48.53690708, 38.66180274, 32.97178512, 29.18578745, 21.04634253,\n",
       "       12.97956791, 12.64911223, 11.7482765 , 10.26793121,  9.51281739,\n",
       "        8.99274211,  8.69294486,  8.14785825,  7.31586907,  7.19563638,\n",
       "        6.44186607,  6.15111847,  5.37659114,  5.13961721,  4.75653332,\n",
       "        4.64352793,  4.4548419 ,  4.1937319 ,  3.99181081,  3.76066431,\n",
       "        3.66332497,  3.50102824,  3.43801557,  3.06485983,  2.94459591,\n",
       "        2.81317401,  2.73277275,  2.68476646,  2.54325974,  2.49991643,\n",
       "        2.38061747,  2.29362476,  2.25160515,  2.16495846,  2.0609124 ,\n",
       "        2.00574757,  1.96144574,  1.88992877,  1.88543583,  1.80772319,\n",
       "        1.78486397,  1.74182449,  1.71894434,  1.7126994 ,  1.66858902]), 'variance_ratio': array([0.04702296, 0.03745588, 0.03194334, 0.02827543, 0.02038987,\n",
       "       0.01257471, 0.01225457, 0.01138183, 0.00994766, 0.0092161 ,\n",
       "       0.00871224, 0.0084218 , 0.00789371, 0.00708767, 0.00697119,\n",
       "       0.00624093, 0.00595925, 0.00520889, 0.0049793 , 0.00460817,\n",
       "       0.00449869, 0.00431589, 0.00406292, 0.0038673 , 0.00364336,\n",
       "       0.00354906, 0.00339183, 0.00333078, 0.00296926, 0.00285275,\n",
       "       0.00272543, 0.00264753, 0.00260102, 0.00246393, 0.00242194,\n",
       "       0.00230636, 0.00222208, 0.00218137, 0.00209743, 0.00199663,\n",
       "       0.00194318, 0.00190026, 0.00183098, 0.00182663, 0.00175134,\n",
       "       0.00172919, 0.00168749, 0.00166533, 0.00165928, 0.00161654])}, 'scDonor_snBatch_colors': array(['#1f77b4', '#ff7f0e', '#279e68', '#d62728', '#aa40fc', '#8c564b',\n",
       "       '#e377c2', '#b5bd61', '#17becf', '#aec7e8', '#ffbb78', '#98df8a',\n",
       "       '#ff9896', '#c5b0d5', '#c49c94'], dtype=object), 'scsn_colors': array(['#1f77b4', '#ff7f0e'], dtype=object), 'umap': {'params': {'a': 0.5830300205483709, 'b': 1.334166992455648}}}\n",
       "With overloaded keys:\n",
       "\t['neighbors']."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# uns\n",
    "**Check for uns schema fields**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "schema_version: MISSING\n",
      "title: MISSING\n"
     ]
    }
   ],
   "source": [
    "uns_schema =['schema_version','title']\n",
    "for p in uns_schema:\n",
    "    print(p + ': ' + adata.uns.get(p,'MISSING'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** define uns.schema_version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['schema_version'] = '3.0.0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['title'] = 'hello'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Browse all of uns**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OverloadedDict, wrapping:\n",
       "\t{'Celltypes_colors': array(['#43cd80', '#54ff9f', '#cd6600', '#ff8c00', '#ff83fa', '#9400d3',\n",
       "       '#ff34b3', '#6e8b3d', '#4169e1', '#add8e6', '#0000ee', '#009acd',\n",
       "       '#2f4f4f', '#00868b', '#00cdcd', '#8b3626', '#ffbbff', '#ffaeb9',\n",
       "       '#c71585', '#ee7621', '#ff6347', '#d02090', '#7cfc00', '#c0ff3e',\n",
       "       '#ffd39b', '#ff7256', '#cd3333', '#cd661d', '#8b4513', '#db7093',\n",
       "       '#cd919e', '#ff0000', '#388e8e', '#7d9ec0', '#bc8f8f', '#ffdead',\n",
       "       '#e3a869', '#cd69c9', '#bcee68', '#473c8b', '#00bfff', '#00cd66',\n",
       "       '#668b8b', '#ee00ee', '#eeaeee', '#68228b', '#87cefa', '#53868b',\n",
       "       '#ff69b4', '#9a32cd', '#ffd700', '#ff00ff', '#ffa54f', '#ff9912',\n",
       "       '#ffd39b', '#cdcd00', '#ffa500', '#ffec8b', '#00c957', '#daa520',\n",
       "       '#8b6914', '#006400', '#8b4c39', '#fa8072', '#97ffff', '#7fffd4',\n",
       "       '#79cdcd', '#00c78c', '#7fff00', '#173e17', '#8fbc8f', '#c1ffc1',\n",
       "       '#228b22', '#8b8b00', '#cdcd00', '#ff7256', '#ffa07a', '#a2cd5a',\n",
       "       '#6495ed', '#00f5ff'], dtype=object), 'Celltypes_int_F_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF', '#3B9700', '#04F757', '#C8A1A1',\n",
       "       '#1E6E00'], dtype=object), 'Celltypes_int_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF', '#3B9700', '#04F757', '#C8A1A1'],\n",
       "      dtype=object), 'Celltypes_int_updB_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF', '#3B9700', '#04F757', '#C8A1A1',\n",
       "       '#1E6E00'], dtype=object), 'Celltypes_master_high_colors': array(['#43cd80', '#54ff9f', '#ff8c00', '#cd00cd', '#8b3626', '#ffbbff',\n",
       "       '#ffd39b', '#ff7256', '#6e8b3d', '#ffa07a', '#ff0000', '#388e8e',\n",
       "       '#ffd700', '#daa520', '#9a32cd', '#228b22', '#4169e1'],\n",
       "      dtype=object), 'ID_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648', '#0086ED', '#886F4C', '#34362D', '#B4A8BD', '#00A6AA',\n",
       "       '#452C2C', '#636375', '#A3C8C9', '#FF913F', '#938A81', '#575329',\n",
       "       '#00FECF', '#B05B6F', '#8CD0FF'], dtype=object), 'Loc_true_colors': array(['#8b5a00', '#cd8500', '#ffa500', '#eecfa1', '#ffefd5'],\n",
       "      dtype=object), 'hvg': {'flavor': 'seurat'}, 'leiden': {'params': {'n_iterations': -1, 'random_state': 0, 'resolution': 2.8}}, 'leiden_colors': array(['#FFFF00', '#1CE6FF', '#FF34FF', '#FF4A46', '#008941', '#006FA6',\n",
       "       '#A30059', '#FFDBE5', '#7A4900', '#0000A6', '#63FFAC', '#B79762',\n",
       "       '#004D43', '#8FB0FF', '#997D87', '#5A0007', '#809693', '#6A3A4C',\n",
       "       '#1B4400', '#4FC601', '#3B5DFF', '#4A3B53', '#FF2F80', '#61615A',\n",
       "       '#BA0900', '#6B7900', '#00C2A0', '#FFAA92', '#FF90C9', '#B903AA',\n",
       "       '#D16100', '#DDEFFF', '#000035', '#7B4F4B', '#A1C299', '#300018',\n",
       "       '#0AA6D8', '#013349', '#00846F', '#372101', '#FFB500', '#C2FFED',\n",
       "       '#A079BF', '#CC0744', '#C0B9B2', '#C2FF99', '#001E09', '#00489C',\n",
       "       '#6F0062', '#0CBD66', '#EEC3FF', '#456D75', '#B77B68', '#7A87A1',\n",
       "       '#788D66', '#885578', '#FAD09F', '#FF8A9A', '#D157A0', '#BEC459',\n",
       "       '#456648'], dtype=object), 'neighbors': {'connectivities_key': 'connectivities', 'distances_key': 'distances', 'params': {'method': 'umap', 'metric': 'euclidean', 'n_neighbors': 10, 'n_pcs': 40, 'random_state': 0, 'use_rep': 'X_pca_hm'}}, 'pca': {'params': {'use_highly_variable': True, 'zero_center': True}, 'variance': array([48.53690708, 38.66180274, 32.97178512, 29.18578745, 21.04634253,\n",
       "       12.97956791, 12.64911223, 11.7482765 , 10.26793121,  9.51281739,\n",
       "        8.99274211,  8.69294486,  8.14785825,  7.31586907,  7.19563638,\n",
       "        6.44186607,  6.15111847,  5.37659114,  5.13961721,  4.75653332,\n",
       "        4.64352793,  4.4548419 ,  4.1937319 ,  3.99181081,  3.76066431,\n",
       "        3.66332497,  3.50102824,  3.43801557,  3.06485983,  2.94459591,\n",
       "        2.81317401,  2.73277275,  2.68476646,  2.54325974,  2.49991643,\n",
       "        2.38061747,  2.29362476,  2.25160515,  2.16495846,  2.0609124 ,\n",
       "        2.00574757,  1.96144574,  1.88992877,  1.88543583,  1.80772319,\n",
       "        1.78486397,  1.74182449,  1.71894434,  1.7126994 ,  1.66858902]), 'variance_ratio': array([0.04702296, 0.03745588, 0.03194334, 0.02827543, 0.02038987,\n",
       "       0.01257471, 0.01225457, 0.01138183, 0.00994766, 0.0092161 ,\n",
       "       0.00871224, 0.0084218 , 0.00789371, 0.00708767, 0.00697119,\n",
       "       0.00624093, 0.00595925, 0.00520889, 0.0049793 , 0.00460817,\n",
       "       0.00449869, 0.00431589, 0.00406292, 0.0038673 , 0.00364336,\n",
       "       0.00354906, 0.00339183, 0.00333078, 0.00296926, 0.00285275,\n",
       "       0.00272543, 0.00264753, 0.00260102, 0.00246393, 0.00242194,\n",
       "       0.00230636, 0.00222208, 0.00218137, 0.00209743, 0.00199663,\n",
       "       0.00194318, 0.00190026, 0.00183098, 0.00182663, 0.00175134,\n",
       "       0.00172919, 0.00168749, 0.00166533, 0.00165928, 0.00161654])}, 'scDonor_snBatch_colors': array(['#1f77b4', '#ff7f0e', '#279e68', '#d62728', '#aa40fc', '#8c564b',\n",
       "       '#e377c2', '#b5bd61', '#17becf', '#aec7e8', '#ffbb78', '#98df8a',\n",
       "       '#ff9896', '#c5b0d5', '#c49c94'], dtype=object), 'scsn_colors': array(['#1f77b4', '#ff7f0e'], dtype=object), 'umap': {'params': {'a': 0.5830300205483709, 'b': 1.334166992455648}}, 'schema_version': '3.0.0', 'title': 'hello'}\n",
       "With overloaded keys:\n",
       "\t['neighbors']."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** remove deprecated schema field"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata.uns['X_normalization']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# *_colors\n",
    "**scanpy & cellxgene allow for specification of cluster colors when coloring by specific obs fields**<br>\n",
    "**A list of color codes is specified in `uns.PROP_colors` where `PROP` is an obs field**<br>\n",
    "**The number of color codes in `uns.PROP_colors` must be at least as long as the number of unique values in `obs.PROP`**<br>\n",
    "<br>\n",
    "**Check for _colors fields & ensure each matches a categorical obs field**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Celltypes_int_F not found in obs, consider DELETING or RENAMING uns.Celltypes_int_F_colors\n",
      "WARNING: Celltypes_int not found in obs, consider DELETING or RENAMING uns.Celltypes_int_colors\n",
      "WARNING: Celltypes_int_updB not found in obs, consider DELETING or RENAMING uns.Celltypes_int_updB_colors\n",
      "WARNING: leiden not found in obs, consider DELETING or RENAMING uns.leiden_colors\n",
      "WARNING: scsn not found in obs, consider DELETING or RENAMING uns.scsn_colors\n"
     ]
    }
   ],
   "source": [
    "numb_types = ['int_', 'int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64','float_', 'float16', 'float32', 'float64']\n",
    "\n",
    "for k in adata.uns.keys():\n",
    "    if k.endswith('_colors'):\n",
    "        colors = len(adata.uns[k])\n",
    "        obs_field = k[:-(len('_colors'))]\n",
    "\n",
    "        if obs_field.endswith('_ontology_term_id'):\n",
    "            label_field = obs_field[:-17]\n",
    "            print(f'WARNING: consider copying uns.{k} to uns.{label_field}_colors so palette transfers to CxG viz')\n",
    "\n",
    "        if obs_field in portal_obs_fields:\n",
    "            obs_field += '_ontology_term_id'\n",
    "        if obs_field not in adata.obs.keys():\n",
    "            print(f'WARNING: {obs_field} not found in obs, consider DELETING or RENAMING uns.{k}')\n",
    "        else:\n",
    "            values = len(adata.obs[obs_field].unique())\n",
    "            if colors < values:\n",
    "                print(f'ERROR: uns.{k} has only {str(colors)} colors but obs.{obs_field} has {str(values)} values')\n",
    "            if adata.obs.dtypes[obs_field].name in numb_types:\n",
    "                print(f'ERROR: uns.{k} is associated with non-categorical {obs_field}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Celltypes',\n",
       " 'Celltypes_master_high',\n",
       " 'Celltypes_master_higher',\n",
       " 'Celltypes_master_higher_immune',\n",
       " 'Loc_true',\n",
       " 'Material',\n",
       " 'Donor',\n",
       " 'Donor_ID_2',\n",
       " 'Gender',\n",
       " 'Sample',\n",
       " 'ID',\n",
       " 'Protocol_plot',\n",
       " 'Version',\n",
       " 'Study',\n",
       " 'PoolDon',\n",
       " 'DonorPool',\n",
       " 'scDonor_snBatch',\n",
       " 'n_genes',\n",
       " 'n_genes_by_counts',\n",
       " 'total_counts',\n",
       " 'total_counts_mt',\n",
       " 'pct_counts_mt']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs_keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Celltypes</th>\n",
       "      <th>Celltypes_master_high</th>\n",
       "      <th>Celltypes_master_higher</th>\n",
       "      <th>Celltypes_master_higher_immune</th>\n",
       "      <th>Loc_true</th>\n",
       "      <th>Material</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Donor_ID_2</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Sample</th>\n",
       "      <th>...</th>\n",
       "      <th>Version</th>\n",
       "      <th>Study</th>\n",
       "      <th>PoolDon</th>\n",
       "      <th>DonorPool</th>\n",
       "      <th>scDonor_snBatch</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGTGAATA-1-WTDAtest7887999-0</th>\n",
       "      <td>AT1</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>AT1</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cells</td>\n",
       "      <td>A32</td>\n",
       "      <td>411C</td>\n",
       "      <td>M</td>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>...</td>\n",
       "      <td>10x3prime_v2</td>\n",
       "      <td>dissociation</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>1943</td>\n",
       "      <td>1943</td>\n",
       "      <td>5356.075684</td>\n",
       "      <td>186.790329</td>\n",
       "      <td>3.487448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGCAACCGCCA-1-WTDAtest7887999-0</th>\n",
       "      <td>B_plasma_IgA</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>B_cell</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cells</td>\n",
       "      <td>A32</td>\n",
       "      <td>411C</td>\n",
       "      <td>M</td>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>...</td>\n",
       "      <td>10x3prime_v2</td>\n",
       "      <td>dissociation</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>226</td>\n",
       "      <td>226</td>\n",
       "      <td>768.994751</td>\n",
       "      <td>5.476773</td>\n",
       "      <td>0.712199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCATGGTACAGCAG-1-WTDAtest7887999-0</th>\n",
       "      <td>Monocyte_CD16</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cells</td>\n",
       "      <td>A32</td>\n",
       "      <td>411C</td>\n",
       "      <td>M</td>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>...</td>\n",
       "      <td>10x3prime_v2</td>\n",
       "      <td>dissociation</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>847</td>\n",
       "      <td>847</td>\n",
       "      <td>1901.789185</td>\n",
       "      <td>108.158295</td>\n",
       "      <td>5.687187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACGTTGGTGTCCTCT-1-WTDAtest7887999-0</th>\n",
       "      <td>CD8_EM/EMRA</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>CD8</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cells</td>\n",
       "      <td>A32</td>\n",
       "      <td>411C</td>\n",
       "      <td>M</td>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>...</td>\n",
       "      <td>10x3prime_v2</td>\n",
       "      <td>dissociation</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>767</td>\n",
       "      <td>767</td>\n",
       "      <td>1783.324097</td>\n",
       "      <td>85.630554</td>\n",
       "      <td>4.801738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACTCCCTCCTAGTGA-1-WTDAtest7887999-0</th>\n",
       "      <td>Macro_intravascular</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Macrophage_other</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cells</td>\n",
       "      <td>A32</td>\n",
       "      <td>411C</td>\n",
       "      <td>M</td>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>...</td>\n",
       "      <td>10x3prime_v2</td>\n",
       "      <td>dissociation</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>799</td>\n",
       "      <td>799</td>\n",
       "      <td>1314.895386</td>\n",
       "      <td>97.600655</td>\n",
       "      <td>7.422694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1</th>\n",
       "      <td>T_reg</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>CD4</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>A42</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A42</td>\n",
       "      <td>A42none</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>1556</td>\n",
       "      <td>1556</td>\n",
       "      <td>2975.571777</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1</th>\n",
       "      <td>SMG_Mucous</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Submucosal_Glands</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>A42</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A42</td>\n",
       "      <td>A42none</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>2723</td>\n",
       "      <td>2723</td>\n",
       "      <td>7334.214355</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1</th>\n",
       "      <td>Basal</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epi_Basal</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>A42</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A42</td>\n",
       "      <td>A42none</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>500</td>\n",
       "      <td>500</td>\n",
       "      <td>626.389465</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1</th>\n",
       "      <td>SMG_Mucous</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Submucosal_Glands</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>A42</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A42</td>\n",
       "      <td>A42none</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>2576</td>\n",
       "      <td>2576</td>\n",
       "      <td>6651.600586</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1</th>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>A42</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>A42</td>\n",
       "      <td>A42none</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>945</td>\n",
       "      <td>945</td>\n",
       "      <td>1351.483398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193108 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Celltypes  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0                   AT1    \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0           B_plasma_IgA   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0          Monocyte_CD16   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0            CD8_EM/EMRA   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0    Macro_intravascular   \n",
       "...                                                     ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1                T_reg   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1           SMG_Mucous   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1                Basal   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1           SMG_Mucous   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1          Chondrocyte   \n",
       "\n",
       "                                       Celltypes_master_high  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0               Epithelia   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0                  Immune   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0                  Immune   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0                  Immune   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0                  Immune   \n",
       "...                                                      ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1                Immune   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1             Epithelia   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1             Epithelia   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1             Epithelia   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1                Stroma   \n",
       "\n",
       "                                       Celltypes_master_higher  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0                 Epithelia   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0                    Immune   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0                    Immune   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0                    Immune   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0                    Immune   \n",
       "...                                                        ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1                  Immune   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1               Epithelia   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1               Epithelia   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1               Epithelia   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1                  Stroma   \n",
       "\n",
       "                                       Celltypes_master_higher_immune  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0                              AT1   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0                           B_cell   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0                         Monocyte   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0                              CD8   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0                 Macrophage_other   \n",
       "...                                                               ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1                            CD4   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1              Submucosal_Glands   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1                      Epi_Basal   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1              Submucosal_Glands   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1                    Chondrocyte   \n",
       "\n",
       "                                             Loc_true Material Donor  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0     d_LowLeftPar    cells   A32   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0     d_LowLeftPar    cells   A32   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0     d_LowLeftPar    cells   A32   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0     d_LowLeftPar    cells   A32   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0     d_LowLeftPar    cells   A32   \n",
       "...                                               ...      ...   ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1  b_Bronchi.2.3   nuclei   A42   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1  b_Bronchi.2.3   nuclei   A42   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1  b_Bronchi.2.3   nuclei   A42   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1  b_Bronchi.2.3   nuclei   A42   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1  b_Bronchi.2.3   nuclei   A42   \n",
       "\n",
       "                                       Donor_ID_2 Gender             Sample  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0         411C      M    WTDAtest7887999   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0         411C      M    WTDAtest7887999   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0         411C      M    WTDAtest7887999   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0         411C      M    WTDAtest7887999   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0         411C      M    WTDAtest7887999   \n",
       "...                                           ...    ...                ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1        nan    nan  WSSS_A_LNG8757929   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1        nan    nan  WSSS_A_LNG8757929   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1        nan    nan  WSSS_A_LNG8757929   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1        nan    nan  WSSS_A_LNG8757929   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1        nan    nan  WSSS_A_LNG8757929   \n",
       "\n",
       "                                        ...       Version         Study  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0    ...  10x3prime_v2  dissociation   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0    ...  10x3prime_v2  dissociation   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0    ...  10x3prime_v2  dissociation   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0    ...  10x3prime_v2  dissociation   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0    ...  10x3prime_v2  dissociation   \n",
       "...                                     ...           ...           ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1  ...           nan           nan   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1  ...           nan           nan   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1  ...           nan           nan   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1  ...           nan           nan   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1  ...           nan           nan   \n",
       "\n",
       "                                       PoolDon DonorPool scDonor_snBatch  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0       nan       nan       A32_cells   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0       nan       nan       A32_cells   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0       nan       nan       A32_cells   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0       nan       nan       A32_cells   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0       nan       nan       A32_cells   \n",
       "...                                        ...       ...             ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1     A42   A42none      A42_nuclei   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1     A42   A42none      A42_nuclei   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1     A42   A42none      A42_nuclei   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1     A42   A42none      A42_nuclei   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1     A42   A42none      A42_nuclei   \n",
       "\n",
       "                                       n_genes n_genes_by_counts  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0      1943              1943   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0       226               226   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0       847               847   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0       767               767   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0       799               799   \n",
       "...                                        ...               ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1    1556              1556   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1    2723              2723   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1     500               500   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1    2576              2576   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1     945               945   \n",
       "\n",
       "                                        total_counts  total_counts_mt  \\\n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0     5356.075684       186.790329   \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0      768.994751         5.476773   \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0     1901.789185       108.158295   \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0     1783.324097        85.630554   \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0     1314.895386        97.600655   \n",
       "...                                              ...              ...   \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1   2975.571777         0.000000   \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1   7334.214355         0.000000   \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1    626.389465         0.000000   \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1   6651.600586         0.000000   \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1   1351.483398         0.000000   \n",
       "\n",
       "                                        pct_counts_mt  \n",
       "AAACCTGGTGTGAATA-1-WTDAtest7887999-0         3.487448  \n",
       "AAACGGGCAACCGCCA-1-WTDAtest7887999-0         0.712199  \n",
       "AACCATGGTACAGCAG-1-WTDAtest7887999-0         5.687187  \n",
       "AACGTTGGTGTCCTCT-1-WTDAtest7887999-0         4.801738  \n",
       "AACTCCCTCCTAGTGA-1-WTDAtest7887999-0         7.422694  \n",
       "...                                               ...  \n",
       "TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1       0.000000  \n",
       "TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1       0.000000  \n",
       "TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1       0.000000  \n",
       "TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1       0.000000  \n",
       "TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1       0.000000  \n",
       "\n",
       "[193108 rows x 22 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Ensure the portal fields are not used**<br>\n",
    "**Ensure schema fields are present and values are valid & precise**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assay\n",
      "NOT IN OBS\n",
      "cell_type\n",
      "NOT IN OBS\n",
      "development_stage\n",
      "NOT IN OBS\n",
      "disease\n",
      "NOT IN OBS\n",
      "self_reported_ethnicity\n",
      "NOT IN OBS\n",
      "organism\n",
      "NOT IN OBS\n",
      "sex\n",
      "NOT IN OBS\n",
      "tissue\n",
      "NOT IN OBS\n",
      "assay_ontology_term_id\n",
      "NOT IN OBS\n",
      "cell_type_ontology_term_id\n",
      "NOT IN OBS\n",
      "development_stage_ontology_term_id\n",
      "NOT IN OBS\n",
      "disease_ontology_term_id\n",
      "NOT IN OBS\n",
      "self_reported_ethnicity_ontology_term_id\n",
      "NOT IN OBS\n",
      "organism_ontology_term_id\n",
      "NOT IN OBS\n",
      "sex_ontology_term_id\n",
      "NOT IN OBS\n",
      "tissue_ontology_term_id\n",
      "NOT IN OBS\n",
      "donor_id\n",
      "NOT IN OBS\n",
      "suspension_type\n",
      "NOT IN OBS\n",
      "is_primary_data\n",
      "NOT IN OBS\n"
     ]
    }
   ],
   "source": [
    "for o in full_obs_standards:\n",
    "    print(o)\n",
    "    if o not in adata.obs_keys():\n",
    "        print('NOT IN OBS')\n",
    "    else:\n",
    "        un = adata.obs[o].unique()\n",
    "        if un.dtype == 'category':\n",
    "            print(un.to_list())\n",
    "        else:\n",
    "            print(un.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare the obs layer\n",
    "\n",
    "First we extract the obs layer from the h5ad file that has been provided to us. Then we extract out the obs layer as a csv file. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_obs = adata.obs\n",
    "a_obs.to_csv(\"a_obs_layer.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_obs = adata.raw.obs\n",
    "r_obs.to_csv(\"r_obs_layer.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the excel spreadsheet from the INGEST data submission, and navigate to the sequence_file page.\n",
    "Select the cell_suspension.biomaterial_core.biomaterial_id, cell_suspension.uuid, and library_preparation_protocol.protocol_core.protocol_id and filter for duplicates on those keys. This results in a unique set of cell_suspensions with their related data. \n",
    "\n",
    "See script here: https://github.com/ebi-ait/ingest-cellxgene-submitter#create-obs-layer-from-multiple-cell-suspension-uuids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The manual work here is  matching up unique HCA cell_suspensions and metadata with the  unique samples in the provided-h5ad file. We then generate a final obs_layer.csv which combines HCA metadata with the provided metadata from the contributor in matching rows. There is an opportunity for future automation here for scripts to perform the matching and to provide cell-type ontology terms. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a list of fields which should be in the final obs_layer.csv see (https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/2.0.0/schema.md#obs-cell-metadata)\n",
    "- assay_ontology_term_id\n",
    "- cell_type_ontology_term_id\n",
    "- development_stage_ontology_term_id\n",
    "- disease_ontology_term_id\n",
    "- ethnicity_ontology_term_id\n",
    "- is_primary_data\n",
    "- organism_ontology_term_id\n",
    "- sex_ontology_term_id\n",
    "- tissue_ontology_term_id\n",
    "- donor_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is also some manual work here to select cell-type ontology terms from the free text cell-type provided by the contributors. There is an opportunity for future automation here to get scripts / work with other experts in ontology to do this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once we have the final combined obs_layer, then save it as \"obs_layer.csv\" and save it as a dobs object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/4r/275m7hv11jv8zl5946vzl1c00000gp/T/ipykernel_78384/3991837842.py:1: DtypeWarning: Columns (11,12,16,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  dobs = pd.read_csv(\"a_obs_layer.csv\",sep=\",\")\n"
     ]
    }
   ],
   "source": [
    "dobs = pd.read_csv(\"a_obs_layer.csv\",sep=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Celltypes</th>\n",
       "      <th>Celltypes_master_high</th>\n",
       "      <th>Celltypes_master_higher</th>\n",
       "      <th>Celltypes_master_higher_immune</th>\n",
       "      <th>Celltypes_master_2</th>\n",
       "      <th>Celltypes_master_3</th>\n",
       "      <th>Loc_true</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>scsn</th>\n",
       "      <th>...</th>\n",
       "      <th>body_mass_index</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>author_tissue_label</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>assay_label</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGGTGTGAATA-1-WTDAtest7887999-0</td>\n",
       "      <td>AT1</td>\n",
       "      <td>AT1</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>AT1</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cell</td>\n",
       "      <td>cells</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Lower Left Lobe</td>\n",
       "      <td>UBERON:0008949</td>\n",
       "      <td>Singlecell_10x3prime_v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0002062</td>\n",
       "      <td>type I pneumocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACGGGCAACCGCCA-1-WTDAtest7887999-0</td>\n",
       "      <td>B_plasma_IgA</td>\n",
       "      <td>B_plasma</td>\n",
       "      <td>Immune</td>\n",
       "      <td>B_cell</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cell</td>\n",
       "      <td>cells</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Lower Left Lobe</td>\n",
       "      <td>UBERON:0008949</td>\n",
       "      <td>Singlecell_10x3prime_v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000786</td>\n",
       "      <td>plasma cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AACCATGGTACAGCAG-1-WTDAtest7887999-0</td>\n",
       "      <td>Monocyte_CD16</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cell</td>\n",
       "      <td>cells</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Lower Left Lobe</td>\n",
       "      <td>UBERON:0008949</td>\n",
       "      <td>Singlecell_10x3prime_v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000576</td>\n",
       "      <td>monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACGTTGGTGTCCTCT-1-WTDAtest7887999-0</td>\n",
       "      <td>CD8_EM/EMRA</td>\n",
       "      <td>T_NK</td>\n",
       "      <td>Immune</td>\n",
       "      <td>CD8</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cell</td>\n",
       "      <td>cells</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Lower Left Lobe</td>\n",
       "      <td>UBERON:0008949</td>\n",
       "      <td>Singlecell_10x3prime_v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000913</td>\n",
       "      <td>effector memory CD8-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACTCCCTCCTAGTGA-1-WTDAtest7887999-0</td>\n",
       "      <td>Macro_int</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Macrophage_other</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>d_LowLeftPar</td>\n",
       "      <td>cell</td>\n",
       "      <td>cells</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Lower Left Lobe</td>\n",
       "      <td>UBERON:0008949</td>\n",
       "      <td>Singlecell_10x3prime_v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:1001603</td>\n",
       "      <td>lung macrophage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193103</th>\n",
       "      <td>TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1</td>\n",
       "      <td>T_reg</td>\n",
       "      <td>T_NK</td>\n",
       "      <td>Immune</td>\n",
       "      <td>CD4</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Immune</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>32.5</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Trachea</td>\n",
       "      <td>UBERON:0003126</td>\n",
       "      <td>Singlenuclei_10x3prime_v3</td>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>CL:0000815</td>\n",
       "      <td>regulatory T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193104</th>\n",
       "      <td>TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1</td>\n",
       "      <td>SMG_Mucous</td>\n",
       "      <td>Submucosal_Glands</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Submucosal_Glands</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>32.5</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Trachea</td>\n",
       "      <td>UBERON:0003126</td>\n",
       "      <td>Singlenuclei_10x3prime_v3</td>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>CL:1000272</td>\n",
       "      <td>lung secretory cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193105</th>\n",
       "      <td>TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Epi_Basal</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epi_Basal</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>32.5</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Trachea</td>\n",
       "      <td>UBERON:0003126</td>\n",
       "      <td>Singlenuclei_10x3prime_v3</td>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>CL:0000646</td>\n",
       "      <td>basal cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193106</th>\n",
       "      <td>TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1</td>\n",
       "      <td>SMG_Mucous</td>\n",
       "      <td>Submucosal_Glands</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Submucosal_Glands</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>32.5</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Trachea</td>\n",
       "      <td>UBERON:0003126</td>\n",
       "      <td>Singlenuclei_10x3prime_v3</td>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>CL:1000272</td>\n",
       "      <td>lung secretory cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193107</th>\n",
       "      <td>TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1</td>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>b_Bronchi.2.3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>32.5</td>\n",
       "      <td>True</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>Trachea</td>\n",
       "      <td>UBERON:0003126</td>\n",
       "      <td>Singlenuclei_10x3prime_v3</td>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>CL:0000138</td>\n",
       "      <td>chondrocyte</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193108 rows  39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Unnamed: 0      Celltypes  \\\n",
       "0         AAACCTGGTGTGAATA-1-WTDAtest7887999-0            AT1   \n",
       "1         AAACGGGCAACCGCCA-1-WTDAtest7887999-0   B_plasma_IgA   \n",
       "2         AACCATGGTACAGCAG-1-WTDAtest7887999-0  Monocyte_CD16   \n",
       "3         AACGTTGGTGTCCTCT-1-WTDAtest7887999-0    CD8_EM/EMRA   \n",
       "4         AACTCCCTCCTAGTGA-1-WTDAtest7887999-0      Macro_int   \n",
       "...                                        ...            ...   \n",
       "193103  TTTGGTTGTAATGCGG-1-WSSS_A_LNG8757929-1          T_reg   \n",
       "193104  TTTGGTTGTGCCCTTT-1-WSSS_A_LNG8757929-1     SMG_Mucous   \n",
       "193105  TTTGGTTTCAAGAGTA-1-WSSS_A_LNG8757929-1          Basal   \n",
       "193106  TTTGTTGAGCGAGTCA-1-WSSS_A_LNG8757929-1     SMG_Mucous   \n",
       "193107  TTTGTTGTCTTCTGGC-1-WSSS_A_LNG8757929-1    Chondrocyte   \n",
       "\n",
       "       Celltypes_master_high Celltypes_master_higher  \\\n",
       "0                        AT1               Epithelia   \n",
       "1                   B_plasma                  Immune   \n",
       "2                    Myeloid                  Immune   \n",
       "3                       T_NK                  Immune   \n",
       "4                    Myeloid                  Immune   \n",
       "...                      ...                     ...   \n",
       "193103                  T_NK                  Immune   \n",
       "193104     Submucosal_Glands               Epithelia   \n",
       "193105             Epi_Basal               Epithelia   \n",
       "193106     Submucosal_Glands               Epithelia   \n",
       "193107           Chondrocyte                  Stroma   \n",
       "\n",
       "       Celltypes_master_higher_immune Celltypes_master_2 Celltypes_master_3  \\\n",
       "0                                 AT1          Epithelia          Epithelia   \n",
       "1                              B_cell             Immune             Immune   \n",
       "2                            Monocyte             Immune             Immune   \n",
       "3                                 CD8             Immune             Immune   \n",
       "4                    Macrophage_other             Immune             Immune   \n",
       "...                               ...                ...                ...   \n",
       "193103                            CD4             Immune             Immune   \n",
       "193104              Submucosal_Glands          Epithelia          Epithelia   \n",
       "193105                      Epi_Basal          Epithelia          Epithelia   \n",
       "193106              Submucosal_Glands          Epithelia          Epithelia   \n",
       "193107                    Chondrocyte             Stroma             Stroma   \n",
       "\n",
       "             Loc_true suspension_type    scsn  ... body_mass_index  \\\n",
       "0        d_LowLeftPar            cell   cells  ...            22.0   \n",
       "1        d_LowLeftPar            cell   cells  ...            22.0   \n",
       "2        d_LowLeftPar            cell   cells  ...            22.0   \n",
       "3        d_LowLeftPar            cell   cells  ...            22.0   \n",
       "4        d_LowLeftPar            cell   cells  ...            22.0   \n",
       "...               ...             ...     ...  ...             ...   \n",
       "193103  b_Bronchi.2.3         nucleus  nuclei  ...            32.5   \n",
       "193104  b_Bronchi.2.3         nucleus  nuclei  ...            32.5   \n",
       "193105  b_Bronchi.2.3         nucleus  nuclei  ...            32.5   \n",
       "193106  b_Bronchi.2.3         nucleus  nuclei  ...            32.5   \n",
       "193107  b_Bronchi.2.3         nucleus  nuclei  ...            32.5   \n",
       "\n",
       "       is_primary_data organism_ontology_term_id sex_ontology_term_id  \\\n",
       "0                 True            NCBITaxon:9606         PATO:0000384   \n",
       "1                 True            NCBITaxon:9606         PATO:0000384   \n",
       "2                 True            NCBITaxon:9606         PATO:0000384   \n",
       "3                 True            NCBITaxon:9606         PATO:0000384   \n",
       "4                 True            NCBITaxon:9606         PATO:0000384   \n",
       "...                ...                       ...                  ...   \n",
       "193103            True            NCBITaxon:9606         PATO:0000384   \n",
       "193104            True            NCBITaxon:9606         PATO:0000384   \n",
       "193105            True            NCBITaxon:9606         PATO:0000384   \n",
       "193106            True            NCBITaxon:9606         PATO:0000384   \n",
       "193107            True            NCBITaxon:9606         PATO:0000384   \n",
       "\n",
       "       author_tissue_label tissue_ontology_term_id                assay_label  \\\n",
       "0          Lower Left Lobe          UBERON:0008949    Singlecell_10x3prime_v2   \n",
       "1          Lower Left Lobe          UBERON:0008949    Singlecell_10x3prime_v2   \n",
       "2          Lower Left Lobe          UBERON:0008949    Singlecell_10x3prime_v2   \n",
       "3          Lower Left Lobe          UBERON:0008949    Singlecell_10x3prime_v2   \n",
       "4          Lower Left Lobe          UBERON:0008949    Singlecell_10x3prime_v2   \n",
       "...                    ...                     ...                        ...   \n",
       "193103             Trachea          UBERON:0003126  Singlenuclei_10x3prime_v3   \n",
       "193104             Trachea          UBERON:0003126  Singlenuclei_10x3prime_v3   \n",
       "193105             Trachea          UBERON:0003126  Singlenuclei_10x3prime_v3   \n",
       "193106             Trachea          UBERON:0003126  Singlenuclei_10x3prime_v3   \n",
       "193107             Trachea          UBERON:0003126  Singlenuclei_10x3prime_v3   \n",
       "\n",
       "       assay_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0                 EFO:0009899                 CL:0002062   \n",
       "1                 EFO:0009899                 CL:0000786   \n",
       "2                 EFO:0009899                 CL:0000576   \n",
       "3                 EFO:0009899                 CL:0000913   \n",
       "4                 EFO:0009899                 CL:1001603   \n",
       "...                       ...                        ...   \n",
       "193103            EFO:0009922                 CL:0000815   \n",
       "193104            EFO:0009922                 CL:1000272   \n",
       "193105            EFO:0009922                 CL:0000646   \n",
       "193106            EFO:0009922                 CL:1000272   \n",
       "193107            EFO:0009922                 CL:0000138   \n",
       "\n",
       "                          cell_type_ontology_term_label  \n",
       "0                                     type I pneumocyte  \n",
       "1                                           plasma cell  \n",
       "2                                              monocyte  \n",
       "3       effector memory CD8-positive, alpha-beta T cell  \n",
       "4                                       lung macrophage  \n",
       "...                                                 ...  \n",
       "193103                                regulatory T cell  \n",
       "193104                              lung secretory cell  \n",
       "193105                                       basal cell  \n",
       "193106                              lung secretory cell  \n",
       "193107                                      chondrocyte  \n",
       "\n",
       "[193108 rows x 39 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assay\n",
      "NOT IN OBS\n",
      "cell_type\n",
      "NOT IN OBS\n",
      "development_stage\n",
      "NOT IN OBS\n",
      "disease\n",
      "NOT IN OBS\n",
      "self_reported_ethnicity\n",
      "NOT IN OBS\n",
      "organism\n",
      "NOT IN OBS\n",
      "sex\n",
      "NOT IN OBS\n",
      "tissue\n",
      "NOT IN OBS\n",
      "assay_ontology_term_id\n",
      "['EFO:0009899', 'EFO:0011025', 'EFO:0009922']\n",
      "cell_type_ontology_term_id\n",
      "['CL:0002062', 'CL:0000786', 'CL:0000576', 'CL:0000913', 'CL:1001603', 'CL:0002063', 'CL:0000583', 'CL:0002503', 'CL:0001065', 'CL:0001054', 'CL:0000623', 'CL:0000940', 'CL:0000815', 'CL:0001044', 'CL:0000909', 'CL:0002138', 'CL:0000071', 'CL:0000451', 'CL:0000897', 'CL:0000898', 'CL:0000097', 'CL:1000272', 'CL:0000646', 'CL:0000067', 'CL:0000158', 'CL:0002591', 'CL:0000787', 'CL:0000798', 'CL:1001568', 'CL:0000186', 'CL:0002553', 'CL:0000160', 'CL:0000057', 'CL:0009089', 'CL:0000232', 'CL:0000788', 'CL:0000082', 'CL:0000077', 'CL:0002539', 'CL:0000556', 'CL:0005006', 'CL:0000125', 'CL:0000165', 'CL:0005012', 'CL:0002376', 'CL:0000138', 'CL:0000218', 'CL:0000185']\n",
      "development_stage_ontology_term_id\n",
      "['HsapDv:0000123', 'HsapDv:0000169', 'HsapDv:0000153', 'HsapDv:0000152', 'HsapDv:0000158', 'HsapDv:0000129', 'HsapDv:0000160', 'HsapDv:0000240']\n",
      "disease_ontology_term_id\n",
      "['PATO:0000461', 'MONDO:0005148']\n",
      "self_reported_ethnicity_ontology_term_id\n",
      "['HANCESTRO:0005', 'unknown']\n",
      "organism_ontology_term_id\n",
      "['NCBITaxon:9606']\n",
      "sex_ontology_term_id\n",
      "['PATO:0000384', 'PATO:0000383']\n",
      "tissue_ontology_term_id\n",
      "['UBERON:0008949', 'UBERON:0002185', 'UBERON:0008948', 'UBERON:0003126']\n",
      "donor_id\n",
      "['A32', 'A26', 'A37', 'A47', 'A40', 'A43', 'A44', 'A41', 'A48', 'A42']\n",
      "suspension_type\n",
      "['cell', 'nucleus']\n",
      "is_primary_data\n",
      "[True]\n"
     ]
    }
   ],
   "source": [
    "for o in full_obs_standards:\n",
    "    print(o)\n",
    "    if o not in dobs.keys():\n",
    "        print('NOT IN OBS')\n",
    "    else:\n",
    "        un = dobs[o].unique()\n",
    "        if un.dtype == 'category':\n",
    "            print(un.to_list())\n",
    "        else:\n",
    "            print(un.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = dobs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check here for redundant fields in the obs layer (e.g. tissue_label, donor_ext). Assay_ontology_term_id should be '10x v2 5' sequencing' as opposed to '10x TCR'.\n",
    "Development_stage and tissue should be made specific (e.g. 35-year old human stage).\n",
    "Unannotated cells use CL:0000003 (native cell).\n",
    "Check that certain fields in the obs layer should not be continuous / numerical, but should be categorical or string. Use the cells below to fix this if necessary."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** set the index to the \"barcodes\" column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "dobs = dobs.set_index(\"Unnamed: 0\", inplace = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** drop \"author_tissue\" and \"donor_ext\" columns from obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.drop(['Unnamed: 18','author_tissue','tissue_label','donor_ext'], axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** set specific obs datatypes to category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.astype({'batch': 'category','cluster':'category'}, copy = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Ensure the portal fields are not used**<br>\n",
    "**Ensure schema fields are present and values are valid & precise**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for o in full_obs_standards:\n",
    "    print(o)\n",
    "    if o not in adata.obs_keys():\n",
    "        print('NOT IN OBS')\n",
    "    else:\n",
    "        un = adata.obs[o].unique()\n",
    "        if un.dtype == 'category':\n",
    "            print(un.to_list())\n",
    "        else:\n",
    "            print(un.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10x barcode checker\n",
    "**Checks a random 1k barcodes against 10x barcode lists**<br>\n",
    "*Can help confirm 3' v2 vs v3 vs multiome*<br>\n",
    "*5' v1 and v2 kits use the same barcode list as 3' v2*<br>\n",
    "*Assumes the barcode is in the index. Suffixes/prefixes are OK*<br>\n",
    "<br>\n",
    "**Define the function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import re\n",
    "from random import randint\n",
    "\n",
    "\n",
    "def TENx_barcode_checker(df):\n",
    "    num_to_check = 1000\n",
    "\n",
    "    v2_file = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/737K-august-2016.txt'\n",
    "    v3_file = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/3M-february-2018.txt'\n",
    "    multiome_file = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/737K-arc-v1.txt'\n",
    "\n",
    "    v2_list = [line.strip() for line in open(v2_file, 'r')]\n",
    "    v3_list = [line.strip() for line in open(v3_file, 'r')]\n",
    "    multiome_list = [line.strip() for line in open(multiome_file, 'r')]\n",
    "\n",
    "    cellcount = df.index.shape[0]\n",
    "    barcode_pattern = '[ACTG]{16}'\n",
    "    barcode_results = ''\n",
    "    if re.search(barcode_pattern, df.index[5]):\n",
    "        cellcount\n",
    "        random_indices = [randint(0, cellcount - 1) for p in range(0, num_to_check)]\n",
    "        barcodes = {'3pv2_5pv1_5pv2': 0,'3pv3': 0,'multiome': 0,'multiple': 0,'none': 0}\n",
    "        for i in random_indices:\n",
    "            if re.search(barcode_pattern, df.index[i]):\n",
    "                barcode = re.search(barcode_pattern, df.index[i]).group(0)\n",
    "                if barcode in v2_list:\n",
    "                    if barcode in multiome_list:\n",
    "                        barcodes['multiple'] += 1\n",
    "                    elif barcode in v3_list:\n",
    "                        barcodes['multiple'] += 1\n",
    "                    else:\n",
    "                        barcodes['3pv2_5pv1_5pv2'] += 1\n",
    "                elif barcode in multiome_list:\n",
    "                    if barcode in v3_list:\n",
    "                        barcodes['multiple'] += 1\n",
    "                    else:\n",
    "                        barcodes['multiome'] += 1\n",
    "                elif barcode in v3_list:\n",
    "                    barcodes['3pv3'] += 1\n",
    "                else:\n",
    "                    barcodes['none'] += 1\n",
    "        return barcodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Check 1k barcodes across the whole obs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3pv2_5pv1_5pv2</th>\n",
       "      <th>3pv3</th>\n",
       "      <th>multiome</th>\n",
       "      <th>multiple</th>\n",
       "      <th>none</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>594</td>\n",
       "      <td>328</td>\n",
       "      <td>0</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   3pv2_5pv1_5pv2  3pv3  multiome  multiple  none\n",
       "0             594   328         0        78     0"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = TENx_barcode_checker(adata.obs)\n",
    "if not results:\n",
    "    print('no barcodes checked')\n",
    "pd.DataFrame([results])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Additionally, can check 1k barcodes from multiple subsets of obs**<br>\n",
    "*Define `prop` and 1k barcodes will be checked for each unique value in `obs.prop`*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prop = 'assay_ontology_term_id'\n",
    "\n",
    "results = []\n",
    "for a in adata.obs[prop].value_counts().keys():\n",
    "    print(a)\n",
    "    r = TENx_barcode_checker(adata.obs[adata.obs[prop] == a])\n",
    "    if r:\n",
    "        r[prop] = a\n",
    "        results.append(r)\n",
    "    else:\n",
    "        print('no barcodes checked')\n",
    "pd.DataFrame(results).set_index(prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** set a column with all the same values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['is_primary_data'] = True\n",
    "adata.obs['suspension_type'] = 'nucleus'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** change column names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_me = {\n",
    "    'cell_type': 'author_cell_type',\n",
    "    'ethnicity_ontology_id': 'self_reported_ethnicity_ontology_term_id',\n",
    "    'disease_ontology_id': 'disease_ontology_term_id'\n",
    "}\n",
    "\n",
    "adata.obs.rename(columns=rename_me, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** fill null values of a specific column with a specified value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sex_ontology_term_id'].cat.add_categories('unknown', inplace=True)\n",
    "adata.obs.fillna({'sex_ontology_term_id': 'unknown'}, inplace=True)\n",
    "adata.obs['sex_ontology_term_id'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** adjust the values in a specific column in a standard way"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_typo(x):\n",
    "    return x.replace('_',':')\n",
    "\n",
    "\n",
    "adata.obs['development_stage_ontology_term_id'] = adata.obs['development_stage_ontology_term_id'].apply(fix_typo)\n",
    "adata.obs['development_stage_ontology_term_id'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** replace specified values in specified columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "replace_me = {\n",
    "    'organism_ontology_term_id':{'human':'NCBITaxon:9606', 'mouse': 'NCBITaxon:10090'},\n",
    "    'assay_ontology_term_id': {'EFO:0030003': 'EFO:0009899'}\n",
    "}\n",
    "\n",
    "adata.obs.replace(replace_me,inplace=True)\n",
    "adata.obs[['organism_ontology_term_id','assay_ontology_term_id']].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** add a new column with values based on values in an existing column - with DataFrame<br>\n",
    "**Step 1:** get the values to map from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in adata.obs['author_cell_type'].unique():\n",
    "    print(k)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** add a new column with values based on values in an existing column<br>\n",
    "**Step 2:** set up a dataframe with the mapping\n",
    "**Option A:** from a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#map in values based on another field - step 2: set up a dataframe with the mapping\n",
    "#option A: from dict\n",
    "celltypes = {\n",
    "    'Myeloid': 'CL:0001082',\n",
    "    'Endothelial': 'CL:0010008',\n",
    "    'Fibroblast': 'CL:0002548',\n",
    "    'Cardiomyocyte': 'CL:0000513',\n",
    "    'Pericyte': 'CL:0000669',\n",
    "    'Lymphoid': 'CL:0000838',\n",
    "    'Cycling cells': 'CL:0000003',\n",
    "    'vSMCs': 'CL:0000514',\n",
    "    'Neuronal': 'CL:0000006'\n",
    "}\n",
    "\n",
    "ct_df = pd.DataFrame.from_dict(celltypes,orient='index',columns=['cell_type_ontology_term_id']).reset_index().rename(columns={'index':'author_cell_type'})\n",
    "ct_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** add a new column with values based on values in an existing column<br>\n",
    "**Step 2:** set up a dataframe with the mapping<br>\n",
    "**Option B:** from a Google Sheet<br>\n",
    "*Google Sheet permissions must be Anyone with Link is a Viewer*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheet_id = '15oG8v5BS6HMPqCehYQcujMZUq9PgQNpo8osKhO7yA5o'\n",
    "tab_name = 'Sheet1'\n",
    "url = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={tab_name}'\n",
    "ct_df = pd.read_csv(url)\n",
    "ct_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** add a new column with values based on values in an existing column<br>\n",
    "**Step 3:** merge the dataframe into obs<br>\n",
    "*`how='left'` is critical to ensure obs order is retained*<br>\n",
    "*`set_index` is critical to ensure the index is retained*<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.merge(ct_df, on='author_cell_type',how='left').set_index(adata.obs.index)\n",
    "adata.obs[['author_cell_type','cell_type_ontology_term_id']].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** add a new column with values based on values in an existing column - with Dictionary<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_map = {\n",
    "    'KL001': 'P21',\n",
    "    'KL002': 'P22',\n",
    "    'KL003': 'P23'\n",
    "}\n",
    "\n",
    "adata.obs['donor_id'] = adata.obs['sample'].map(donor_map)\n",
    "adata.obs[['donor_id','sample']].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Look for general obs field issues and collect obs information to check for redundant information**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "long_fields = []\n",
    "gradient_fields = []\n",
    "uber_dict = {}\n",
    "for o in adata.obs.keys():\n",
    "    vc_dict = adata.obs[o].value_counts(dropna=False).to_dict()\n",
    "    counts = '_'.join([str(c) for c in vc_dict.values()])\n",
    "    count_len = len(vc_dict.keys())\n",
    "    values = [str(i) for i in vc_dict.keys()]\n",
    "\n",
    "    if o.startswith(' ') or o.endswith(' ') or '  ' in o:\n",
    "        print('leading/trailing whitespace:' + o)\n",
    "\n",
    "    if o not in full_obs_standards and ' '.join(o.split()).lower() in full_obs_standards:\n",
    "        print('schema conflict:' + o)\n",
    "\n",
    "    if count_len == 1:\n",
    "        lone_v = str(list(vc_dict.keys())[0])\n",
    "        if o not in full_obs_standards:\n",
    "            print('all same value:' + o + ',' + lone_v)\n",
    "\n",
    "    numb_types = ['int_', 'int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64','float_', 'float16', 'float32', 'float64']\n",
    "    if adata.obs.dtypes[o].name in numb_types:\n",
    "        gradient_fields.append(o)\n",
    "    #check for long categories as they will not be enabled for coloring\n",
    "    elif count_len > 200:\n",
    "        long_fields.append(o)\n",
    "\n",
    "    #report value_counts to later look for redundancy\n",
    "    metadata = {\n",
    "        'values': values,\n",
    "        'property': o\n",
    "    }\n",
    "    if counts in uber_dict:\n",
    "        uber_dict[counts].append(metadata)\n",
    "    else:\n",
    "        uber_dict[counts] = [metadata]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Comb value_counts to report possible redundancy**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k,v in uber_dict.items():\n",
    "    if '_' in k and not k.startswith('1_1'):\n",
    "        props = [e['property'] for e in v]\n",
    "        if len(v) > 1 and not all(elem in full_obs_standards for elem in props):\n",
    "            print('cells breakdown: ' + k)\n",
    "            for e in v:\n",
    "                print(e['property'])\n",
    "                #print(e['values'])\n",
    "            print('----------------------------------------------------------------------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Investigate any fields that may be redundant**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['sample','patient','age','development_stage_ontology_term_id']].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Check for fields that aren't appropriate as gradient (e.g. cluster number)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gradient_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Update a gradient field to categorical, if needed**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['cluster_id'] = adata.obs['cluster_id'].map(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**List any categorical fields with more than 200 categories as they may not be useful in the visualization**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "long_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**List any obs columns to remove and remove them**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_remove = [\n",
    "    'tissue',\n",
    "    'organism',\n",
    "    'self_reported_ethnicity',\n",
    "    'assay',\n",
    "    'disease',\n",
    "    'sex',\n",
    "    'cell_type'\n",
    "]\n",
    "\n",
    "obs_remove = [o for o in obs_remove if o in adata.obs.columns]\n",
    "adata.obs.drop(columns=obs_remove, inplace=True)\n",
    "if obs_remove:\n",
    "    print('removed: ' + ','.join(obs_remove))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Ensure the portal fields are not used**<br>\n",
    "**Ensure schema fields are present and values are valid & precise**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assay\n",
      "NOT IN OBS\n",
      "cell_type\n",
      "NOT IN OBS\n",
      "development_stage\n",
      "NOT IN OBS\n",
      "disease\n",
      "NOT IN OBS\n",
      "self_reported_ethnicity\n",
      "NOT IN OBS\n",
      "organism\n",
      "NOT IN OBS\n",
      "sex\n",
      "NOT IN OBS\n",
      "tissue\n",
      "NOT IN OBS\n",
      "assay_ontology_term_id\n",
      "['EFO:0009899', 'EFO:0011025', 'EFO:0009922']\n",
      "cell_type_ontology_term_id\n",
      "['CL:0002062', 'CL:0000786', 'CL:0000576', 'CL:0000913', 'CL:1001603', 'CL:0002063', 'CL:0000583', 'CL:0002503', 'CL:0001065', 'CL:0001054', 'CL:0000623', 'CL:0000940', 'CL:0000815', 'CL:0001044', 'CL:0000909', 'CL:0002138', 'CL:0000071', 'CL:0000451', 'CL:0000897', 'CL:0000898', 'CL:0000097', 'CL:1000272', 'CL:0000646', 'CL:0000067', 'CL:0000158', 'CL:0002591', 'CL:0000787', 'CL:0000798', 'CL:1001568', 'CL:0000186', 'CL:0002553', 'CL:0000160', 'CL:0000057', 'CL:0009089', 'CL:0000232', 'CL:0000788', 'CL:0000082', 'CL:0000077', 'CL:0002539', 'CL:0000556', 'CL:0005006', 'CL:0000125', 'CL:0000165', 'CL:0005012', 'CL:0002376', 'CL:0000138', 'CL:0000218', 'CL:0000185']\n",
      "development_stage_ontology_term_id\n",
      "['HsapDv:0000123', 'HsapDv:0000169', 'HsapDv:0000153', 'HsapDv:0000152', 'HsapDv:0000158', 'HsapDv:0000129', 'HsapDv:0000160', 'HsapDv:0000240']\n",
      "disease_ontology_term_id\n",
      "['PATO:0000461', 'MONDO:0005148']\n",
      "self_reported_ethnicity_ontology_term_id\n",
      "['HANCESTRO:0005', 'unknown']\n",
      "organism_ontology_term_id\n",
      "['NCBITaxon:9606']\n",
      "sex_ontology_term_id\n",
      "['PATO:0000384', 'PATO:0000383']\n",
      "tissue_ontology_term_id\n",
      "['UBERON:0008949', 'UBERON:0002185', 'UBERON:0008948', 'UBERON:0003126']\n",
      "donor_id\n",
      "['A32', 'A26', 'A37', 'A47', 'A40', 'A43', 'A44', 'A41', 'A48', 'A42']\n",
      "suspension_type\n",
      "['cell', 'nucleus']\n",
      "is_primary_data\n",
      "[True]\n"
     ]
    }
   ],
   "source": [
    "for o in full_obs_standards:\n",
    "    print(o)\n",
    "    if o not in adata.obs_keys():\n",
    "        print('NOT IN OBS')\n",
    "    else:\n",
    "        un = adata.obs[o].unique()\n",
    "        if un.dtype == 'category':\n",
    "            print(un.to_list())\n",
    "        else:\n",
    "            print(un.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# var\n",
    "**Check for Ensembl IDs, redundant fields, etc.**<br>\n",
    "**Check for schema fields**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map Gene Symbols to ENSEMBL IDs\n",
    "**Skip this section if there are already ENSEMBL IDs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells_by_counts-0</th>\n",
       "      <th>mean_counts-0</th>\n",
       "      <th>pct_dropout_by_counts-0</th>\n",
       "      <th>total_counts-0</th>\n",
       "      <th>gene_ids-1</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells_by_counts-1</th>\n",
       "      <th>...</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HES4</th>\n",
       "      <td>False</td>\n",
       "      <td>HES4</td>\n",
       "      <td>16580</td>\n",
       "      <td>16580</td>\n",
       "      <td>0.302786</td>\n",
       "      <td>88.689619</td>\n",
       "      <td>44385.750000</td>\n",
       "      <td>ENSG00000188290</td>\n",
       "      <td>5369</td>\n",
       "      <td>5369</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>18879</td>\n",
       "      <td>True</td>\n",
       "      <td>0.348015</td>\n",
       "      <td>2.407703</td>\n",
       "      <td>0.858933</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-4.553251e-11</td>\n",
       "      <td>0.512652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNFRSF18</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>5391</td>\n",
       "      <td>5391</td>\n",
       "      <td>0.055908</td>\n",
       "      <td>96.322421</td>\n",
       "      <td>8195.559570</td>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>570</td>\n",
       "      <td>570</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5383</td>\n",
       "      <td>True</td>\n",
       "      <td>0.090583</td>\n",
       "      <td>2.411408</td>\n",
       "      <td>0.957952</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-2.240157e-11</td>\n",
       "      <td>0.252654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNFRSF4</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113</td>\n",
       "      <td>0.054750</td>\n",
       "      <td>96.512064</td>\n",
       "      <td>8025.812988</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5170</td>\n",
       "      <td>True</td>\n",
       "      <td>0.103728</td>\n",
       "      <td>2.507761</td>\n",
       "      <td>1.110371</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>1.818045e-11</td>\n",
       "      <td>0.274860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL645728.1</th>\n",
       "      <td>False</td>\n",
       "      <td>AL645728.1</td>\n",
       "      <td>1166</td>\n",
       "      <td>1166</td>\n",
       "      <td>0.008344</td>\n",
       "      <td>99.204590</td>\n",
       "      <td>1223.174438</td>\n",
       "      <td>ENSG00000215014</td>\n",
       "      <td>367</td>\n",
       "      <td>367</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1251</td>\n",
       "      <td>True</td>\n",
       "      <td>0.020191</td>\n",
       "      <td>2.200457</td>\n",
       "      <td>0.519287</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-9.707694e-12</td>\n",
       "      <td>0.115668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC34</th>\n",
       "      <td>False</td>\n",
       "      <td>TTC34</td>\n",
       "      <td>707</td>\n",
       "      <td>707</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>99.517706</td>\n",
       "      <td>853.601624</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>2557</td>\n",
       "      <td>2557</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1571</td>\n",
       "      <td>True</td>\n",
       "      <td>0.040115</td>\n",
       "      <td>2.175078</td>\n",
       "      <td>0.452723</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-1.262883e-11</td>\n",
       "      <td>0.133543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBASH3A</th>\n",
       "      <td>False</td>\n",
       "      <td>UBASH3A</td>\n",
       "      <td>2057</td>\n",
       "      <td>2057</td>\n",
       "      <td>0.015399</td>\n",
       "      <td>98.596776</td>\n",
       "      <td>2257.323975</td>\n",
       "      <td>ENSG00000160185</td>\n",
       "      <td>1330</td>\n",
       "      <td>1330</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2619</td>\n",
       "      <td>True</td>\n",
       "      <td>0.076424</td>\n",
       "      <td>2.345723</td>\n",
       "      <td>0.911611</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2.340285e-12</td>\n",
       "      <td>0.203516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RSPH1</th>\n",
       "      <td>False</td>\n",
       "      <td>RSPH1</td>\n",
       "      <td>3765</td>\n",
       "      <td>3765</td>\n",
       "      <td>0.404490</td>\n",
       "      <td>97.431629</td>\n",
       "      <td>59294.632812</td>\n",
       "      <td>ENSG00000160188</td>\n",
       "      <td>3671</td>\n",
       "      <td>3671</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5024</td>\n",
       "      <td>True</td>\n",
       "      <td>0.237395</td>\n",
       "      <td>2.757026</td>\n",
       "      <td>1.647012</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-2.383287e-11</td>\n",
       "      <td>0.381427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL18A1</th>\n",
       "      <td>False</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>13232</td>\n",
       "      <td>13232</td>\n",
       "      <td>0.149136</td>\n",
       "      <td>90.973525</td>\n",
       "      <td>21861.937500</td>\n",
       "      <td>ENSG00000182871</td>\n",
       "      <td>6512</td>\n",
       "      <td>6512</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>16562</td>\n",
       "      <td>True</td>\n",
       "      <td>0.321884</td>\n",
       "      <td>2.337140</td>\n",
       "      <td>0.801880</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>9.382796e-12</td>\n",
       "      <td>0.458006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL6A1</th>\n",
       "      <td>False</td>\n",
       "      <td>COL6A1</td>\n",
       "      <td>14127</td>\n",
       "      <td>14127</td>\n",
       "      <td>0.513421</td>\n",
       "      <td>90.362983</td>\n",
       "      <td>75262.882812</td>\n",
       "      <td>ENSG00000142156</td>\n",
       "      <td>10117</td>\n",
       "      <td>10117</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>19712</td>\n",
       "      <td>True</td>\n",
       "      <td>0.514242</td>\n",
       "      <td>2.488781</td>\n",
       "      <td>0.826094</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-8.717349e-12</td>\n",
       "      <td>0.586709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL6A2</th>\n",
       "      <td>False</td>\n",
       "      <td>COL6A2</td>\n",
       "      <td>16267</td>\n",
       "      <td>16267</td>\n",
       "      <td>0.612974</td>\n",
       "      <td>88.903139</td>\n",
       "      <td>89856.523438</td>\n",
       "      <td>ENSG00000142173</td>\n",
       "      <td>10974</td>\n",
       "      <td>10974</td>\n",
       "      <td>...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>22146</td>\n",
       "      <td>True</td>\n",
       "      <td>0.657928</td>\n",
       "      <td>2.699069</td>\n",
       "      <td>1.050830</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.136420e-11</td>\n",
       "      <td>0.685193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1147 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               mt  gene_ids-0  n_cells-0  n_cells_by_counts-0  mean_counts-0  \\\n",
       "HES4        False        HES4      16580                16580       0.302786   \n",
       "TNFRSF18    False    TNFRSF18       5391                 5391       0.055908   \n",
       "TNFRSF4     False     TNFRSF4       5113                 5113       0.054750   \n",
       "AL645728.1  False  AL645728.1       1166                 1166       0.008344   \n",
       "TTC34       False       TTC34        707                  707       0.005823   \n",
       "...           ...         ...        ...                  ...            ...   \n",
       "UBASH3A     False     UBASH3A       2057                 2057       0.015399   \n",
       "RSPH1       False       RSPH1       3765                 3765       0.404490   \n",
       "COL18A1     False     COL18A1      13232                13232       0.149136   \n",
       "COL6A1      False      COL6A1      14127                14127       0.513421   \n",
       "COL6A2      False      COL6A2      16267                16267       0.612974   \n",
       "\n",
       "            pct_dropout_by_counts-0  total_counts-0       gene_ids-1  \\\n",
       "HES4                      88.689619    44385.750000  ENSG00000188290   \n",
       "TNFRSF18                  96.322421     8195.559570  ENSG00000186891   \n",
       "TNFRSF4                   96.512064     8025.812988  ENSG00000186827   \n",
       "AL645728.1                99.204590     1223.174438  ENSG00000215014   \n",
       "TTC34                     99.517706      853.601624  ENSG00000215912   \n",
       "...                             ...             ...              ...   \n",
       "UBASH3A                   98.596776     2257.323975  ENSG00000160185   \n",
       "RSPH1                     97.431629    59294.632812  ENSG00000160188   \n",
       "COL18A1                   90.973525    21861.937500  ENSG00000182871   \n",
       "COL6A1                    90.362983    75262.882812  ENSG00000142156   \n",
       "COL6A2                    88.903139    89856.523438  ENSG00000142173   \n",
       "\n",
       "            n_cells-1  n_cells_by_counts-1  ...  feature_types-1  n_cells  \\\n",
       "HES4             5369                 5369  ...  Gene Expression    18879   \n",
       "TNFRSF18          570                  570  ...  Gene Expression     5383   \n",
       "TNFRSF4           521                  521  ...  Gene Expression     5170   \n",
       "AL645728.1        367                  367  ...  Gene Expression     1251   \n",
       "TTC34            2557                 2557  ...  Gene Expression     1571   \n",
       "...               ...                  ...  ...              ...      ...   \n",
       "UBASH3A          1330                 1330  ...  Gene Expression     2619   \n",
       "RSPH1            3671                 3671  ...  Gene Expression     5024   \n",
       "COL18A1          6512                 6512  ...  Gene Expression    16562   \n",
       "COL6A1          10117                10117  ...  Gene Expression    19712   \n",
       "COL6A2          10974                10974  ...  Gene Expression    22146   \n",
       "\n",
       "            highly_variable     means  dispersions  dispersions_norm  \\\n",
       "HES4                   True  0.348015     2.407703          0.858933   \n",
       "TNFRSF18               True  0.090583     2.411408          0.957952   \n",
       "TNFRSF4                True  0.103728     2.507761          1.110371   \n",
       "AL645728.1             True  0.020191     2.200457          0.519287   \n",
       "TTC34                  True  0.040115     2.175078          0.452723   \n",
       "...                     ...       ...          ...               ...   \n",
       "UBASH3A                True  0.076424     2.345723          0.911611   \n",
       "RSPH1                  True  0.237395     2.757026          1.647012   \n",
       "COL18A1                True  0.321884     2.337140          0.801880   \n",
       "COL6A1                 True  0.514242     2.488781          0.826094   \n",
       "COL6A2                 True  0.657928     2.699069          1.050830   \n",
       "\n",
       "            highly_variable_nbatches  highly_variable_intersection  \\\n",
       "HES4                               2                          True   \n",
       "TNFRSF18                           2                          True   \n",
       "TNFRSF4                            2                          True   \n",
       "AL645728.1                         2                          True   \n",
       "TTC34                              2                          True   \n",
       "...                              ...                           ...   \n",
       "UBASH3A                            2                          True   \n",
       "RSPH1                              2                          True   \n",
       "COL18A1                            2                          True   \n",
       "COL6A1                             2                          True   \n",
       "COL6A2                             2                          True   \n",
       "\n",
       "                    mean       std  \n",
       "HES4       -4.553251e-11  0.512652  \n",
       "TNFRSF18   -2.240157e-11  0.252654  \n",
       "TNFRSF4     1.818045e-11  0.274860  \n",
       "AL645728.1 -9.707694e-12  0.115668  \n",
       "TTC34      -1.262883e-11  0.133543  \n",
       "...                  ...       ...  \n",
       "UBASH3A     2.340285e-12  0.203516  \n",
       "RSPH1      -2.383287e-11  0.381427  \n",
       "COL18A1     9.382796e-12  0.458006  \n",
       "COL6A1     -8.717349e-12  0.586709  \n",
       "COL6A2      4.136420e-11  0.685193  \n",
       "\n",
       "[1147 rows x 23 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells_by_counts-0</th>\n",
       "      <th>mean_counts-0</th>\n",
       "      <th>pct_dropout_by_counts-0</th>\n",
       "      <th>total_counts-0</th>\n",
       "      <th>gene_ids-1</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells_by_counts-1</th>\n",
       "      <th>...</th>\n",
       "      <th>pct_dropout_by_counts-1</th>\n",
       "      <th>total_counts-1</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FAM87B</th>\n",
       "      <td>False</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>176</td>\n",
       "      <td>176</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>99.879938</td>\n",
       "      <td>178.966705</td>\n",
       "      <td>ENSG00000177757</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>99.966985</td>\n",
       "      <td>37.157669</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>196</td>\n",
       "      <td>False</td>\n",
       "      <td>0.001895</td>\n",
       "      <td>1.718146</td>\n",
       "      <td>-0.593497</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>False</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>2468</td>\n",
       "      <td>2468</td>\n",
       "      <td>0.017432</td>\n",
       "      <td>98.316404</td>\n",
       "      <td>2555.424072</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>911</td>\n",
       "      <td>911</td>\n",
       "      <td>...</td>\n",
       "      <td>99.228803</td>\n",
       "      <td>851.700073</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2926</td>\n",
       "      <td>False</td>\n",
       "      <td>0.034289</td>\n",
       "      <td>1.777689</td>\n",
       "      <td>-0.360488</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM41C</th>\n",
       "      <td>False</td>\n",
       "      <td>FAM41C</td>\n",
       "      <td>955</td>\n",
       "      <td>955</td>\n",
       "      <td>0.006955</td>\n",
       "      <td>99.348528</td>\n",
       "      <td>1019.574158</td>\n",
       "      <td>ENSG00000230368</td>\n",
       "      <td>175</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>99.851856</td>\n",
       "      <td>165.668961</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>986</td>\n",
       "      <td>False</td>\n",
       "      <td>0.010345</td>\n",
       "      <td>1.972112</td>\n",
       "      <td>-0.150652</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD11</th>\n",
       "      <td>False</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1750</td>\n",
       "      <td>1750</td>\n",
       "      <td>0.019036</td>\n",
       "      <td>98.806202</td>\n",
       "      <td>2790.467041</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1109</td>\n",
       "      <td>1109</td>\n",
       "      <td>...</td>\n",
       "      <td>99.061188</td>\n",
       "      <td>1118.028442</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2284</td>\n",
       "      <td>False</td>\n",
       "      <td>0.035187</td>\n",
       "      <td>1.811885</td>\n",
       "      <td>-0.101987</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>False</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>17570</td>\n",
       "      <td>17570</td>\n",
       "      <td>0.147330</td>\n",
       "      <td>88.014271</td>\n",
       "      <td>21597.226562</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>5334</td>\n",
       "      <td>5334</td>\n",
       "      <td>...</td>\n",
       "      <td>95.484559</td>\n",
       "      <td>5078.837891</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>19626</td>\n",
       "      <td>False</td>\n",
       "      <td>0.212764</td>\n",
       "      <td>1.889519</td>\n",
       "      <td>-0.077502</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>False</td>\n",
       "      <td>AC023491.2</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>99.976124</td>\n",
       "      <td>38.998932</td>\n",
       "      <td>ENSG00000278633</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>99.977143</td>\n",
       "      <td>26.716158</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>1.519368</td>\n",
       "      <td>-0.584777</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004556.1</th>\n",
       "      <td>False</td>\n",
       "      <td>AC004556.1</td>\n",
       "      <td>2975</td>\n",
       "      <td>2975</td>\n",
       "      <td>0.025639</td>\n",
       "      <td>97.970544</td>\n",
       "      <td>3758.375000</td>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>354</td>\n",
       "      <td>354</td>\n",
       "      <td>...</td>\n",
       "      <td>99.700325</td>\n",
       "      <td>266.433228</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2949</td>\n",
       "      <td>False</td>\n",
       "      <td>0.040713</td>\n",
       "      <td>2.134076</td>\n",
       "      <td>0.372613</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>False</td>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>104</td>\n",
       "      <td>104</td>\n",
       "      <td>0.064795</td>\n",
       "      <td>99.929054</td>\n",
       "      <td>9498.426758</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>99.991535</td>\n",
       "      <td>24.979130</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>90</td>\n",
       "      <td>False</td>\n",
       "      <td>0.021040</td>\n",
       "      <td>4.567000</td>\n",
       "      <td>4.135742</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>False</td>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>0.135452</td>\n",
       "      <td>99.913364</td>\n",
       "      <td>19856.105469</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>99.989842</td>\n",
       "      <td>12.580947</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>113</td>\n",
       "      <td>False</td>\n",
       "      <td>0.029828</td>\n",
       "      <td>4.531441</td>\n",
       "      <td>3.646755</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>False</td>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>3564</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.026531</td>\n",
       "      <td>97.568746</td>\n",
       "      <td>3889.149414</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>1964</td>\n",
       "      <td>1964</td>\n",
       "      <td>...</td>\n",
       "      <td>98.337397</td>\n",
       "      <td>1933.262695</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4611</td>\n",
       "      <td>False</td>\n",
       "      <td>0.056425</td>\n",
       "      <td>1.857525</td>\n",
       "      <td>-0.079161</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20922 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               mt  gene_ids-0  n_cells-0  n_cells_by_counts-0  mean_counts-0  \\\n",
       "FAM87B      False      FAM87B        176                  176       0.001221   \n",
       "LINC00115   False   LINC00115       2468                 2468       0.017432   \n",
       "FAM41C      False      FAM41C        955                  955       0.006955   \n",
       "SAMD11      False      SAMD11       1750                 1750       0.019036   \n",
       "NOC2L       False       NOC2L      17570                17570       0.147330   \n",
       "...           ...         ...        ...                  ...            ...   \n",
       "AC023491.2  False  AC023491.2         35                   35       0.000266   \n",
       "AC004556.1  False  AC004556.1       2975                 2975       0.025639   \n",
       "AC233755.2  False  AC233755.2        104                  104       0.064795   \n",
       "AC233755.1  False  AC233755.1        127                  127       0.135452   \n",
       "AC240274.1  False  AC240274.1       3564                 3564       0.026531   \n",
       "\n",
       "            pct_dropout_by_counts-0  total_counts-0       gene_ids-1  \\\n",
       "FAM87B                    99.879938      178.966705  ENSG00000177757   \n",
       "LINC00115                 98.316404     2555.424072  ENSG00000225880   \n",
       "FAM41C                    99.348528     1019.574158  ENSG00000230368   \n",
       "SAMD11                    98.806202     2790.467041  ENSG00000187634   \n",
       "NOC2L                     88.014271    21597.226562  ENSG00000188976   \n",
       "...                             ...             ...              ...   \n",
       "AC023491.2                99.976124       38.998932  ENSG00000278633   \n",
       "AC004556.1                97.970544     3758.375000  ENSG00000276345   \n",
       "AC233755.2                99.929054     9498.426758  ENSG00000277856   \n",
       "AC233755.1                99.913364    19856.105469  ENSG00000275063   \n",
       "AC240274.1                97.568746     3889.149414  ENSG00000271254   \n",
       "\n",
       "            n_cells-1  n_cells_by_counts-1  ...  pct_dropout_by_counts-1  \\\n",
       "FAM87B             39                   39  ...                99.966985   \n",
       "LINC00115         911                  911  ...                99.228803   \n",
       "FAM41C            175                  175  ...                99.851856   \n",
       "SAMD11           1109                 1109  ...                99.061188   \n",
       "NOC2L            5334                 5334  ...                95.484559   \n",
       "...               ...                  ...  ...                      ...   \n",
       "AC023491.2         27                   27  ...                99.977143   \n",
       "AC004556.1        354                  354  ...                99.700325   \n",
       "AC233755.2         10                   10  ...                99.991535   \n",
       "AC233755.1         12                   12  ...                99.989842   \n",
       "AC240274.1       1964                 1964  ...                98.337397   \n",
       "\n",
       "            total_counts-1  feature_types-1 n_cells  highly_variable  \\\n",
       "FAM87B           37.157669  Gene Expression     196            False   \n",
       "LINC00115       851.700073  Gene Expression    2926            False   \n",
       "FAM41C          165.668961  Gene Expression     986            False   \n",
       "SAMD11         1118.028442  Gene Expression    2284            False   \n",
       "NOC2L          5078.837891  Gene Expression   19626            False   \n",
       "...                    ...              ...     ...              ...   \n",
       "AC023491.2       26.716158  Gene Expression      40            False   \n",
       "AC004556.1      266.433228  Gene Expression    2949            False   \n",
       "AC233755.2       24.979130  Gene Expression      90            False   \n",
       "AC233755.1       12.580947  Gene Expression     113            False   \n",
       "AC240274.1     1933.262695  Gene Expression    4611            False   \n",
       "\n",
       "               means  dispersions  dispersions_norm  highly_variable_nbatches  \\\n",
       "FAM87B      0.001895     1.718146         -0.593497                         0   \n",
       "LINC00115   0.034289     1.777689         -0.360488                         0   \n",
       "FAM41C      0.010345     1.972112         -0.150652                         1   \n",
       "SAMD11      0.035187     1.811885         -0.101987                         0   \n",
       "NOC2L       0.212764     1.889519         -0.077502                         0   \n",
       "...              ...          ...               ...                       ...   \n",
       "AC023491.2  0.000331     1.519368         -0.584777                         0   \n",
       "AC004556.1  0.040713     2.134076          0.372613                         1   \n",
       "AC233755.2  0.021040     4.567000          4.135742                         1   \n",
       "AC233755.1  0.029828     4.531441          3.646755                         1   \n",
       "AC240274.1  0.056425     1.857525         -0.079161                         0   \n",
       "\n",
       "            highly_variable_intersection  \n",
       "FAM87B                             False  \n",
       "LINC00115                          False  \n",
       "FAM41C                             False  \n",
       "SAMD11                             False  \n",
       "NOC2L                              False  \n",
       "...                                  ...  \n",
       "AC023491.2                         False  \n",
       "AC004556.1                         False  \n",
       "AC233755.2                         False  \n",
       "AC233755.1                         False  \n",
       "AC240274.1                         False  \n",
       "\n",
       "[20922 rows x 21 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_adata = ad.AnnData(adata.raw.X, var=adata.raw.var, obs=adata.obs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Set ENSEMBL IDs as barcodes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var = adata.var.set_index(\"gene_ids-1\", inplace = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_adata.var = raw_adata.var.set_index(\"gene_ids-1\", inplace = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**If CellRanger may have been used for alignment, check against the default CellRanger references for matches in order to inform symbol-to-ID mapping**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CR_12 = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/refdata-cellranger-GRCh38-1_2_0_genes_gtf.tsv'\n",
    "CR_30 = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/refdata-cellranger-GRCh38-3_0_0_genes_gtf.tsv'\n",
    "CR_2020 = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/refdata-gex-GRCh38-2020-A_genes_gtf.tsv'\n",
    "CR_hg19 = '/Users/wteh/Documents/Wrangling/Cellxgene/Notebooks/refdata-cellranger-hg19-1_2_0_genes_gtf.tsv'\n",
    "for v in [CR_12,CR_30,CR_2020,CR_hg19]:\n",
    "    map_df = pd.read_csv(v, sep='\\t')\n",
    "    print(v)\n",
    "    print(adata.var.merge(map_df,left_index=True,right_on='gene_symbols',how='inner').shape[0])\n",
    "    print('----------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fill in the mapping file to use to map symbols to Ensembl IDs**<br>\n",
    "*Expecting a .tsv with columns `gene_symbols` & `gene_ids`*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_mapping_file = CR_12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**View what features are not mapped in this**<br>\n",
    "*Check for typos or other alterations to the symbols that can be fixed*<br>\n",
    "*Common to see many ending in `.1` or `-1` resulting from duplicated symbols in the reference*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Shows list of gene features that are not properly mapped "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_map_df = pd.read_csv(var_mapping_file, sep='\\t')\n",
    "adata.var[adata.var.index.isin(var_map_df['gene_symbols']) != True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Filtered out genes from raw to normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Skip this section if there is the identical number of genes in raw and normalized (if cell below = True)**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- If any genes genes have been filtered out when the authors processed the raw matrix, they will not available in the annotated matrix. We need to add them to the processed matrix. We checked this by inspecting the matrices, but can be checked again by running the following cell:\n",
    "- Check that both dataframes have the same number of rows. If they are different, the authors filtered out some genes from the PROCESSED, and we will need to add them in. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dvar = pd.DataFrame(data=adata.var)\n",
    "rvar = pd.DataFrame(data=raw_adata.var)\n",
    "dvar.shape[0] == rvar.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If some of the features have been filtered out of the processed matrices, we have to add the filtered-out genes at the end of the matrices. For that, we are going to first fill in the *feature_is_filtered* column at the rvar dataframe. We can then create a new dataframe dropping all the non filtered gene, and add this dataframe with the filtered genes at the end of dvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_add = [x for x in rvar.index.to_list() if x not in adata.var.index.to_list()]\n",
    "all_genes = adata.var.index.to_list()\n",
    "all_genes.extend(genes_add)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSG00000177757',\n",
       " 'ENSG00000225880',\n",
       " 'ENSG00000230368',\n",
       " 'ENSG00000187634',\n",
       " 'ENSG00000188976',\n",
       " 'ENSG00000187961',\n",
       " 'ENSG00000187583',\n",
       " 'ENSG00000187642',\n",
       " 'ENSG00000187608',\n",
       " 'ENSG00000188157',\n",
       " 'ENSG00000237330',\n",
       " 'ENSG00000131591',\n",
       " 'ENSG00000223823',\n",
       " 'ENSG00000205231',\n",
       " 'ENSG00000162571',\n",
       " 'ENSG00000078808',\n",
       " 'ENSG00000176022',\n",
       " 'ENSG00000160087',\n",
       " 'ENSG00000162572',\n",
       " 'ENSG00000131584',\n",
       " 'ENSG00000169972',\n",
       " 'ENSG00000224051',\n",
       " 'ENSG00000169962',\n",
       " 'ENSG00000107404',\n",
       " 'ENSG00000162576',\n",
       " 'ENSG00000175756',\n",
       " 'ENSG00000221978',\n",
       " 'ENSG00000242485',\n",
       " 'ENSG00000235098',\n",
       " 'ENSG00000205116',\n",
       " 'ENSG00000179403',\n",
       " 'ENSG00000215915',\n",
       " 'ENSG00000160072',\n",
       " 'ENSG00000197785',\n",
       " 'ENSG00000205090',\n",
       " 'ENSG00000160075',\n",
       " 'ENSG00000197530',\n",
       " 'ENSG00000189409',\n",
       " 'ENSG00000248333',\n",
       " 'ENSG00000189339',\n",
       " 'ENSG00000008128',\n",
       " 'ENSG00000008130',\n",
       " 'ENSG00000078369',\n",
       " 'ENSG00000169885',\n",
       " 'ENSG00000178821',\n",
       " 'ENSG00000142609',\n",
       " 'ENSG00000187730',\n",
       " 'ENSG00000067606',\n",
       " 'ENSG00000162585',\n",
       " 'ENSG00000157933',\n",
       " 'ENSG00000116151',\n",
       " 'ENSG00000157916',\n",
       " 'ENSG00000157911',\n",
       " 'ENSG00000149527',\n",
       " 'ENSG00000157881',\n",
       " 'ENSG00000197921',\n",
       " 'ENSG00000157873',\n",
       " 'ENSG00000157870',\n",
       " 'ENSG00000142606',\n",
       " 'ENSG00000177133',\n",
       " 'ENSG00000142611',\n",
       " 'ENSG00000130762',\n",
       " 'ENSG00000162591',\n",
       " 'ENSG00000158109',\n",
       " 'ENSG00000116213',\n",
       " 'ENSG00000078900',\n",
       " 'ENSG00000162592',\n",
       " 'ENSG00000235169',\n",
       " 'ENSG00000130764',\n",
       " 'ENSG00000116198',\n",
       " 'ENSG00000169598',\n",
       " 'ENSG00000198912',\n",
       " 'ENSG00000236423',\n",
       " 'ENSG00000233304',\n",
       " 'ENSG00000196581',\n",
       " 'ENSG00000131697',\n",
       " 'ENSG00000069424',\n",
       " 'ENSG00000116254',\n",
       " 'ENSG00000116251',\n",
       " 'ENSG00000158286',\n",
       " 'ENSG00000116237',\n",
       " 'ENSG00000225077',\n",
       " 'ENSG00000158292',\n",
       " 'ENSG00000097021',\n",
       " 'ENSG00000069812',\n",
       " 'ENSG00000187017',\n",
       " 'ENSG00000171680',\n",
       " 'ENSG00000162408',\n",
       " 'ENSG00000173662',\n",
       " 'ENSG00000204859',\n",
       " 'ENSG00000162413',\n",
       " 'ENSG00000116273',\n",
       " 'ENSG00000041988',\n",
       " 'ENSG00000007923',\n",
       " 'ENSG00000171735',\n",
       " 'ENSG00000049245',\n",
       " 'ENSG00000049246',\n",
       " 'ENSG00000049247',\n",
       " 'ENSG00000116288',\n",
       " 'ENSG00000162426',\n",
       " 'ENSG00000142599',\n",
       " 'ENSG00000074800',\n",
       " 'ENSG00000230679',\n",
       " 'ENSG00000131686',\n",
       " 'ENSG00000197241',\n",
       " 'ENSG00000180758',\n",
       " 'ENSG00000228526',\n",
       " 'ENSG00000049239',\n",
       " 'ENSG00000171621',\n",
       " 'ENSG00000171612',\n",
       " 'ENSG00000188807',\n",
       " 'ENSG00000171608',\n",
       " 'ENSG00000179840',\n",
       " 'ENSG00000231789',\n",
       " 'ENSG00000171603',\n",
       " 'ENSG00000178585',\n",
       " 'ENSG00000223989',\n",
       " 'ENSG00000162441',\n",
       " 'ENSG00000173614',\n",
       " 'ENSG00000162444',\n",
       " 'ENSG00000130939',\n",
       " 'ENSG00000054523',\n",
       " 'ENSG00000142657',\n",
       " 'ENSG00000241563',\n",
       " 'ENSG00000160049',\n",
       " 'ENSG00000142655',\n",
       " 'ENSG00000130940',\n",
       " 'ENSG00000175262',\n",
       " 'ENSG00000120948',\n",
       " 'ENSG00000009724',\n",
       " 'ENSG00000116649',\n",
       " 'ENSG00000171824',\n",
       " 'ENSG00000198793',\n",
       " 'ENSG00000225602',\n",
       " 'ENSG00000171819',\n",
       " 'ENSG00000120942',\n",
       " 'ENSG00000116661',\n",
       " 'ENSG00000132879',\n",
       " 'ENSG00000116663',\n",
       " 'ENSG00000116670',\n",
       " 'ENSG00000162490',\n",
       " 'ENSG00000177674',\n",
       " 'ENSG00000215910',\n",
       " 'ENSG00000177000',\n",
       " 'ENSG00000011021',\n",
       " 'ENSG00000242349',\n",
       " 'ENSG00000175206',\n",
       " 'ENSG00000120937',\n",
       " 'ENSG00000116685',\n",
       " 'ENSG00000083444',\n",
       " 'ENSG00000116688',\n",
       " 'ENSG00000116691',\n",
       " 'ENSG00000120949',\n",
       " 'ENSG00000028137',\n",
       " 'ENSG00000048707',\n",
       " 'ENSG00000162496',\n",
       " 'ENSG00000204518',\n",
       " 'ENSG00000188984',\n",
       " 'ENSG00000157330',\n",
       " 'ENSG00000179172',\n",
       " 'ENSG00000275774',\n",
       " 'ENSG00000204478',\n",
       " 'ENSG00000162494',\n",
       " 'ENSG00000162493',\n",
       " 'ENSG00000116731',\n",
       " 'ENSG00000189337',\n",
       " 'ENSG00000175147',\n",
       " 'ENSG00000171729',\n",
       " 'ENSG00000204464',\n",
       " 'ENSG00000142621',\n",
       " 'ENSG00000142634',\n",
       " 'ENSG00000162438',\n",
       " 'ENSG00000142615',\n",
       " 'ENSG00000215704',\n",
       " 'ENSG00000132906',\n",
       " 'ENSG00000116138',\n",
       " 'ENSG00000116771',\n",
       " 'ENSG00000197312',\n",
       " 'ENSG00000116786',\n",
       " 'ENSG00000162461',\n",
       " 'ENSG00000162460',\n",
       " 'ENSG00000162458',\n",
       " 'ENSG00000233954',\n",
       " 'ENSG00000065526',\n",
       " 'ENSG00000116809',\n",
       " 'ENSG00000173641',\n",
       " 'ENSG00000186510',\n",
       " 'ENSG00000184908',\n",
       " 'ENSG00000185519',\n",
       " 'ENSG00000142627',\n",
       " 'ENSG00000234166',\n",
       " 'ENSG00000142632',\n",
       " 'ENSG00000037637',\n",
       " 'ENSG00000055070',\n",
       " 'ENSG00000187144',\n",
       " 'ENSG00000157191',\n",
       " 'ENSG00000219481',\n",
       " 'ENSG00000058453',\n",
       " 'ENSG00000117122',\n",
       " 'ENSG00000159363',\n",
       " 'ENSG00000117118',\n",
       " 'ENSG00000117115',\n",
       " 'ENSG00000142623',\n",
       " 'ENSG00000142619',\n",
       " 'ENSG00000159339',\n",
       " 'ENSG00000276747',\n",
       " 'ENSG00000179051',\n",
       " 'ENSG00000074964',\n",
       " 'ENSG00000117148',\n",
       " 'ENSG00000179023',\n",
       " 'ENSG00000009709',\n",
       " 'ENSG00000159423',\n",
       " 'ENSG00000169991',\n",
       " 'ENSG00000127481',\n",
       " 'ENSG00000127463',\n",
       " 'ENSG00000053372',\n",
       " 'ENSG00000162482',\n",
       " 'ENSG00000053371',\n",
       " 'ENSG00000040487',\n",
       " 'ENSG00000077549',\n",
       " 'ENSG00000173436',\n",
       " 'ENSG00000158748',\n",
       " 'ENSG00000162542',\n",
       " 'ENSG00000178828',\n",
       " 'ENSG00000169914',\n",
       " 'ENSG00000127472',\n",
       " 'ENSG00000117215',\n",
       " 'ENSG00000158786',\n",
       " 'ENSG00000187980',\n",
       " 'ENSG00000225986',\n",
       " 'ENSG00000162543',\n",
       " 'ENSG00000158816',\n",
       " 'ENSG00000236963',\n",
       " 'ENSG00000090432',\n",
       " 'ENSG00000183114',\n",
       " 'ENSG00000158825',\n",
       " 'ENSG00000158828',\n",
       " 'ENSG00000117242',\n",
       " 'ENSG00000244038',\n",
       " 'ENSG00000117245',\n",
       " 'ENSG00000189410',\n",
       " 'ENSG00000127483',\n",
       " 'ENSG00000075151',\n",
       " 'ENSG00000117298',\n",
       " 'ENSG00000142794',\n",
       " 'ENSG00000076864',\n",
       " 'ENSG00000090686',\n",
       " 'ENSG00000187942',\n",
       " 'ENSG00000142798',\n",
       " 'ENSG00000219073',\n",
       " 'ENSG00000142789',\n",
       " 'ENSG00000070831',\n",
       " 'ENSG00000162552',\n",
       " 'ENSG00000184677',\n",
       " 'ENSG00000070886',\n",
       " 'ENSG00000133216',\n",
       " 'ENSG00000215906',\n",
       " 'ENSG00000004487',\n",
       " 'ENSG00000169641',\n",
       " 'ENSG00000179546',\n",
       " 'ENSG00000261326',\n",
       " 'ENSG00000125944',\n",
       " 'ENSG00000125945',\n",
       " 'ENSG00000249087',\n",
       " 'ENSG00000204219',\n",
       " 'ENSG00000088280',\n",
       " 'ENSG00000007968',\n",
       " 'ENSG00000197880',\n",
       " 'ENSG00000142676',\n",
       " 'ENSG00000057757',\n",
       " 'ENSG00000011009',\n",
       " 'ENSG00000117308',\n",
       " 'ENSG00000117305',\n",
       " 'ENSG00000179163',\n",
       " 'ENSG00000188822',\n",
       " 'ENSG00000189266',\n",
       " 'ENSG00000188529',\n",
       " 'ENSG00000142661',\n",
       " 'ENSG00000142677',\n",
       " 'ENSG00000185436',\n",
       " 'ENSG00000158055',\n",
       " 'ENSG00000001460',\n",
       " 'ENSG00000001461',\n",
       " 'ENSG00000117602',\n",
       " 'ENSG00000184454',\n",
       " 'ENSG00000133226',\n",
       " 'ENSG00000169504',\n",
       " 'ENSG00000117614',\n",
       " 'ENSG00000117616',\n",
       " 'ENSG00000187010',\n",
       " 'ENSG00000183726',\n",
       " 'ENSG00000188672',\n",
       " 'ENSG00000157978',\n",
       " 'ENSG00000117643',\n",
       " 'ENSG00000117640',\n",
       " 'ENSG00000127423',\n",
       " 'ENSG00000182749',\n",
       " 'ENSG00000117632',\n",
       " 'ENSG00000158006',\n",
       " 'ENSG00000158008',\n",
       " 'ENSG00000158014',\n",
       " 'ENSG00000158022',\n",
       " 'ENSG00000175087',\n",
       " 'ENSG00000197245',\n",
       " 'ENSG00000142684',\n",
       " 'ENSG00000142675',\n",
       " 'ENSG00000188782',\n",
       " 'ENSG00000130695',\n",
       " 'ENSG00000142669',\n",
       " 'ENSG00000158062',\n",
       " 'ENSG00000169442',\n",
       " 'ENSG00000131914',\n",
       " 'ENSG00000117682',\n",
       " 'ENSG00000198830',\n",
       " 'ENSG00000117676',\n",
       " 'ENSG00000117713',\n",
       " 'ENSG00000060642',\n",
       " 'ENSG00000204160',\n",
       " 'ENSG00000142751',\n",
       " 'ENSG00000198746',\n",
       " 'ENSG00000090273',\n",
       " 'ENSG00000131910',\n",
       " 'ENSG00000175707',\n",
       " 'ENSG00000253368',\n",
       " 'ENSG00000090020',\n",
       " 'ENSG00000142784',\n",
       " 'ENSG00000186501',\n",
       " 'ENSG00000142765',\n",
       " 'ENSG00000142733',\n",
       " 'ENSG00000174950',\n",
       " 'ENSG00000181773',\n",
       " 'ENSG00000158195',\n",
       " 'ENSG00000126705',\n",
       " 'ENSG00000000938',\n",
       " 'ENSG00000126709',\n",
       " 'ENSG00000009780',\n",
       " 'ENSG00000117758',\n",
       " 'ENSG00000117751',\n",
       " 'ENSG00000130775',\n",
       " 'ENSG00000117748',\n",
       " 'ENSG00000130768',\n",
       " 'ENSG00000158156',\n",
       " 'ENSG00000158161',\n",
       " 'ENSG00000169403',\n",
       " 'ENSG00000126698',\n",
       " 'ENSG00000130766',\n",
       " 'ENSG00000130772',\n",
       " 'ENSG00000204138',\n",
       " 'ENSG00000180198',\n",
       " 'ENSG00000180098',\n",
       " 'ENSG00000197989',\n",
       " 'ENSG00000120656',\n",
       " 'ENSG00000188060',\n",
       " 'ENSG00000162419',\n",
       " 'ENSG00000198492',\n",
       " 'ENSG00000116329',\n",
       " 'ENSG00000159023',\n",
       " 'ENSG00000253304',\n",
       " 'ENSG00000116350',\n",
       " 'ENSG00000116353',\n",
       " 'ENSG00000060656',\n",
       " 'ENSG00000162510',\n",
       " 'ENSG00000186056',\n",
       " 'ENSG00000162511',\n",
       " 'ENSG00000162512',\n",
       " 'ENSG00000134644',\n",
       " 'ENSG00000084628',\n",
       " 'ENSG00000060688',\n",
       " 'ENSG00000121766',\n",
       " 'ENSG00000121769',\n",
       " 'ENSG00000168528',\n",
       " 'ENSG00000142910',\n",
       " 'ENSG00000121764',\n",
       " 'ENSG00000162517',\n",
       " 'ENSG00000084636',\n",
       " 'ENSG00000121753',\n",
       " 'ENSG00000134668',\n",
       " 'ENSG00000184007',\n",
       " 'ENSG00000121774',\n",
       " 'ENSG00000121775',\n",
       " 'ENSG00000025800',\n",
       " 'ENSG00000084652',\n",
       " 'ENSG00000160050',\n",
       " 'ENSG00000160051',\n",
       " 'ENSG00000222046',\n",
       " 'ENSG00000160055',\n",
       " 'ENSG00000084623',\n",
       " 'ENSG00000183615',\n",
       " 'ENSG00000116478',\n",
       " 'ENSG00000175130',\n",
       " 'ENSG00000162526',\n",
       " 'ENSG00000225828',\n",
       " 'ENSG00000160058',\n",
       " 'ENSG00000273274',\n",
       " 'ENSG00000160062',\n",
       " 'ENSG00000176261',\n",
       " 'ENSG00000162521',\n",
       " 'ENSG00000162520',\n",
       " 'ENSG00000162522',\n",
       " 'ENSG00000134684',\n",
       " 'ENSG00000116497',\n",
       " 'ENSG00000160097',\n",
       " 'ENSG00000121905',\n",
       " 'ENSG00000121900',\n",
       " 'ENSG00000116514',\n",
       " 'ENSG00000004455',\n",
       " 'ENSG00000142920',\n",
       " 'ENSG00000116525',\n",
       " 'ENSG00000160094',\n",
       " 'ENSG00000184389',\n",
       " 'ENSG00000134686',\n",
       " 'ENSG00000121903',\n",
       " 'ENSG00000121904',\n",
       " 'ENSG00000231163',\n",
       " 'ENSG00000142698',\n",
       " 'ENSG00000163866',\n",
       " 'ENSG00000189280',\n",
       " 'ENSG00000189433',\n",
       " 'ENSG00000188910',\n",
       " 'ENSG00000116544',\n",
       " 'ENSG00000163867',\n",
       " 'ENSG00000197056',\n",
       " 'ENSG00000116560',\n",
       " 'ENSG00000146463',\n",
       " 'ENSG00000227409',\n",
       " 'ENSG00000142687',\n",
       " 'ENSG00000020129',\n",
       " 'ENSG00000116819',\n",
       " 'ENSG00000126067',\n",
       " 'ENSG00000142686',\n",
       " 'ENSG00000092853',\n",
       " 'ENSG00000134698',\n",
       " 'ENSG00000092847',\n",
       " 'ENSG00000126070',\n",
       " 'ENSG00000092850',\n",
       " 'ENSG00000116863',\n",
       " 'ENSG00000171812',\n",
       " 'ENSG00000054116',\n",
       " 'ENSG00000116871',\n",
       " 'ENSG00000054118',\n",
       " 'ENSG00000214193',\n",
       " 'ENSG00000142694',\n",
       " 'ENSG00000196182',\n",
       " 'ENSG00000181817',\n",
       " 'ENSG00000116885',\n",
       " 'ENSG00000116898',\n",
       " 'ENSG00000119535',\n",
       " 'ENSG00000163873',\n",
       " 'ENSG00000233621',\n",
       " 'ENSG00000163874',\n",
       " 'ENSG00000163875',\n",
       " 'ENSG00000163877',\n",
       " 'ENSG00000163879',\n",
       " 'ENSG00000134697',\n",
       " 'ENSG00000169218',\n",
       " 'ENSG00000116922',\n",
       " 'ENSG00000134690',\n",
       " 'ENSG00000183317',\n",
       " 'ENSG00000185090',\n",
       " 'ENSG00000196449',\n",
       " 'ENSG00000197982',\n",
       " 'ENSG00000188786',\n",
       " 'ENSG00000204084',\n",
       " 'ENSG00000183431',\n",
       " 'ENSG00000183386',\n",
       " 'ENSG00000185668',\n",
       " 'ENSG00000237290',\n",
       " 'ENSG00000116954',\n",
       " 'ENSG00000214114',\n",
       " 'ENSG00000131233',\n",
       " 'ENSG00000158315',\n",
       " 'ENSG00000174574',\n",
       " 'ENSG00000168653',\n",
       " 'ENSG00000127603',\n",
       " 'ENSG00000183682',\n",
       " 'ENSG00000090621',\n",
       " 'ENSG00000163909',\n",
       " 'ENSG00000116981',\n",
       " 'ENSG00000116983',\n",
       " 'ENSG00000084072',\n",
       " 'ENSG00000116985',\n",
       " 'ENSG00000198754',\n",
       " 'ENSG00000043514',\n",
       " 'ENSG00000116990',\n",
       " 'ENSG00000168389',\n",
       " 'ENSG00000131236',\n",
       " 'ENSG00000131238',\n",
       " 'ENSG00000117000',\n",
       " 'ENSG00000188800',\n",
       " 'ENSG00000084073',\n",
       " 'ENSG00000049089',\n",
       " 'ENSG00000084070',\n",
       " 'ENSG00000187801',\n",
       " 'ENSG00000187815',\n",
       " 'ENSG00000164002',\n",
       " 'ENSG00000117010',\n",
       " 'ENSG00000117016',\n",
       " 'ENSG00000272145',\n",
       " 'ENSG00000066136',\n",
       " 'ENSG00000117013',\n",
       " 'ENSG00000179862',\n",
       " 'ENSG00000171793',\n",
       " 'ENSG00000281207',\n",
       " 'ENSG00000171790',\n",
       " 'ENSG00000010803',\n",
       " 'ENSG00000204060',\n",
       " 'ENSG00000127124',\n",
       " 'ENSG00000044012',\n",
       " 'ENSG00000197273',\n",
       " 'ENSG00000198815',\n",
       " 'ENSG00000177181',\n",
       " 'ENSG00000066185',\n",
       " 'ENSG00000127125',\n",
       " 'ENSG00000186409',\n",
       " 'ENSG00000171960',\n",
       " 'ENSG00000065978',\n",
       " 'ENSG00000164007',\n",
       " 'ENSG00000117385',\n",
       " 'ENSG00000164008',\n",
       " 'ENSG00000177868',\n",
       " 'ENSG00000164010',\n",
       " 'ENSG00000164011',\n",
       " 'ENSG00000117394',\n",
       " 'ENSG00000227533',\n",
       " 'ENSG00000186973',\n",
       " 'ENSG00000117395',\n",
       " 'ENSG00000243710',\n",
       " 'ENSG00000179178',\n",
       " 'ENSG00000253313',\n",
       " 'ENSG00000117400',\n",
       " 'ENSG00000117399',\n",
       " 'ENSG00000066322',\n",
       " 'ENSG00000159479',\n",
       " 'ENSG00000198198',\n",
       " 'ENSG00000229372',\n",
       " 'ENSG00000178922',\n",
       " 'ENSG00000142949',\n",
       " 'ENSG00000066135',\n",
       " 'ENSG00000236200',\n",
       " 'ENSG00000126091',\n",
       " 'ENSG00000117407',\n",
       " 'ENSG00000117408',\n",
       " 'ENSG00000132768',\n",
       " 'ENSG00000117410',\n",
       " 'ENSG00000117411',\n",
       " 'ENSG00000159214',\n",
       " 'ENSG00000196517',\n",
       " 'ENSG00000171872',\n",
       " 'ENSG00000178028',\n",
       " 'ENSG00000117419',\n",
       " 'ENSG00000126106',\n",
       " 'ENSG00000142945',\n",
       " 'ENSG00000142937',\n",
       " 'ENSG00000142959',\n",
       " 'ENSG00000173846',\n",
       " 'ENSG00000188396',\n",
       " 'ENSG00000222009',\n",
       " 'ENSG00000117425',\n",
       " 'ENSG00000070785',\n",
       " 'ENSG00000126107',\n",
       " 'ENSG00000126088',\n",
       " 'ENSG00000162415',\n",
       " 'ENSG00000281912',\n",
       " 'ENSG00000186603',\n",
       " 'ENSG00000132781',\n",
       " 'ENSG00000132773',\n",
       " 'ENSG00000070759',\n",
       " 'ENSG00000132763',\n",
       " 'ENSG00000117450',\n",
       " 'ENSG00000117448',\n",
       " 'ENSG00000132780',\n",
       " 'ENSG00000159592',\n",
       " 'ENSG00000159596',\n",
       " 'ENSG00000197429',\n",
       " 'ENSG00000086015',\n",
       " 'ENSG00000117461',\n",
       " 'ENSG00000117472',\n",
       " 'ENSG00000085998',\n",
       " 'ENSG00000171357',\n",
       " 'ENSG00000085999',\n",
       " 'ENSG00000132128',\n",
       " 'ENSG00000173660',\n",
       " 'ENSG00000117481',\n",
       " 'ENSG00000117480',\n",
       " 'ENSG00000197587',\n",
       " 'ENSG00000269956',\n",
       " 'ENSG00000162456',\n",
       " 'ENSG00000079277',\n",
       " 'ENSG00000142961',\n",
       " 'ENSG00000123472',\n",
       " 'ENSG00000186118',\n",
       " 'ENSG00000228237',\n",
       " 'ENSG00000159658',\n",
       " 'ENSG00000142973',\n",
       " 'ENSG00000187048',\n",
       " 'ENSG00000186377',\n",
       " 'ENSG00000186160',\n",
       " 'ENSG00000225506',\n",
       " 'ENSG00000162365',\n",
       " 'ENSG00000224805',\n",
       " 'ENSG00000162366',\n",
       " 'ENSG00000162367',\n",
       " 'ENSG00000123473',\n",
       " 'ENSG00000162368',\n",
       " 'ENSG00000225762',\n",
       " 'ENSG00000186790',\n",
       " 'ENSG00000237424',\n",
       " 'ENSG00000186564',\n",
       " 'ENSG00000269113',\n",
       " 'ENSG00000226133',\n",
       " 'ENSG00000117834',\n",
       " 'ENSG00000132122',\n",
       " 'ENSG00000186094',\n",
       " 'ENSG00000162373',\n",
       " 'ENSG00000162374',\n",
       " 'ENSG00000142700',\n",
       " 'ENSG00000185104',\n",
       " 'ENSG00000123080',\n",
       " 'ENSG00000204006',\n",
       " 'ENSG00000203356',\n",
       " 'ENSG00000123091',\n",
       " 'ENSG00000085831',\n",
       " 'ENSG00000261664',\n",
       " 'ENSG00000085832',\n",
       " 'ENSG00000117859',\n",
       " 'ENSG00000078618',\n",
       " 'ENSG00000169213',\n",
       " 'ENSG00000117862',\n",
       " 'ENSG00000198841',\n",
       " 'ENSG00000228369',\n",
       " 'ENSG00000134717',\n",
       " 'ENSG00000157077',\n",
       " 'ENSG00000154222',\n",
       " 'ENSG00000085840',\n",
       " 'ENSG00000134748',\n",
       " 'ENSG00000116157',\n",
       " 'ENSG00000162377',\n",
       " 'ENSG00000162378',\n",
       " 'ENSG00000203995',\n",
       " 'ENSG00000121310',\n",
       " 'ENSG00000116171',\n",
       " 'ENSG00000232993',\n",
       " 'ENSG00000162383',\n",
       " 'ENSG00000157184',\n",
       " 'ENSG00000162384',\n",
       " 'ENSG00000162385',\n",
       " 'ENSG00000157193',\n",
       " 'ENSG00000143006',\n",
       " 'ENSG00000174332',\n",
       " 'ENSG00000058804',\n",
       " 'ENSG00000058799',\n",
       " 'ENSG00000211452',\n",
       " 'ENSG00000081870',\n",
       " 'ENSG00000116212',\n",
       " 'ENSG00000203985',\n",
       " 'ENSG00000116209',\n",
       " 'ENSG00000116205',\n",
       " 'ENSG00000157211',\n",
       " 'ENSG00000215883',\n",
       " 'ENSG00000116221',\n",
       " 'ENSG00000157216',\n",
       " 'ENSG00000198711',\n",
       " 'ENSG00000162390',\n",
       " 'ENSG00000162391',\n",
       " 'ENSG00000184313',\n",
       " 'ENSG00000271723',\n",
       " 'ENSG00000243725',\n",
       " 'ENSG00000162396',\n",
       " 'ENSG00000006555',\n",
       " 'ENSG00000162398',\n",
       " 'ENSG00000116133',\n",
       " 'ENSG00000143001',\n",
       " 'ENSG00000162399',\n",
       " 'ENSG00000169174',\n",
       " 'ENSG00000162402',\n",
       " 'ENSG00000162407',\n",
       " 'ENSG00000162409',\n",
       " 'ENSG00000157131',\n",
       " 'ENSG00000021852',\n",
       " 'ENSG00000173406',\n",
       " 'ENSG00000226759',\n",
       " 'ENSG00000162600',\n",
       " 'ENSG00000184292',\n",
       " 'ENSG00000162601',\n",
       " 'ENSG00000177606',\n",
       " 'ENSG00000234807',\n",
       " 'ENSG00000237352',\n",
       " 'ENSG00000172456',\n",
       " 'ENSG00000134709',\n",
       " 'ENSG00000134716',\n",
       " 'ENSG00000162598',\n",
       " 'ENSG00000162599',\n",
       " 'ENSG00000237928',\n",
       " 'ENSG00000162604',\n",
       " 'ENSG00000240563',\n",
       " 'ENSG00000132854',\n",
       " 'ENSG00000162607',\n",
       " 'ENSG00000116641',\n",
       " 'ENSG00000132855',\n",
       " 'ENSG00000125703',\n",
       " 'ENSG00000230798',\n",
       " 'ENSG00000187140',\n",
       " 'ENSG00000088035',\n",
       " 'ENSG00000142856',\n",
       " 'ENSG00000203965',\n",
       " 'ENSG00000079739',\n",
       " 'ENSG00000185483',\n",
       " 'ENSG00000223949',\n",
       " 'ENSG00000177414',\n",
       " 'ENSG00000158966',\n",
       " 'ENSG00000162437',\n",
       " 'ENSG00000162434',\n",
       " 'ENSG00000162433',\n",
       " 'ENSG00000116675',\n",
       " 'ENSG00000213625',\n",
       " 'ENSG00000118473',\n",
       " 'ENSG00000152760',\n",
       " 'ENSG00000172410',\n",
       " 'ENSG00000152763',\n",
       " 'ENSG00000198160',\n",
       " 'ENSG00000116704',\n",
       " 'ENSG00000203963',\n",
       " 'ENSG00000142864',\n",
       " 'ENSG00000116717',\n",
       " 'ENSG00000172380',\n",
       " 'ENSG00000232284',\n",
       " 'ENSG00000116729',\n",
       " 'ENSG00000116745',\n",
       " 'ENSG00000024526',\n",
       " 'ENSG00000234264',\n",
       " 'ENSG00000033122',\n",
       " 'ENSG00000066557',\n",
       " 'ENSG00000116754',\n",
       " 'ENSG00000118454',\n",
       " 'ENSG00000197568',\n",
       " 'ENSG00000226088',\n",
       " 'ENSG00000116761',\n",
       " 'ENSG00000050628',\n",
       " 'ENSG00000235079',\n",
       " 'ENSG00000132485',\n",
       " 'ENSG00000172260',\n",
       " 'ENSG00000233973',\n",
       " 'ENSG00000162620',\n",
       " 'ENSG00000254685',\n",
       " 'ENSG00000259030',\n",
       " 'ENSG00000116783',\n",
       " 'ENSG00000162621',\n",
       " 'ENSG00000234497',\n",
       " 'ENSG00000116791',\n",
       " 'ENSG00000162623',\n",
       " 'ENSG00000162624',\n",
       " 'ENSG00000137968',\n",
       " 'ENSG00000117054',\n",
       " 'ENSG00000137955',\n",
       " 'ENSG00000057468',\n",
       " 'ENSG00000154007',\n",
       " 'ENSG00000184005',\n",
       " 'ENSG00000117069',\n",
       " 'ENSG00000142892',\n",
       " 'ENSG00000077254',\n",
       " 'ENSG00000235927',\n",
       " 'ENSG00000162613',\n",
       " 'ENSG00000162616',\n",
       " 'ENSG00000137960',\n",
       " 'ENSG00000122420',\n",
       " 'ENSG00000137959',\n",
       " 'ENSG00000137965',\n",
       " 'ENSG00000230817',\n",
       " 'ENSG00000236268',\n",
       " 'ENSG00000137941',\n",
       " 'ENSG00000142875',\n",
       " 'ENSG00000203943',\n",
       " 'ENSG00000137976',\n",
       " 'ENSG00000117133',\n",
       " 'ENSG00000174021',\n",
       " 'ENSG00000117151',\n",
       " 'ENSG00000180869',\n",
       " 'ENSG00000117155',\n",
       " 'ENSG00000171517',\n",
       " 'ENSG00000162643',\n",
       " 'ENSG00000055732',\n",
       " 'ENSG00000097096',\n",
       " 'ENSG00000162642',\n",
       " 'ENSG00000142867',\n",
       " 'ENSG00000153904',\n",
       " 'ENSG00000117174',\n",
       " 'ENSG00000171502',\n",
       " 'ENSG00000122417',\n",
       " 'ENSG00000137975',\n",
       " 'ENSG00000016490',\n",
       " 'ENSG00000016602',\n",
       " 'ENSG00000097033',\n",
       " 'ENSG00000153936',\n",
       " 'ENSG00000267272',\n",
       " 'ENSG00000143013',\n",
       " 'ENSG00000227290',\n",
       " 'ENSG00000237505',\n",
       " 'ENSG00000065243',\n",
       " 'ENSG00000137947',\n",
       " 'ENSG00000213516',\n",
       " 'ENSG00000117226',\n",
       " 'ENSG00000117228',\n",
       " 'ENSG00000162645',\n",
       " 'ENSG00000213512',\n",
       " 'ENSG00000183347',\n",
       " 'ENSG00000197147',\n",
       " 'ENSG00000171488',\n",
       " 'ENSG00000171492',\n",
       " 'ENSG00000162664',\n",
       " 'ENSG00000122482',\n",
       " 'ENSG00000162669',\n",
       " 'ENSG00000097046',\n",
       " 'ENSG00000069702',\n",
       " 'ENSG00000137948',\n",
       " 'ENSG00000172031',\n",
       " 'ENSG00000230667',\n",
       " 'ENSG00000189195',\n",
       " 'ENSG00000203910',\n",
       " 'ENSG00000174842',\n",
       " 'ENSG00000122484',\n",
       " 'ENSG00000067208',\n",
       " 'ENSG00000122406',\n",
       " 'ENSG00000154511',\n",
       " 'ENSG00000143033',\n",
       " 'ENSG00000117500',\n",
       " 'ENSG00000122483',\n",
       " 'ENSG00000117505',\n",
       " 'ENSG00000137942',\n",
       " 'ENSG00000137936',\n",
       " 'ENSG00000067334',\n",
       " 'ENSG00000023909',\n",
       " 'ENSG00000198691',\n",
       " 'ENSG00000117528',\n",
       " 'ENSG00000143036',\n",
       " 'ENSG00000117519',\n",
       " 'ENSG00000172339',\n",
       " 'ENSG00000152078',\n",
       " 'ENSG00000271092',\n",
       " 'ENSG00000122481',\n",
       " 'ENSG00000117569',\n",
       " 'ENSG00000188641',\n",
       " 'ENSG00000232878',\n",
       " 'ENSG00000235777',\n",
       " 'ENSG00000225206',\n",
       " 'ENSG00000162627',\n",
       " 'ENSG00000117598',\n",
       " 'ENSG00000117600',\n",
       " 'ENSG00000099260',\n",
       " 'ENSG00000156869',\n",
       " 'ENSG00000162688',\n",
       " 'ENSG00000117620',\n",
       " 'ENSG00000156875',\n",
       " 'ENSG00000156876',\n",
       " 'ENSG00000122435',\n",
       " 'ENSG00000122477',\n",
       " 'ENSG00000137992',\n",
       " 'ENSG00000224616',\n",
       " 'ENSG00000137996',\n",
       " 'ENSG00000079335',\n",
       " 'ENSG00000181656',\n",
       " 'ENSG00000230402',\n",
       " 'ENSG00000162694',\n",
       " 'ENSG00000162695',\n",
       " 'ENSG00000117543',\n",
       " 'ENSG00000231671',\n",
       " 'ENSG00000118733',\n",
       " 'ENSG00000060718',\n",
       " 'ENSG00000185946',\n",
       " 'ENSG00000240038',\n",
       " 'ENSG00000243480',\n",
       " 'ENSG00000237763',\n",
       " 'ENSG00000174876',\n",
       " 'ENSG00000187733',\n",
       " 'ENSG00000198890',\n",
       " 'ENSG00000162631',\n",
       " 'ENSG00000134215',\n",
       " 'ENSG00000230489',\n",
       " 'ENSG00000085491',\n",
       " 'ENSG00000196427',\n",
       " 'ENSG00000186086',\n",
       " 'ENSG00000162636',\n",
       " 'ENSG00000162639',\n",
       " 'ENSG00000134186',\n",
       " 'ENSG00000143107',\n",
       " 'ENSG00000116266',\n",
       " 'ENSG00000162641',\n",
       " 'ENSG00000203897',\n",
       " 'ENSG00000121957',\n",
       " 'ENSG00000121940',\n",
       " 'ENSG00000085433',\n",
       " 'ENSG00000197780',\n",
       " 'ENSG00000215717',\n",
       " 'ENSG00000179902',\n",
       " 'ENSG00000116299',\n",
       " 'ENSG00000031698',\n",
       " 'ENSG00000143126',\n",
       " 'ENSG00000134222',\n",
       " 'ENSG00000221986',\n",
       " 'ENSG00000134243',\n",
       " 'ENSG00000143106',\n",
       " 'ENSG00000143028',\n",
       " 'ENSG00000162650',\n",
       " 'ENSG00000174151',\n",
       " 'ENSG00000181754',\n",
       " 'ENSG00000156097',\n",
       " 'ENSG00000065135',\n",
       " 'ENSG00000134183',\n",
       " 'ENSG00000116337',\n",
       " 'ENSG00000168765',\n",
       " 'ENSG00000213366',\n",
       " 'ENSG00000134184',\n",
       " 'ENSG00000134201',\n",
       " 'ENSG00000134202',\n",
       " 'ENSG00000198758',\n",
       " 'ENSG00000184371',\n",
       " 'ENSG00000168710',\n",
       " 'ENSG00000143093',\n",
       " 'ENSG00000156150',\n",
       " 'ENSG00000258673',\n",
       " 'ENSG00000186150',\n",
       " 'ENSG00000197106',\n",
       " 'ENSG00000116396',\n",
       " 'ENSG00000162775',\n",
       " 'ENSG00000168679',\n",
       " 'ENSG00000134248',\n",
       " 'ENSG00000143125',\n",
       " 'ENSG00000143105',\n",
       " 'ENSG00000177301',\n",
       " 'ENSG00000143119',\n",
       " 'ENSG00000121931',\n",
       " 'ENSG00000156171',\n",
       " 'ENSG00000134255',\n",
       " 'ENSG00000162777',\n",
       " 'ENSG00000134216',\n",
       " 'ENSG00000173947',\n",
       " 'ENSG00000085465',\n",
       " 'ENSG00000116455',\n",
       " 'ENSG00000143110',\n",
       " 'ENSG00000121933',\n",
       " 'ENSG00000282608',\n",
       " 'ENSG00000116473',\n",
       " 'ENSG00000231346',\n",
       " 'ENSG00000064703',\n",
       " 'ENSG00000171385',\n",
       " 'ENSG00000237556',\n",
       " 'ENSG00000143079',\n",
       " 'ENSG00000134245',\n",
       " 'ENSG00000007341',\n",
       " 'ENSG00000116489',\n",
       " 'ENSG00000155363',\n",
       " 'ENSG00000155366',\n",
       " 'ENSG00000155367',\n",
       " 'ENSG00000184599',\n",
       " 'ENSG00000215866',\n",
       " 'ENSG00000155380',\n",
       " 'ENSG00000226419',\n",
       " 'ENSG00000198799',\n",
       " 'ENSG00000081026',\n",
       " 'ENSG00000116793',\n",
       " 'ENSG00000081019',\n",
       " 'ENSG00000226167',\n",
       " 'ENSG00000188761',\n",
       " 'ENSG00000134262',\n",
       " 'ENSG00000118655',\n",
       " 'ENSG00000235527',\n",
       " 'ENSG00000163349',\n",
       " 'ENSG00000134207',\n",
       " 'ENSG00000197323',\n",
       " 'ENSG00000116752',\n",
       " 'ENSG00000175984',\n",
       " 'ENSG00000116748',\n",
       " 'ENSG00000213281',\n",
       " 'ENSG00000009307',\n",
       " 'ENSG00000052723',\n",
       " 'ENSG00000198765',\n",
       " 'ENSG00000134200',\n",
       " 'ENSG00000134198',\n",
       " 'ENSG00000134259',\n",
       " 'ENSG00000173218',\n",
       " 'ENSG00000118729',\n",
       " 'ENSG00000177551',\n",
       " 'ENSG00000163393',\n",
       " 'ENSG00000173212',\n",
       " 'ENSG00000163399',\n",
       " 'ENSG00000116815',\n",
       " 'ENSG00000143061',\n",
       " 'ENSG00000203864',\n",
       " 'ENSG00000134247',\n",
       " 'ENSG00000134256',\n",
       " 'ENSG00000116830',\n",
       " 'ENSG00000134253',\n",
       " 'ENSG00000134258',\n",
       " 'ENSG00000235202',\n",
       " 'ENSG00000198162',\n",
       " 'ENSG00000196505',\n",
       " 'ENSG00000065183',\n",
       " 'ENSG00000092607',\n",
       " 'ENSG00000116874',\n",
       " 'ENSG00000224238',\n",
       " 'ENSG00000116882',\n",
       " ...]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_var = pd.DataFrame(index=all_genes)\n",
    "new_var = pd.merge(new_var, rvar, left_index=True, right_index=True, how='left')\n",
    "new_var['feature_is_filtered'] = False\n",
    "new_var.loc[genes_add, 'feature_is_filtered'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells_by_counts-0</th>\n",
       "      <th>mean_counts-0</th>\n",
       "      <th>pct_dropout_by_counts-0</th>\n",
       "      <th>total_counts-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells_by_counts-1</th>\n",
       "      <th>mean_counts-1</th>\n",
       "      <th>...</th>\n",
       "      <th>total_counts-1</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "      <th>feature_is_filtered</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>False</td>\n",
       "      <td>HES4</td>\n",
       "      <td>16580</td>\n",
       "      <td>16580</td>\n",
       "      <td>0.302786</td>\n",
       "      <td>88.689619</td>\n",
       "      <td>44385.750000</td>\n",
       "      <td>5369</td>\n",
       "      <td>5369</td>\n",
       "      <td>0.047040</td>\n",
       "      <td>...</td>\n",
       "      <td>5556.719727</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>18879</td>\n",
       "      <td>True</td>\n",
       "      <td>0.348015</td>\n",
       "      <td>2.407703</td>\n",
       "      <td>0.858933</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>5391</td>\n",
       "      <td>5391</td>\n",
       "      <td>0.055908</td>\n",
       "      <td>96.322421</td>\n",
       "      <td>8195.559570</td>\n",
       "      <td>570</td>\n",
       "      <td>570</td>\n",
       "      <td>0.004692</td>\n",
       "      <td>...</td>\n",
       "      <td>554.286743</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5383</td>\n",
       "      <td>True</td>\n",
       "      <td>0.090583</td>\n",
       "      <td>2.411408</td>\n",
       "      <td>0.957952</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113</td>\n",
       "      <td>0.054750</td>\n",
       "      <td>96.512064</td>\n",
       "      <td>8025.812988</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "      <td>0.004022</td>\n",
       "      <td>...</td>\n",
       "      <td>475.164764</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5170</td>\n",
       "      <td>True</td>\n",
       "      <td>0.103728</td>\n",
       "      <td>2.507761</td>\n",
       "      <td>1.110371</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215014</th>\n",
       "      <td>False</td>\n",
       "      <td>AL645728.1</td>\n",
       "      <td>1166</td>\n",
       "      <td>1166</td>\n",
       "      <td>0.008344</td>\n",
       "      <td>99.204590</td>\n",
       "      <td>1223.174438</td>\n",
       "      <td>367</td>\n",
       "      <td>367</td>\n",
       "      <td>0.002795</td>\n",
       "      <td>...</td>\n",
       "      <td>330.200958</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1251</td>\n",
       "      <td>True</td>\n",
       "      <td>0.020191</td>\n",
       "      <td>2.200457</td>\n",
       "      <td>0.519287</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215912</th>\n",
       "      <td>False</td>\n",
       "      <td>TTC34</td>\n",
       "      <td>707</td>\n",
       "      <td>707</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>99.517706</td>\n",
       "      <td>853.601624</td>\n",
       "      <td>2557</td>\n",
       "      <td>2557</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>...</td>\n",
       "      <td>4058.525635</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1571</td>\n",
       "      <td>True</td>\n",
       "      <td>0.040115</td>\n",
       "      <td>2.175078</td>\n",
       "      <td>0.452723</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278633</th>\n",
       "      <td>False</td>\n",
       "      <td>AC023491.2</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>99.976124</td>\n",
       "      <td>38.998932</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>...</td>\n",
       "      <td>26.716158</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>1.519368</td>\n",
       "      <td>-0.584777</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000276345</th>\n",
       "      <td>False</td>\n",
       "      <td>AC004556.1</td>\n",
       "      <td>2975</td>\n",
       "      <td>2975</td>\n",
       "      <td>0.025639</td>\n",
       "      <td>97.970544</td>\n",
       "      <td>3758.375000</td>\n",
       "      <td>354</td>\n",
       "      <td>354</td>\n",
       "      <td>0.002255</td>\n",
       "      <td>...</td>\n",
       "      <td>266.433228</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2949</td>\n",
       "      <td>False</td>\n",
       "      <td>0.040713</td>\n",
       "      <td>2.134076</td>\n",
       "      <td>0.372613</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>False</td>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>104</td>\n",
       "      <td>104</td>\n",
       "      <td>0.064795</td>\n",
       "      <td>99.929054</td>\n",
       "      <td>9498.426758</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>...</td>\n",
       "      <td>24.979130</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>90</td>\n",
       "      <td>False</td>\n",
       "      <td>0.021040</td>\n",
       "      <td>4.567000</td>\n",
       "      <td>4.135742</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>False</td>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>0.135452</td>\n",
       "      <td>99.913364</td>\n",
       "      <td>19856.105469</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>...</td>\n",
       "      <td>12.580947</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>113</td>\n",
       "      <td>False</td>\n",
       "      <td>0.029828</td>\n",
       "      <td>4.531441</td>\n",
       "      <td>3.646755</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>False</td>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>3564</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.026531</td>\n",
       "      <td>97.568746</td>\n",
       "      <td>3889.149414</td>\n",
       "      <td>1964</td>\n",
       "      <td>1964</td>\n",
       "      <td>0.016366</td>\n",
       "      <td>...</td>\n",
       "      <td>1933.262695</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4611</td>\n",
       "      <td>False</td>\n",
       "      <td>0.056425</td>\n",
       "      <td>1.857525</td>\n",
       "      <td>-0.079161</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20922 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mt  gene_ids-0  n_cells-0  n_cells_by_counts-0  \\\n",
       "ENSG00000188290  False        HES4      16580                16580   \n",
       "ENSG00000186891  False    TNFRSF18       5391                 5391   \n",
       "ENSG00000186827  False     TNFRSF4       5113                 5113   \n",
       "ENSG00000215014  False  AL645728.1       1166                 1166   \n",
       "ENSG00000215912  False       TTC34        707                  707   \n",
       "...                ...         ...        ...                  ...   \n",
       "ENSG00000278633  False  AC023491.2         35                   35   \n",
       "ENSG00000276345  False  AC004556.1       2975                 2975   \n",
       "ENSG00000277856  False  AC233755.2        104                  104   \n",
       "ENSG00000275063  False  AC233755.1        127                  127   \n",
       "ENSG00000271254  False  AC240274.1       3564                 3564   \n",
       "\n",
       "                 mean_counts-0  pct_dropout_by_counts-0  total_counts-0  \\\n",
       "ENSG00000188290       0.302786                88.689619    44385.750000   \n",
       "ENSG00000186891       0.055908                96.322421     8195.559570   \n",
       "ENSG00000186827       0.054750                96.512064     8025.812988   \n",
       "ENSG00000215014       0.008344                99.204590     1223.174438   \n",
       "ENSG00000215912       0.005823                99.517706      853.601624   \n",
       "...                        ...                      ...             ...   \n",
       "ENSG00000278633       0.000266                99.976124       38.998932   \n",
       "ENSG00000276345       0.025639                97.970544     3758.375000   \n",
       "ENSG00000277856       0.064795                99.929054     9498.426758   \n",
       "ENSG00000275063       0.135452                99.913364    19856.105469   \n",
       "ENSG00000271254       0.026531                97.568746     3889.149414   \n",
       "\n",
       "                 n_cells-1  n_cells_by_counts-1  mean_counts-1  ...  \\\n",
       "ENSG00000188290       5369                 5369       0.047040  ...   \n",
       "ENSG00000186891        570                  570       0.004692  ...   \n",
       "ENSG00000186827        521                  521       0.004022  ...   \n",
       "ENSG00000215014        367                  367       0.002795  ...   \n",
       "ENSG00000215912       2557                 2557       0.034357  ...   \n",
       "...                    ...                  ...            ...  ...   \n",
       "ENSG00000278633         27                   27       0.000226  ...   \n",
       "ENSG00000276345        354                  354       0.002255  ...   \n",
       "ENSG00000277856         10                   10       0.000211  ...   \n",
       "ENSG00000275063         12                   12       0.000107  ...   \n",
       "ENSG00000271254       1964                 1964       0.016366  ...   \n",
       "\n",
       "                 total_counts-1  feature_types-1 n_cells  highly_variable  \\\n",
       "ENSG00000188290     5556.719727  Gene Expression   18879             True   \n",
       "ENSG00000186891      554.286743  Gene Expression    5383             True   \n",
       "ENSG00000186827      475.164764  Gene Expression    5170             True   \n",
       "ENSG00000215014      330.200958  Gene Expression    1251             True   \n",
       "ENSG00000215912     4058.525635  Gene Expression    1571             True   \n",
       "...                         ...              ...     ...              ...   \n",
       "ENSG00000278633       26.716158  Gene Expression      40            False   \n",
       "ENSG00000276345      266.433228  Gene Expression    2949            False   \n",
       "ENSG00000277856       24.979130  Gene Expression      90            False   \n",
       "ENSG00000275063       12.580947  Gene Expression     113            False   \n",
       "ENSG00000271254     1933.262695  Gene Expression    4611            False   \n",
       "\n",
       "                    means  dispersions  dispersions_norm  \\\n",
       "ENSG00000188290  0.348015     2.407703          0.858933   \n",
       "ENSG00000186891  0.090583     2.411408          0.957952   \n",
       "ENSG00000186827  0.103728     2.507761          1.110371   \n",
       "ENSG00000215014  0.020191     2.200457          0.519287   \n",
       "ENSG00000215912  0.040115     2.175078          0.452723   \n",
       "...                   ...          ...               ...   \n",
       "ENSG00000278633  0.000331     1.519368         -0.584777   \n",
       "ENSG00000276345  0.040713     2.134076          0.372613   \n",
       "ENSG00000277856  0.021040     4.567000          4.135742   \n",
       "ENSG00000275063  0.029828     4.531441          3.646755   \n",
       "ENSG00000271254  0.056425     1.857525         -0.079161   \n",
       "\n",
       "                 highly_variable_nbatches  highly_variable_intersection  \\\n",
       "ENSG00000188290                         2                          True   \n",
       "ENSG00000186891                         2                          True   \n",
       "ENSG00000186827                         2                          True   \n",
       "ENSG00000215014                         2                          True   \n",
       "ENSG00000215912                         2                          True   \n",
       "...                                   ...                           ...   \n",
       "ENSG00000278633                         0                         False   \n",
       "ENSG00000276345                         1                         False   \n",
       "ENSG00000277856                         1                         False   \n",
       "ENSG00000275063                         1                         False   \n",
       "ENSG00000271254                         0                         False   \n",
       "\n",
       "                 feature_is_filtered  \n",
       "ENSG00000188290                False  \n",
       "ENSG00000186891                False  \n",
       "ENSG00000186827                False  \n",
       "ENSG00000215014                False  \n",
       "ENSG00000215912                False  \n",
       "...                              ...  \n",
       "ENSG00000278633                 True  \n",
       "ENSG00000276345                 True  \n",
       "ENSG00000277856                 True  \n",
       "ENSG00000275063                 True  \n",
       "ENSG00000271254                 True  \n",
       "\n",
       "[20922 rows x 21 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Also editing the X layer to add filtered out genes from raw to normalized matrix**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "dExprs = pd.DataFrame(data=adata.X)\n",
    "dExprs = dExprs.set_axis(dvar.index.to_list(), axis=1, inplace=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <th>ENSG00000215014</th>\n",
       "      <th>ENSG00000215912</th>\n",
       "      <th>ENSG00000215788</th>\n",
       "      <th>ENSG00000049249</th>\n",
       "      <th>ENSG00000116285</th>\n",
       "      <th>ENSG00000142583</th>\n",
       "      <th>ENSG00000117154</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000234883</th>\n",
       "      <th>ENSG00000154734</th>\n",
       "      <th>ENSG00000231106</th>\n",
       "      <th>ENSG00000160180</th>\n",
       "      <th>ENSG00000160182</th>\n",
       "      <th>ENSG00000160185</th>\n",
       "      <th>ENSG00000160188</th>\n",
       "      <th>ENSG00000182871</th>\n",
       "      <th>ENSG00000142156</th>\n",
       "      <th>ENSG00000142173</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.723888</td>\n",
       "      <td>-0.139995</td>\n",
       "      <td>-0.140796</td>\n",
       "      <td>-0.065739</td>\n",
       "      <td>-0.083806</td>\n",
       "      <td>-0.140786</td>\n",
       "      <td>-0.075423</td>\n",
       "      <td>-0.462988</td>\n",
       "      <td>-0.076909</td>\n",
       "      <td>-0.092391</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.174554</td>\n",
       "      <td>-0.302418</td>\n",
       "      <td>-0.108493</td>\n",
       "      <td>-0.258152</td>\n",
       "      <td>-0.057768</td>\n",
       "      <td>-0.098407</td>\n",
       "      <td>-0.161561</td>\n",
       "      <td>-0.272683</td>\n",
       "      <td>-0.335242</td>\n",
       "      <td>-0.350405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.197353</td>\n",
       "      <td>-0.091179</td>\n",
       "      <td>-0.143029</td>\n",
       "      <td>-0.062913</td>\n",
       "      <td>-0.085007</td>\n",
       "      <td>-0.159376</td>\n",
       "      <td>-0.072150</td>\n",
       "      <td>-0.477511</td>\n",
       "      <td>-0.061004</td>\n",
       "      <td>-0.119909</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.141530</td>\n",
       "      <td>-0.298021</td>\n",
       "      <td>-0.086605</td>\n",
       "      <td>-0.134693</td>\n",
       "      <td>-0.013729</td>\n",
       "      <td>-0.130186</td>\n",
       "      <td>-0.015604</td>\n",
       "      <td>-0.272482</td>\n",
       "      <td>-0.286995</td>\n",
       "      <td>-0.333254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.281528</td>\n",
       "      <td>-0.152716</td>\n",
       "      <td>-0.162285</td>\n",
       "      <td>-0.074706</td>\n",
       "      <td>-0.062901</td>\n",
       "      <td>-0.149338</td>\n",
       "      <td>-0.095287</td>\n",
       "      <td>-0.357703</td>\n",
       "      <td>-0.048933</td>\n",
       "      <td>-0.062510</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.223301</td>\n",
       "      <td>-0.258535</td>\n",
       "      <td>-0.077394</td>\n",
       "      <td>-0.203656</td>\n",
       "      <td>-0.069490</td>\n",
       "      <td>-0.114936</td>\n",
       "      <td>-0.137211</td>\n",
       "      <td>-0.249522</td>\n",
       "      <td>-0.229784</td>\n",
       "      <td>-0.271228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.267558</td>\n",
       "      <td>-0.142068</td>\n",
       "      <td>-0.158585</td>\n",
       "      <td>-0.072528</td>\n",
       "      <td>-0.067069</td>\n",
       "      <td>-0.150852</td>\n",
       "      <td>-0.090966</td>\n",
       "      <td>-0.380127</td>\n",
       "      <td>-0.051557</td>\n",
       "      <td>-0.072820</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.208483</td>\n",
       "      <td>-0.266118</td>\n",
       "      <td>-0.079602</td>\n",
       "      <td>-0.193169</td>\n",
       "      <td>-0.059837</td>\n",
       "      <td>-0.117161</td>\n",
       "      <td>-0.117210</td>\n",
       "      <td>-0.253882</td>\n",
       "      <td>-0.241652</td>\n",
       "      <td>-0.283356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.298954</td>\n",
       "      <td>-0.170585</td>\n",
       "      <td>-0.172224</td>\n",
       "      <td>-0.079752</td>\n",
       "      <td>-0.052425</td>\n",
       "      <td>-0.149062</td>\n",
       "      <td>-0.105750</td>\n",
       "      <td>-0.302920</td>\n",
       "      <td>-0.039100</td>\n",
       "      <td>-0.041369</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.254953</td>\n",
       "      <td>-0.238199</td>\n",
       "      <td>-0.067471</td>\n",
       "      <td>-0.206611</td>\n",
       "      <td>-0.085730</td>\n",
       "      <td>-0.115394</td>\n",
       "      <td>-0.160236</td>\n",
       "      <td>-0.238282</td>\n",
       "      <td>-0.189923</td>\n",
       "      <td>-0.236741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193103</th>\n",
       "      <td>-0.214100</td>\n",
       "      <td>-0.091046</td>\n",
       "      <td>5.200045</td>\n",
       "      <td>-0.058977</td>\n",
       "      <td>-0.094838</td>\n",
       "      <td>-0.153033</td>\n",
       "      <td>9.562877</td>\n",
       "      <td>-0.525991</td>\n",
       "      <td>-0.076291</td>\n",
       "      <td>-0.130821</td>\n",
       "      <td>...</td>\n",
       "      <td>7.411474</td>\n",
       "      <td>-0.319499</td>\n",
       "      <td>-0.104110</td>\n",
       "      <td>-0.175720</td>\n",
       "      <td>-0.013492</td>\n",
       "      <td>-0.118431</td>\n",
       "      <td>-0.044981</td>\n",
       "      <td>-0.283560</td>\n",
       "      <td>-0.343194</td>\n",
       "      <td>-0.373127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193104</th>\n",
       "      <td>-0.267087</td>\n",
       "      <td>-0.107013</td>\n",
       "      <td>-0.118204</td>\n",
       "      <td>-0.054842</td>\n",
       "      <td>-0.107104</td>\n",
       "      <td>-0.138718</td>\n",
       "      <td>-0.052452</td>\n",
       "      <td>-0.105977</td>\n",
       "      <td>-0.101240</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.108407</td>\n",
       "      <td>-0.348619</td>\n",
       "      <td>-0.133893</td>\n",
       "      <td>2.105271</td>\n",
       "      <td>-0.027745</td>\n",
       "      <td>-0.092786</td>\n",
       "      <td>-0.131069</td>\n",
       "      <td>-0.297897</td>\n",
       "      <td>-0.431532</td>\n",
       "      <td>-0.430158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193105</th>\n",
       "      <td>-0.185541</td>\n",
       "      <td>-0.082441</td>\n",
       "      <td>-0.140208</td>\n",
       "      <td>-0.061206</td>\n",
       "      <td>-0.088227</td>\n",
       "      <td>-0.160748</td>\n",
       "      <td>-0.068790</td>\n",
       "      <td>-0.494921</td>\n",
       "      <td>-0.062845</td>\n",
       "      <td>-0.128148</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.129762</td>\n",
       "      <td>-0.303804</td>\n",
       "      <td>-0.088058</td>\n",
       "      <td>-0.125241</td>\n",
       "      <td>-0.005810</td>\n",
       "      <td>-0.132254</td>\n",
       "      <td>0.001419</td>\n",
       "      <td>-0.275834</td>\n",
       "      <td>-0.295581</td>\n",
       "      <td>-0.342390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193106</th>\n",
       "      <td>-0.258789</td>\n",
       "      <td>-0.104512</td>\n",
       "      <td>-0.120443</td>\n",
       "      <td>-0.055490</td>\n",
       "      <td>-0.105183</td>\n",
       "      <td>-0.140960</td>\n",
       "      <td>-0.054115</td>\n",
       "      <td>0.032701</td>\n",
       "      <td>-0.097333</td>\n",
       "      <td>-0.135005</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.110581</td>\n",
       "      <td>-0.344059</td>\n",
       "      <td>-0.129229</td>\n",
       "      <td>1.750257</td>\n",
       "      <td>-0.025513</td>\n",
       "      <td>-0.096802</td>\n",
       "      <td>-0.117587</td>\n",
       "      <td>-0.295652</td>\n",
       "      <td>-0.417698</td>\n",
       "      <td>-0.421226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193107</th>\n",
       "      <td>-0.194356</td>\n",
       "      <td>-0.085097</td>\n",
       "      <td>-0.137830</td>\n",
       "      <td>-0.060518</td>\n",
       "      <td>-0.090268</td>\n",
       "      <td>-0.158367</td>\n",
       "      <td>-0.067024</td>\n",
       "      <td>2.282134</td>\n",
       "      <td>-0.066995</td>\n",
       "      <td>-0.128973</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.127454</td>\n",
       "      <td>-0.308649</td>\n",
       "      <td>-0.093012</td>\n",
       "      <td>-0.140822</td>\n",
       "      <td>-0.008181</td>\n",
       "      <td>-0.127987</td>\n",
       "      <td>-0.012903</td>\n",
       "      <td>-0.278219</td>\n",
       "      <td>-0.310277</td>\n",
       "      <td>-0.351877</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193108 rows  1147 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ENSG00000188290  ENSG00000186891  ENSG00000186827  ENSG00000215014  \\\n",
       "0              2.723888        -0.139995        -0.140796        -0.065739   \n",
       "1             -0.197353        -0.091179        -0.143029        -0.062913   \n",
       "2             -0.281528        -0.152716        -0.162285        -0.074706   \n",
       "3             -0.267558        -0.142068        -0.158585        -0.072528   \n",
       "4             -0.298954        -0.170585        -0.172224        -0.079752   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103        -0.214100        -0.091046         5.200045        -0.058977   \n",
       "193104        -0.267087        -0.107013        -0.118204        -0.054842   \n",
       "193105        -0.185541        -0.082441        -0.140208        -0.061206   \n",
       "193106        -0.258789        -0.104512        -0.120443        -0.055490   \n",
       "193107        -0.194356        -0.085097        -0.137830        -0.060518   \n",
       "\n",
       "        ENSG00000215912  ENSG00000215788  ENSG00000049249  ENSG00000116285  \\\n",
       "0             -0.083806        -0.140786        -0.075423        -0.462988   \n",
       "1             -0.085007        -0.159376        -0.072150        -0.477511   \n",
       "2             -0.062901        -0.149338        -0.095287        -0.357703   \n",
       "3             -0.067069        -0.150852        -0.090966        -0.380127   \n",
       "4             -0.052425        -0.149062        -0.105750        -0.302920   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103        -0.094838        -0.153033         9.562877        -0.525991   \n",
       "193104        -0.107104        -0.138718        -0.052452        -0.105977   \n",
       "193105        -0.088227        -0.160748        -0.068790        -0.494921   \n",
       "193106        -0.105183        -0.140960        -0.054115         0.032701   \n",
       "193107        -0.090268        -0.158367        -0.067024         2.282134   \n",
       "\n",
       "        ENSG00000142583  ENSG00000117154  ...  ENSG00000234883  \\\n",
       "0             -0.076909        -0.092391  ...        -0.174554   \n",
       "1             -0.061004        -0.119909  ...        -0.141530   \n",
       "2             -0.048933        -0.062510  ...        -0.223301   \n",
       "3             -0.051557        -0.072820  ...        -0.208483   \n",
       "4             -0.039100        -0.041369  ...        -0.254953   \n",
       "...                 ...              ...  ...              ...   \n",
       "193103        -0.076291        -0.130821  ...         7.411474   \n",
       "193104        -0.101240        10.000000  ...        -0.108407   \n",
       "193105        -0.062845        -0.128148  ...        -0.129762   \n",
       "193106        -0.097333        -0.135005  ...        -0.110581   \n",
       "193107        -0.066995        -0.128973  ...        -0.127454   \n",
       "\n",
       "        ENSG00000154734  ENSG00000231106  ENSG00000160180  ENSG00000160182  \\\n",
       "0             -0.302418        -0.108493        -0.258152        -0.057768   \n",
       "1             -0.298021        -0.086605        -0.134693        -0.013729   \n",
       "2             -0.258535        -0.077394        -0.203656        -0.069490   \n",
       "3             -0.266118        -0.079602        -0.193169        -0.059837   \n",
       "4             -0.238199        -0.067471        -0.206611        -0.085730   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103        -0.319499        -0.104110        -0.175720        -0.013492   \n",
       "193104        -0.348619        -0.133893         2.105271        -0.027745   \n",
       "193105        -0.303804        -0.088058        -0.125241        -0.005810   \n",
       "193106        -0.344059        -0.129229         1.750257        -0.025513   \n",
       "193107        -0.308649        -0.093012        -0.140822        -0.008181   \n",
       "\n",
       "        ENSG00000160185  ENSG00000160188  ENSG00000182871  ENSG00000142156  \\\n",
       "0             -0.098407        -0.161561        -0.272683        -0.335242   \n",
       "1             -0.130186        -0.015604        -0.272482        -0.286995   \n",
       "2             -0.114936        -0.137211        -0.249522        -0.229784   \n",
       "3             -0.117161        -0.117210        -0.253882        -0.241652   \n",
       "4             -0.115394        -0.160236        -0.238282        -0.189923   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103        -0.118431        -0.044981        -0.283560        -0.343194   \n",
       "193104        -0.092786        -0.131069        -0.297897        -0.431532   \n",
       "193105        -0.132254         0.001419        -0.275834        -0.295581   \n",
       "193106        -0.096802        -0.117587        -0.295652        -0.417698   \n",
       "193107        -0.127987        -0.012903        -0.278219        -0.310277   \n",
       "\n",
       "        ENSG00000142173  \n",
       "0             -0.350405  \n",
       "1             -0.333254  \n",
       "2             -0.271228  \n",
       "3             -0.283356  \n",
       "4             -0.236741  \n",
       "...                 ...  \n",
       "193103        -0.373127  \n",
       "193104        -0.430158  \n",
       "193105        -0.342390  \n",
       "193106        -0.421226  \n",
       "193107        -0.351877  \n",
       "\n",
       "[193108 rows x 1147 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dExprs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "dExprsgenesToAdd = new_var.loc[new_var['feature_is_filtered'] == True]\n",
    "dExprs = dExprs.reindex(columns=[*dExprs.columns.tolist(), *dExprsgenesToAdd.index.to_list()], fill_value=0.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <th>ENSG00000215014</th>\n",
       "      <th>ENSG00000215912</th>\n",
       "      <th>ENSG00000215788</th>\n",
       "      <th>ENSG00000049249</th>\n",
       "      <th>ENSG00000116285</th>\n",
       "      <th>ENSG00000142583</th>\n",
       "      <th>ENSG00000117154</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000276017</th>\n",
       "      <th>ENSG00000278817</th>\n",
       "      <th>ENSG00000277196</th>\n",
       "      <th>ENSG00000277630</th>\n",
       "      <th>ENSG00000278384</th>\n",
       "      <th>ENSG00000278633</th>\n",
       "      <th>ENSG00000276345</th>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <th>ENSG00000271254</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.723888</td>\n",
       "      <td>-0.139995</td>\n",
       "      <td>-0.140796</td>\n",
       "      <td>-0.065739</td>\n",
       "      <td>-0.083806</td>\n",
       "      <td>-0.140786</td>\n",
       "      <td>-0.075423</td>\n",
       "      <td>-0.462988</td>\n",
       "      <td>-0.076909</td>\n",
       "      <td>-0.092391</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.197353</td>\n",
       "      <td>-0.091179</td>\n",
       "      <td>-0.143029</td>\n",
       "      <td>-0.062913</td>\n",
       "      <td>-0.085007</td>\n",
       "      <td>-0.159376</td>\n",
       "      <td>-0.072150</td>\n",
       "      <td>-0.477511</td>\n",
       "      <td>-0.061004</td>\n",
       "      <td>-0.119909</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.281528</td>\n",
       "      <td>-0.152716</td>\n",
       "      <td>-0.162285</td>\n",
       "      <td>-0.074706</td>\n",
       "      <td>-0.062901</td>\n",
       "      <td>-0.149338</td>\n",
       "      <td>-0.095287</td>\n",
       "      <td>-0.357703</td>\n",
       "      <td>-0.048933</td>\n",
       "      <td>-0.062510</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.267558</td>\n",
       "      <td>-0.142068</td>\n",
       "      <td>-0.158585</td>\n",
       "      <td>-0.072528</td>\n",
       "      <td>-0.067069</td>\n",
       "      <td>-0.150852</td>\n",
       "      <td>-0.090966</td>\n",
       "      <td>-0.380127</td>\n",
       "      <td>-0.051557</td>\n",
       "      <td>-0.072820</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.298954</td>\n",
       "      <td>-0.170585</td>\n",
       "      <td>-0.172224</td>\n",
       "      <td>-0.079752</td>\n",
       "      <td>-0.052425</td>\n",
       "      <td>-0.149062</td>\n",
       "      <td>-0.105750</td>\n",
       "      <td>-0.302920</td>\n",
       "      <td>-0.039100</td>\n",
       "      <td>-0.041369</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193103</th>\n",
       "      <td>-0.214100</td>\n",
       "      <td>-0.091046</td>\n",
       "      <td>5.200045</td>\n",
       "      <td>-0.058977</td>\n",
       "      <td>-0.094838</td>\n",
       "      <td>-0.153033</td>\n",
       "      <td>9.562877</td>\n",
       "      <td>-0.525991</td>\n",
       "      <td>-0.076291</td>\n",
       "      <td>-0.130821</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193104</th>\n",
       "      <td>-0.267087</td>\n",
       "      <td>-0.107013</td>\n",
       "      <td>-0.118204</td>\n",
       "      <td>-0.054842</td>\n",
       "      <td>-0.107104</td>\n",
       "      <td>-0.138718</td>\n",
       "      <td>-0.052452</td>\n",
       "      <td>-0.105977</td>\n",
       "      <td>-0.101240</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193105</th>\n",
       "      <td>-0.185541</td>\n",
       "      <td>-0.082441</td>\n",
       "      <td>-0.140208</td>\n",
       "      <td>-0.061206</td>\n",
       "      <td>-0.088227</td>\n",
       "      <td>-0.160748</td>\n",
       "      <td>-0.068790</td>\n",
       "      <td>-0.494921</td>\n",
       "      <td>-0.062845</td>\n",
       "      <td>-0.128148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193106</th>\n",
       "      <td>-0.258789</td>\n",
       "      <td>-0.104512</td>\n",
       "      <td>-0.120443</td>\n",
       "      <td>-0.055490</td>\n",
       "      <td>-0.105183</td>\n",
       "      <td>-0.140960</td>\n",
       "      <td>-0.054115</td>\n",
       "      <td>0.032701</td>\n",
       "      <td>-0.097333</td>\n",
       "      <td>-0.135005</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193107</th>\n",
       "      <td>-0.194356</td>\n",
       "      <td>-0.085097</td>\n",
       "      <td>-0.137830</td>\n",
       "      <td>-0.060518</td>\n",
       "      <td>-0.090268</td>\n",
       "      <td>-0.158367</td>\n",
       "      <td>-0.067024</td>\n",
       "      <td>2.282134</td>\n",
       "      <td>-0.066995</td>\n",
       "      <td>-0.128973</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193108 rows  20922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ENSG00000188290  ENSG00000186891  ENSG00000186827  ENSG00000215014  \\\n",
       "0              2.723888        -0.139995        -0.140796        -0.065739   \n",
       "1             -0.197353        -0.091179        -0.143029        -0.062913   \n",
       "2             -0.281528        -0.152716        -0.162285        -0.074706   \n",
       "3             -0.267558        -0.142068        -0.158585        -0.072528   \n",
       "4             -0.298954        -0.170585        -0.172224        -0.079752   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103        -0.214100        -0.091046         5.200045        -0.058977   \n",
       "193104        -0.267087        -0.107013        -0.118204        -0.054842   \n",
       "193105        -0.185541        -0.082441        -0.140208        -0.061206   \n",
       "193106        -0.258789        -0.104512        -0.120443        -0.055490   \n",
       "193107        -0.194356        -0.085097        -0.137830        -0.060518   \n",
       "\n",
       "        ENSG00000215912  ENSG00000215788  ENSG00000049249  ENSG00000116285  \\\n",
       "0             -0.083806        -0.140786        -0.075423        -0.462988   \n",
       "1             -0.085007        -0.159376        -0.072150        -0.477511   \n",
       "2             -0.062901        -0.149338        -0.095287        -0.357703   \n",
       "3             -0.067069        -0.150852        -0.090966        -0.380127   \n",
       "4             -0.052425        -0.149062        -0.105750        -0.302920   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103        -0.094838        -0.153033         9.562877        -0.525991   \n",
       "193104        -0.107104        -0.138718        -0.052452        -0.105977   \n",
       "193105        -0.088227        -0.160748        -0.068790        -0.494921   \n",
       "193106        -0.105183        -0.140960        -0.054115         0.032701   \n",
       "193107        -0.090268        -0.158367        -0.067024         2.282134   \n",
       "\n",
       "        ENSG00000142583  ENSG00000117154  ...  ENSG00000276017  \\\n",
       "0             -0.076909        -0.092391  ...              0.0   \n",
       "1             -0.061004        -0.119909  ...              0.0   \n",
       "2             -0.048933        -0.062510  ...              0.0   \n",
       "3             -0.051557        -0.072820  ...              0.0   \n",
       "4             -0.039100        -0.041369  ...              0.0   \n",
       "...                 ...              ...  ...              ...   \n",
       "193103        -0.076291        -0.130821  ...              0.0   \n",
       "193104        -0.101240        10.000000  ...              0.0   \n",
       "193105        -0.062845        -0.128148  ...              0.0   \n",
       "193106        -0.097333        -0.135005  ...              0.0   \n",
       "193107        -0.066995        -0.128973  ...              0.0   \n",
       "\n",
       "        ENSG00000278817  ENSG00000277196  ENSG00000277630  ENSG00000278384  \\\n",
       "0                   0.0              0.0              0.0              0.0   \n",
       "1                   0.0              0.0              0.0              0.0   \n",
       "2                   0.0              0.0              0.0              0.0   \n",
       "3                   0.0              0.0              0.0              0.0   \n",
       "4                   0.0              0.0              0.0              0.0   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103              0.0              0.0              0.0              0.0   \n",
       "193104              0.0              0.0              0.0              0.0   \n",
       "193105              0.0              0.0              0.0              0.0   \n",
       "193106              0.0              0.0              0.0              0.0   \n",
       "193107              0.0              0.0              0.0              0.0   \n",
       "\n",
       "        ENSG00000278633  ENSG00000276345  ENSG00000277856  ENSG00000275063  \\\n",
       "0                   0.0              0.0              0.0              0.0   \n",
       "1                   0.0              0.0              0.0              0.0   \n",
       "2                   0.0              0.0              0.0              0.0   \n",
       "3                   0.0              0.0              0.0              0.0   \n",
       "4                   0.0              0.0              0.0              0.0   \n",
       "...                 ...              ...              ...              ...   \n",
       "193103              0.0              0.0              0.0              0.0   \n",
       "193104              0.0              0.0              0.0              0.0   \n",
       "193105              0.0              0.0              0.0              0.0   \n",
       "193106              0.0              0.0              0.0              0.0   \n",
       "193107              0.0              0.0              0.0              0.0   \n",
       "\n",
       "        ENSG00000271254  \n",
       "0                   0.0  \n",
       "1                   0.0  \n",
       "2                   0.0  \n",
       "3                   0.0  \n",
       "4                   0.0  \n",
       "...                 ...  \n",
       "193103              0.0  \n",
       "193104              0.0  \n",
       "193105              0.0  \n",
       "193106              0.0  \n",
       "193107              0.0  \n",
       "\n",
       "[193108 rows x 20922 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dExprs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Similar checks for raw.var, if present**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "rExprs = pd.DataFrame(data=raw_adata.X.toarray())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "rExprs = rExprs.set_axis(rvar.index.to_list(), axis = 1, inplace=False)\n",
    "rExprs = rExprs.reindex(columns = dExprs.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "rvar = rvar.reindex(index=dExprs.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells_by_counts-0</th>\n",
       "      <th>mean_counts-0</th>\n",
       "      <th>pct_dropout_by_counts-0</th>\n",
       "      <th>total_counts-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells_by_counts-1</th>\n",
       "      <th>mean_counts-1</th>\n",
       "      <th>pct_dropout_by_counts-1</th>\n",
       "      <th>total_counts-1</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>False</td>\n",
       "      <td>HES4</td>\n",
       "      <td>16580</td>\n",
       "      <td>16580</td>\n",
       "      <td>0.302786</td>\n",
       "      <td>88.689619</td>\n",
       "      <td>44385.750000</td>\n",
       "      <td>5369</td>\n",
       "      <td>5369</td>\n",
       "      <td>0.047040</td>\n",
       "      <td>95.454930</td>\n",
       "      <td>5556.719727</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>18879</td>\n",
       "      <td>True</td>\n",
       "      <td>0.348015</td>\n",
       "      <td>2.407703</td>\n",
       "      <td>0.858933</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>5391</td>\n",
       "      <td>5391</td>\n",
       "      <td>0.055908</td>\n",
       "      <td>96.322421</td>\n",
       "      <td>8195.559570</td>\n",
       "      <td>570</td>\n",
       "      <td>570</td>\n",
       "      <td>0.004692</td>\n",
       "      <td>99.517473</td>\n",
       "      <td>554.286743</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5383</td>\n",
       "      <td>True</td>\n",
       "      <td>0.090583</td>\n",
       "      <td>2.411408</td>\n",
       "      <td>0.957952</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113</td>\n",
       "      <td>0.054750</td>\n",
       "      <td>96.512064</td>\n",
       "      <td>8025.812988</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "      <td>0.004022</td>\n",
       "      <td>99.558953</td>\n",
       "      <td>475.164764</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5170</td>\n",
       "      <td>True</td>\n",
       "      <td>0.103728</td>\n",
       "      <td>2.507761</td>\n",
       "      <td>1.110371</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215014</th>\n",
       "      <td>False</td>\n",
       "      <td>AL645728.1</td>\n",
       "      <td>1166</td>\n",
       "      <td>1166</td>\n",
       "      <td>0.008344</td>\n",
       "      <td>99.204590</td>\n",
       "      <td>1223.174438</td>\n",
       "      <td>367</td>\n",
       "      <td>367</td>\n",
       "      <td>0.002795</td>\n",
       "      <td>99.689320</td>\n",
       "      <td>330.200958</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1251</td>\n",
       "      <td>True</td>\n",
       "      <td>0.020191</td>\n",
       "      <td>2.200457</td>\n",
       "      <td>0.519287</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215912</th>\n",
       "      <td>False</td>\n",
       "      <td>TTC34</td>\n",
       "      <td>707</td>\n",
       "      <td>707</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>99.517706</td>\n",
       "      <td>853.601624</td>\n",
       "      <td>2557</td>\n",
       "      <td>2557</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>97.835399</td>\n",
       "      <td>4058.525635</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1571</td>\n",
       "      <td>True</td>\n",
       "      <td>0.040115</td>\n",
       "      <td>2.175078</td>\n",
       "      <td>0.452723</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278633</th>\n",
       "      <td>False</td>\n",
       "      <td>AC023491.2</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>99.976124</td>\n",
       "      <td>38.998932</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>99.977143</td>\n",
       "      <td>26.716158</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>1.519368</td>\n",
       "      <td>-0.584777</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000276345</th>\n",
       "      <td>False</td>\n",
       "      <td>AC004556.1</td>\n",
       "      <td>2975</td>\n",
       "      <td>2975</td>\n",
       "      <td>0.025639</td>\n",
       "      <td>97.970544</td>\n",
       "      <td>3758.375000</td>\n",
       "      <td>354</td>\n",
       "      <td>354</td>\n",
       "      <td>0.002255</td>\n",
       "      <td>99.700325</td>\n",
       "      <td>266.433228</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2949</td>\n",
       "      <td>False</td>\n",
       "      <td>0.040713</td>\n",
       "      <td>2.134076</td>\n",
       "      <td>0.372613</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>False</td>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>104</td>\n",
       "      <td>104</td>\n",
       "      <td>0.064795</td>\n",
       "      <td>99.929054</td>\n",
       "      <td>9498.426758</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>99.991535</td>\n",
       "      <td>24.979130</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>90</td>\n",
       "      <td>False</td>\n",
       "      <td>0.021040</td>\n",
       "      <td>4.567000</td>\n",
       "      <td>4.135742</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>False</td>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>0.135452</td>\n",
       "      <td>99.913364</td>\n",
       "      <td>19856.105469</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>99.989842</td>\n",
       "      <td>12.580947</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>113</td>\n",
       "      <td>False</td>\n",
       "      <td>0.029828</td>\n",
       "      <td>4.531441</td>\n",
       "      <td>3.646755</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>False</td>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>3564</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.026531</td>\n",
       "      <td>97.568746</td>\n",
       "      <td>3889.149414</td>\n",
       "      <td>1964</td>\n",
       "      <td>1964</td>\n",
       "      <td>0.016366</td>\n",
       "      <td>98.337397</td>\n",
       "      <td>1933.262695</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4611</td>\n",
       "      <td>False</td>\n",
       "      <td>0.056425</td>\n",
       "      <td>1.857525</td>\n",
       "      <td>-0.079161</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20922 rows  20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mt  gene_ids-0  n_cells-0  n_cells_by_counts-0  \\\n",
       "ENSG00000188290  False        HES4      16580                16580   \n",
       "ENSG00000186891  False    TNFRSF18       5391                 5391   \n",
       "ENSG00000186827  False     TNFRSF4       5113                 5113   \n",
       "ENSG00000215014  False  AL645728.1       1166                 1166   \n",
       "ENSG00000215912  False       TTC34        707                  707   \n",
       "...                ...         ...        ...                  ...   \n",
       "ENSG00000278633  False  AC023491.2         35                   35   \n",
       "ENSG00000276345  False  AC004556.1       2975                 2975   \n",
       "ENSG00000277856  False  AC233755.2        104                  104   \n",
       "ENSG00000275063  False  AC233755.1        127                  127   \n",
       "ENSG00000271254  False  AC240274.1       3564                 3564   \n",
       "\n",
       "                 mean_counts-0  pct_dropout_by_counts-0  total_counts-0  \\\n",
       "ENSG00000188290       0.302786                88.689619    44385.750000   \n",
       "ENSG00000186891       0.055908                96.322421     8195.559570   \n",
       "ENSG00000186827       0.054750                96.512064     8025.812988   \n",
       "ENSG00000215014       0.008344                99.204590     1223.174438   \n",
       "ENSG00000215912       0.005823                99.517706      853.601624   \n",
       "...                        ...                      ...             ...   \n",
       "ENSG00000278633       0.000266                99.976124       38.998932   \n",
       "ENSG00000276345       0.025639                97.970544     3758.375000   \n",
       "ENSG00000277856       0.064795                99.929054     9498.426758   \n",
       "ENSG00000275063       0.135452                99.913364    19856.105469   \n",
       "ENSG00000271254       0.026531                97.568746     3889.149414   \n",
       "\n",
       "                 n_cells-1  n_cells_by_counts-1  mean_counts-1  \\\n",
       "ENSG00000188290       5369                 5369       0.047040   \n",
       "ENSG00000186891        570                  570       0.004692   \n",
       "ENSG00000186827        521                  521       0.004022   \n",
       "ENSG00000215014        367                  367       0.002795   \n",
       "ENSG00000215912       2557                 2557       0.034357   \n",
       "...                    ...                  ...            ...   \n",
       "ENSG00000278633         27                   27       0.000226   \n",
       "ENSG00000276345        354                  354       0.002255   \n",
       "ENSG00000277856         10                   10       0.000211   \n",
       "ENSG00000275063         12                   12       0.000107   \n",
       "ENSG00000271254       1964                 1964       0.016366   \n",
       "\n",
       "                 pct_dropout_by_counts-1  total_counts-1  feature_types-1  \\\n",
       "ENSG00000188290                95.454930     5556.719727  Gene Expression   \n",
       "ENSG00000186891                99.517473      554.286743  Gene Expression   \n",
       "ENSG00000186827                99.558953      475.164764  Gene Expression   \n",
       "ENSG00000215014                99.689320      330.200958  Gene Expression   \n",
       "ENSG00000215912                97.835399     4058.525635  Gene Expression   \n",
       "...                                  ...             ...              ...   \n",
       "ENSG00000278633                99.977143       26.716158  Gene Expression   \n",
       "ENSG00000276345                99.700325      266.433228  Gene Expression   \n",
       "ENSG00000277856                99.991535       24.979130  Gene Expression   \n",
       "ENSG00000275063                99.989842       12.580947  Gene Expression   \n",
       "ENSG00000271254                98.337397     1933.262695  Gene Expression   \n",
       "\n",
       "                 n_cells  highly_variable     means  dispersions  \\\n",
       "ENSG00000188290    18879             True  0.348015     2.407703   \n",
       "ENSG00000186891     5383             True  0.090583     2.411408   \n",
       "ENSG00000186827     5170             True  0.103728     2.507761   \n",
       "ENSG00000215014     1251             True  0.020191     2.200457   \n",
       "ENSG00000215912     1571             True  0.040115     2.175078   \n",
       "...                  ...              ...       ...          ...   \n",
       "ENSG00000278633       40            False  0.000331     1.519368   \n",
       "ENSG00000276345     2949            False  0.040713     2.134076   \n",
       "ENSG00000277856       90            False  0.021040     4.567000   \n",
       "ENSG00000275063      113            False  0.029828     4.531441   \n",
       "ENSG00000271254     4611            False  0.056425     1.857525   \n",
       "\n",
       "                 dispersions_norm  highly_variable_nbatches  \\\n",
       "ENSG00000188290          0.858933                         2   \n",
       "ENSG00000186891          0.957952                         2   \n",
       "ENSG00000186827          1.110371                         2   \n",
       "ENSG00000215014          0.519287                         2   \n",
       "ENSG00000215912          0.452723                         2   \n",
       "...                           ...                       ...   \n",
       "ENSG00000278633         -0.584777                         0   \n",
       "ENSG00000276345          0.372613                         1   \n",
       "ENSG00000277856          4.135742                         1   \n",
       "ENSG00000275063          3.646755                         1   \n",
       "ENSG00000271254         -0.079161                         0   \n",
       "\n",
       "                 highly_variable_intersection  \n",
       "ENSG00000188290                          True  \n",
       "ENSG00000186891                          True  \n",
       "ENSG00000186827                          True  \n",
       "ENSG00000215014                          True  \n",
       "ENSG00000215912                          True  \n",
       "...                                       ...  \n",
       "ENSG00000278633                         False  \n",
       "ENSG00000276345                         False  \n",
       "ENSG00000277856                         False  \n",
       "ENSG00000275063                         False  \n",
       "ENSG00000271254                         False  \n",
       "\n",
       "[20922 rows x 20 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rExprs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dExprs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter out not-approved genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Create the list of approved IDs to filter on**<br>\n",
    "*For the initial run, download the 4 genes_ csv files from https://github.com/chanzuckerberg/single-cell-curation/tree/main/cellxgene_schema_cli/cellxgene_schema/ontology_files*<br>\n",
    "*After that, if the `genes_approved.csv` is available locally, then the 4 genes_ files won't be necessary*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_files = [\n",
    "    'genes_ercc.csv',\n",
    "    'genes_homo_sapiens.csv',\n",
    "    'genes_mus_musculus.csv',\n",
    "    'genes_sars_cov_2.csv'\n",
    "]\n",
    "\n",
    "if not os.path.exists('genes_approved.csv'):\n",
    "    ids = pd.DataFrame()\n",
    "    for f in ref_files:\n",
    "        df = pd.read_csv(f, names=['feature_id','symb','num','length'],dtype='str',index_col=False)\n",
    "        ids = ids.append(df)\n",
    "        os.remove(f)\n",
    "    ids.to_csv('genes_approved.csv', index=False)\n",
    "\n",
    "approved = pd.read_csv('genes_approved.csv',dtype='str')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sample:** set columns with all the same values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['feature_is_filtered'] = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove any fields (typically symbols as the portal will add those)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_remove = [\n",
    "    'gene_symbols'\n",
    "]\n",
    "\n",
    "adata.var.drop(columns=var_remove, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Map the Ensembl IDs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var = adata.var.merge(var_map_df,left_index=True,right_on='gene_symbols',how='left').set_index(adata.var.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Filter out genes that don't appear in the approved annotation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_to_keep = adata.var.index.tolist()\n",
    "var_in_approved = adata.var.index[adata.var['gene_ids'].isin(approved['feature_id'])].tolist()\n",
    "var_to_keep = [e for e in var_to_keep if e in var_in_approved]\n",
    "adata = adata[:, var_to_keep]\n",
    "adata.var.set_index('gene_ids',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Repeat much of the same steps for the `raw.var`, if it exists**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells_by_counts-0</th>\n",
       "      <th>mean_counts-0</th>\n",
       "      <th>pct_dropout_by_counts-0</th>\n",
       "      <th>total_counts-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells_by_counts-1</th>\n",
       "      <th>mean_counts-1</th>\n",
       "      <th>pct_dropout_by_counts-1</th>\n",
       "      <th>total_counts-1</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_ids-1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>False</td>\n",
       "      <td>HES4</td>\n",
       "      <td>16580</td>\n",
       "      <td>16580</td>\n",
       "      <td>0.302786</td>\n",
       "      <td>88.689619</td>\n",
       "      <td>44385.750000</td>\n",
       "      <td>5369</td>\n",
       "      <td>5369</td>\n",
       "      <td>0.047040</td>\n",
       "      <td>95.454930</td>\n",
       "      <td>5556.719727</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>18879</td>\n",
       "      <td>True</td>\n",
       "      <td>0.348015</td>\n",
       "      <td>2.407703</td>\n",
       "      <td>0.858933</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>5391</td>\n",
       "      <td>5391</td>\n",
       "      <td>0.055908</td>\n",
       "      <td>96.322421</td>\n",
       "      <td>8195.559570</td>\n",
       "      <td>570</td>\n",
       "      <td>570</td>\n",
       "      <td>0.004692</td>\n",
       "      <td>99.517473</td>\n",
       "      <td>554.286743</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5383</td>\n",
       "      <td>True</td>\n",
       "      <td>0.090583</td>\n",
       "      <td>2.411408</td>\n",
       "      <td>0.957952</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113</td>\n",
       "      <td>0.054750</td>\n",
       "      <td>96.512064</td>\n",
       "      <td>8025.812988</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "      <td>0.004022</td>\n",
       "      <td>99.558953</td>\n",
       "      <td>475.164764</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5170</td>\n",
       "      <td>True</td>\n",
       "      <td>0.103728</td>\n",
       "      <td>2.507761</td>\n",
       "      <td>1.110371</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215014</th>\n",
       "      <td>False</td>\n",
       "      <td>AL645728.1</td>\n",
       "      <td>1166</td>\n",
       "      <td>1166</td>\n",
       "      <td>0.008344</td>\n",
       "      <td>99.204590</td>\n",
       "      <td>1223.174438</td>\n",
       "      <td>367</td>\n",
       "      <td>367</td>\n",
       "      <td>0.002795</td>\n",
       "      <td>99.689320</td>\n",
       "      <td>330.200958</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1251</td>\n",
       "      <td>True</td>\n",
       "      <td>0.020191</td>\n",
       "      <td>2.200457</td>\n",
       "      <td>0.519287</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215912</th>\n",
       "      <td>False</td>\n",
       "      <td>TTC34</td>\n",
       "      <td>707</td>\n",
       "      <td>707</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>99.517706</td>\n",
       "      <td>853.601624</td>\n",
       "      <td>2557</td>\n",
       "      <td>2557</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>97.835399</td>\n",
       "      <td>4058.525635</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>1571</td>\n",
       "      <td>True</td>\n",
       "      <td>0.040115</td>\n",
       "      <td>2.175078</td>\n",
       "      <td>0.452723</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160185</th>\n",
       "      <td>False</td>\n",
       "      <td>UBASH3A</td>\n",
       "      <td>2057</td>\n",
       "      <td>2057</td>\n",
       "      <td>0.015399</td>\n",
       "      <td>98.596776</td>\n",
       "      <td>2257.323975</td>\n",
       "      <td>1330</td>\n",
       "      <td>1330</td>\n",
       "      <td>0.011449</td>\n",
       "      <td>98.874103</td>\n",
       "      <td>1352.410767</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2619</td>\n",
       "      <td>True</td>\n",
       "      <td>0.076424</td>\n",
       "      <td>2.345723</td>\n",
       "      <td>0.911611</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160188</th>\n",
       "      <td>False</td>\n",
       "      <td>RSPH1</td>\n",
       "      <td>3765</td>\n",
       "      <td>3765</td>\n",
       "      <td>0.404490</td>\n",
       "      <td>97.431629</td>\n",
       "      <td>59294.632812</td>\n",
       "      <td>3671</td>\n",
       "      <td>3671</td>\n",
       "      <td>0.087606</td>\n",
       "      <td>96.892354</td>\n",
       "      <td>10348.668945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5024</td>\n",
       "      <td>True</td>\n",
       "      <td>0.237395</td>\n",
       "      <td>2.757026</td>\n",
       "      <td>1.647012</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000182871</th>\n",
       "      <td>False</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>13232</td>\n",
       "      <td>13232</td>\n",
       "      <td>0.149136</td>\n",
       "      <td>90.973525</td>\n",
       "      <td>21861.937500</td>\n",
       "      <td>6512</td>\n",
       "      <td>6512</td>\n",
       "      <td>0.064490</td>\n",
       "      <td>94.487336</td>\n",
       "      <td>7618.034180</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>16562</td>\n",
       "      <td>True</td>\n",
       "      <td>0.321884</td>\n",
       "      <td>2.337140</td>\n",
       "      <td>0.801880</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142156</th>\n",
       "      <td>False</td>\n",
       "      <td>COL6A1</td>\n",
       "      <td>14127</td>\n",
       "      <td>14127</td>\n",
       "      <td>0.513421</td>\n",
       "      <td>90.362983</td>\n",
       "      <td>75262.882812</td>\n",
       "      <td>10117</td>\n",
       "      <td>10117</td>\n",
       "      <td>0.120751</td>\n",
       "      <td>91.435561</td>\n",
       "      <td>14264.091797</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>19712</td>\n",
       "      <td>True</td>\n",
       "      <td>0.514242</td>\n",
       "      <td>2.488781</td>\n",
       "      <td>0.826094</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142173</th>\n",
       "      <td>False</td>\n",
       "      <td>COL6A2</td>\n",
       "      <td>16267</td>\n",
       "      <td>16267</td>\n",
       "      <td>0.612974</td>\n",
       "      <td>88.903139</td>\n",
       "      <td>89856.523438</td>\n",
       "      <td>10974</td>\n",
       "      <td>10974</td>\n",
       "      <td>0.138058</td>\n",
       "      <td>90.710077</td>\n",
       "      <td>16308.473633</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>22146</td>\n",
       "      <td>True</td>\n",
       "      <td>0.657928</td>\n",
       "      <td>2.699069</td>\n",
       "      <td>1.050830</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1147 rows  20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mt  gene_ids-0  n_cells-0  n_cells_by_counts-0  \\\n",
       "gene_ids-1                                                           \n",
       "ENSG00000188290  False        HES4      16580                16580   \n",
       "ENSG00000186891  False    TNFRSF18       5391                 5391   \n",
       "ENSG00000186827  False     TNFRSF4       5113                 5113   \n",
       "ENSG00000215014  False  AL645728.1       1166                 1166   \n",
       "ENSG00000215912  False       TTC34        707                  707   \n",
       "...                ...         ...        ...                  ...   \n",
       "ENSG00000160185  False     UBASH3A       2057                 2057   \n",
       "ENSG00000160188  False       RSPH1       3765                 3765   \n",
       "ENSG00000182871  False     COL18A1      13232                13232   \n",
       "ENSG00000142156  False      COL6A1      14127                14127   \n",
       "ENSG00000142173  False      COL6A2      16267                16267   \n",
       "\n",
       "                 mean_counts-0  pct_dropout_by_counts-0  total_counts-0  \\\n",
       "gene_ids-1                                                                \n",
       "ENSG00000188290       0.302786                88.689619    44385.750000   \n",
       "ENSG00000186891       0.055908                96.322421     8195.559570   \n",
       "ENSG00000186827       0.054750                96.512064     8025.812988   \n",
       "ENSG00000215014       0.008344                99.204590     1223.174438   \n",
       "ENSG00000215912       0.005823                99.517706      853.601624   \n",
       "...                        ...                      ...             ...   \n",
       "ENSG00000160185       0.015399                98.596776     2257.323975   \n",
       "ENSG00000160188       0.404490                97.431629    59294.632812   \n",
       "ENSG00000182871       0.149136                90.973525    21861.937500   \n",
       "ENSG00000142156       0.513421                90.362983    75262.882812   \n",
       "ENSG00000142173       0.612974                88.903139    89856.523438   \n",
       "\n",
       "                 n_cells-1  n_cells_by_counts-1  mean_counts-1  \\\n",
       "gene_ids-1                                                       \n",
       "ENSG00000188290       5369                 5369       0.047040   \n",
       "ENSG00000186891        570                  570       0.004692   \n",
       "ENSG00000186827        521                  521       0.004022   \n",
       "ENSG00000215014        367                  367       0.002795   \n",
       "ENSG00000215912       2557                 2557       0.034357   \n",
       "...                    ...                  ...            ...   \n",
       "ENSG00000160185       1330                 1330       0.011449   \n",
       "ENSG00000160188       3671                 3671       0.087606   \n",
       "ENSG00000182871       6512                 6512       0.064490   \n",
       "ENSG00000142156      10117                10117       0.120751   \n",
       "ENSG00000142173      10974                10974       0.138058   \n",
       "\n",
       "                 pct_dropout_by_counts-1  total_counts-1  feature_types-1  \\\n",
       "gene_ids-1                                                                  \n",
       "ENSG00000188290                95.454930     5556.719727  Gene Expression   \n",
       "ENSG00000186891                99.517473      554.286743  Gene Expression   \n",
       "ENSG00000186827                99.558953      475.164764  Gene Expression   \n",
       "ENSG00000215014                99.689320      330.200958  Gene Expression   \n",
       "ENSG00000215912                97.835399     4058.525635  Gene Expression   \n",
       "...                                  ...             ...              ...   \n",
       "ENSG00000160185                98.874103     1352.410767  Gene Expression   \n",
       "ENSG00000160188                96.892354    10348.668945  Gene Expression   \n",
       "ENSG00000182871                94.487336     7618.034180  Gene Expression   \n",
       "ENSG00000142156                91.435561    14264.091797  Gene Expression   \n",
       "ENSG00000142173                90.710077    16308.473633  Gene Expression   \n",
       "\n",
       "                 n_cells  highly_variable     means  dispersions  \\\n",
       "gene_ids-1                                                         \n",
       "ENSG00000188290    18879             True  0.348015     2.407703   \n",
       "ENSG00000186891     5383             True  0.090583     2.411408   \n",
       "ENSG00000186827     5170             True  0.103728     2.507761   \n",
       "ENSG00000215014     1251             True  0.020191     2.200457   \n",
       "ENSG00000215912     1571             True  0.040115     2.175078   \n",
       "...                  ...              ...       ...          ...   \n",
       "ENSG00000160185     2619             True  0.076424     2.345723   \n",
       "ENSG00000160188     5024             True  0.237395     2.757026   \n",
       "ENSG00000182871    16562             True  0.321884     2.337140   \n",
       "ENSG00000142156    19712             True  0.514242     2.488781   \n",
       "ENSG00000142173    22146             True  0.657928     2.699069   \n",
       "\n",
       "                 dispersions_norm  highly_variable_nbatches  \\\n",
       "gene_ids-1                                                    \n",
       "ENSG00000188290          0.858933                         2   \n",
       "ENSG00000186891          0.957952                         2   \n",
       "ENSG00000186827          1.110371                         2   \n",
       "ENSG00000215014          0.519287                         2   \n",
       "ENSG00000215912          0.452723                         2   \n",
       "...                           ...                       ...   \n",
       "ENSG00000160185          0.911611                         2   \n",
       "ENSG00000160188          1.647012                         2   \n",
       "ENSG00000182871          0.801880                         2   \n",
       "ENSG00000142156          0.826094                         2   \n",
       "ENSG00000142173          1.050830                         2   \n",
       "\n",
       "                 highly_variable_intersection  \n",
       "gene_ids-1                                     \n",
       "ENSG00000188290                          True  \n",
       "ENSG00000186891                          True  \n",
       "ENSG00000186827                          True  \n",
       "ENSG00000215014                          True  \n",
       "ENSG00000215912                          True  \n",
       "...                                       ...  \n",
       "ENSG00000160185                          True  \n",
       "ENSG00000160188                          True  \n",
       "ENSG00000182871                          True  \n",
       "ENSG00000142156                          True  \n",
       "ENSG00000142173                          True  \n",
       "\n",
       "[1147 rows x 20 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000188290', 'ENSG00000186891', 'ENSG00000186827',\n",
       "       'ENSG00000215014', 'ENSG00000215912', 'ENSG00000215788',\n",
       "       'ENSG00000049249', 'ENSG00000116285', 'ENSG00000142583',\n",
       "       'ENSG00000117154',\n",
       "       ...\n",
       "       'ENSG00000276017', 'ENSG00000278817', 'ENSG00000277196',\n",
       "       'ENSG00000277630', 'ENSG00000278384', 'ENSG00000278633',\n",
       "       'ENSG00000276345', 'ENSG00000277856', 'ENSG00000275063',\n",
       "       'ENSG00000271254'],\n",
       "      dtype='object', length=20922)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rExprs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000188290', 'ENSG00000186891', 'ENSG00000186827',\n",
       "       'ENSG00000215014', 'ENSG00000215912', 'ENSG00000215788',\n",
       "       'ENSG00000049249', 'ENSG00000116285', 'ENSG00000142583',\n",
       "       'ENSG00000117154',\n",
       "       ...\n",
       "       'ENSG00000234883', 'ENSG00000154734', 'ENSG00000231106',\n",
       "       'ENSG00000160180', 'ENSG00000160182', 'ENSG00000160185',\n",
       "       'ENSG00000160188', 'ENSG00000182871', 'ENSG00000142156',\n",
       "       'ENSG00000142173'],\n",
       "      dtype='object', name='gene_ids-1', length=1147)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rvar.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_var_remove = [\n",
    "    'gene_ids-0'\n",
    "]\n",
    "rvar.drop(columns=raw_var_remove, inplace=True)\n",
    "\n",
    "raw_adata = ad.AnnData(rExprs, var=rvar, obs=adata.obs)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'var_map_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [62]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m raw_adata\u001b[38;5;241m.\u001b[39mvar \u001b[38;5;241m=\u001b[39m raw_adata\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mmerge(\u001b[43mvar_map_df\u001b[49m,left_index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,right_on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_symbols\u001b[39m\u001b[38;5;124m'\u001b[39m,how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mset_index(raw_adata\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mindex)\n\u001b[1;32m      3\u001b[0m raw_adata \u001b[38;5;241m=\u001b[39m raw_adata[:, var_to_keep]\n\u001b[1;32m      4\u001b[0m raw_adata\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_ids\u001b[39m\u001b[38;5;124m'\u001b[39m,inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'var_map_df' is not defined"
     ]
    }
   ],
   "source": [
    "raw_adata.var = raw_adata.var.merge(var_map_df,left_index=True,right_on='gene_symbols',how='left').set_index(raw_adata.var.index)\n",
    "\n",
    "raw_adata = raw_adata[:, var_to_keep]\n",
    "raw_adata.var.set_index('gene_ids',inplace=True)\n",
    "adata.raw = raw_adata\n",
    "adata.raw.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate\n",
    "**Determine the embedding by which to plot**\\\n",
    "May need to overwrite if the first obsm is not informative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "default_embedding = adata.uns.get('default_embedding')\n",
    "umap_embedding = None\n",
    "tsne_embdding = None\n",
    "for k in adata.obsm_keys():\n",
    "    if 'umap' in k.lower():\n",
    "        umap_embedding = k\n",
    "    elif 'tsne' in k.lower():\n",
    "        tsne_embdding = k\n",
    "if not default_embedding:\n",
    "    if umap_embedding:\n",
    "        default_embedding = umap_embedding\n",
    "    elif tsne_embdding:\n",
    "        default_embedding = tsne_embdding\n",
    "    else:\n",
    "        default_embedding = adata.obsm_keys()[0]\n",
    "default_embedding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Plot the cells to ensure they cluster by cell type**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi=150)\n",
    "sc.pl.embedding(adata, basis=default_embedding, color=['cell_type_ontology_term_id'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The above plot will set a color palette in uns, so remove that**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata.uns['cell_type_ontology_term_id_colors']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Plot by multiple genes using the normalized counts**<br>\n",
    "*It is best to get a list of genes relevant to the specific data from the contributor/publication*"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "symbol_list = [\n",
    "    'SKAP1', #lymphoid\n",
    "    'RIPOR2', #lymphoid\n",
    "    'THEMIS', #lymphoid\n",
    "    'RBM47', #myeloid\n",
    "    'TBXAS1', #myeloid\n",
    "    'CD163', #myeloid\n",
    "    'VWF', #endothelial\n",
    "    'ANO2', #endothelial\n",
    "    'EGFL7', #endothelial\n",
    "    'BICC1', #fibroblast\n",
    "    'MEG3', #fibroblast\n",
    "    'COL6A3', #fibroblast\n",
    "    'RMB20', #cardiomyocyte\n",
    "    'ACTN2', #cardiomyocyte\n",
    "    'LDB3' #cardiomyocyte\n",
    "]\n",
    "\n",
    "ensg_list = []\n",
    "for s in symbol_list:\n",
    "    if s in approved['symb'].tolist():\n",
    "        ensg_id = approved.loc[approved['symb'] == s, 'feature_id'].iloc[0]\n",
    "        if ensg_id in adata.var.index:\n",
    "            ensg_list.append(ensg_id)\n",
    "        else:\n",
    "            print(f'{s}/{ensg_id} not found in var')\n",
    "    else:\n",
    "        print(f'{s} not found in gene file')\n",
    "\n",
    "ensg_list"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.embedding(adata, basis=default_embedding, color=ensg_list, use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Compare with the same genes using the raw counts to confirm they are correlated**"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.embedding(adata, basis=default_embedding, color=ensg_list, use_raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Additionally, you could compare dotplots of those genes in each cell population**<br>\n",
    "*This will scale all genes based on the max range of any gene so 1 gene with high values may make others difficult to distinguish*"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(adata, ensg_list, 'cell_type_ontology_term_id', use_raw=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(adata, ensg_list, 'cell_type_ontology_term_id', use_raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Write the file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_one = file + '_revised.h5ad'\n",
    "adata.write(filename=new_one, compression='gzip')\n",
    "new_one"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run the CELLxGENE validator**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "validate_process = subprocess.run(['cellxgene-schema', 'validate', new_one], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "for line in validate_process.stdout.decode('utf-8').split('\\n'):\n",
    "    print(line)\n",
    "for line in validate_process.stderr.decode('utf-8').split('\\n'):\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellxgene",
   "language": "python",
   "name": "cellxgene"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
